// oxlint@1.55.0/dist/utils.js
//#region src-js/utils/globals.ts
/**
* Properties of global objects exported as variables.
*
* TSDown will replace e.g. `Object.keys` with import of `ObjectKeys` from this file.
*
* If you use any globals in code in `src-js` directory, you should add them to this file.
*
* See TSDown config file for more details.
*/
const { prototype: ObjectPrototype, hasOwn: ObjectHasOwn, keys: ObjectKeys, values: ObjectValues, freeze: ObjectFreeze, preventExtensions: ObjectPreventExtensions, defineProperty: ObjectDefineProperty, defineProperties: ObjectDefineProperties, create: ObjectCreate, assign: ObjectAssign, getPrototypeOf: ObjectGetPrototypeOf, setPrototypeOf: ObjectSetPrototypeOf, entries: ObjectEntries } = Object, { prototype: ArrayPrototype, isArray: ArrayIsArray, from: ArrayFrom } = Array, { min: MathMin, max: MathMax, floor: MathFloor } = Math, { parse: JSONParse, stringify: JSONStringify } = JSON, { ownKeys: ReflectOwnKeys } = Reflect, { iterator: SymbolIterator } = Symbol, { fromCodePoint: StringFromCodePoint } = String, { now: DateNow } = Date;
//#endregion
//#region src-js/utils/utils.ts
/**
* Get error message from an error.
*
* `err` is expected to be an `Error` object, but can be anything.
*
* * If it's an `Error`, the error message and stack trace is returned.
* * If it's another object with a string `message` property, `message` is returned.
* * Otherwise, a generic "Unknown error" message is returned.
*
* This function will never throw, and always returns a non-empty string, even if:
*
* * `err` is `null` or `undefined`.
* * `err` is an object with a getter for `message` property which throws.
* * `err` has a getter for `stack` or `message` property which returns a different value each time it's accessed.
*
* @param err - Error
* @returns Error message
*/
function getErrorMessage(err) {
	try {
		if (err instanceof Error) {
			let { stack } = err;
			if (typeof stack == "string" && stack !== "") return stack;
		}
		let { message } = err;
		if (typeof message == "string" && message !== "") return message;
	} catch {}
	return "Unknown error";
}
//#endregion
export { ObjectValues as _, JSONParse as a, MathMin as c, ObjectDefineProperty as d, ObjectEntries as f, ObjectPreventExtensions as g, ObjectKeys as h, DateNow as i, ObjectAssign as l, ObjectHasOwn as m, ArrayFrom as n, JSONStringify as o, ObjectFreeze as p, ArrayIsArray as r, MathMax as s, getErrorMessage as t, ObjectCreate as u, StringFromCodePoint as v, SymbolIterator as y };
