// vite-plus@0.1.12/dist/utils/command.js
import spawn from 'cross-spawn';
export async function runCommandSilently(options) {
    const child = spawn(options.command, options.args, {
        stdio: 'pipe',
        cwd: options.cwd,
        env: options.envs,
    });
    const promise = new Promise((resolve, reject) => {
        const stdout = [];
        const stderr = [];
        child.stdout?.on('data', (data) => {
            stdout.push(data);
        });
        child.stderr?.on('data', (data) => {
            stderr.push(data);
        });
        child.on('close', (code) => {
            resolve({
                exitCode: code ?? 0,
                stdout: Buffer.concat(stdout),
                stderr: Buffer.concat(stderr),
            });
        });
        child.on('error', (err) => {
            reject(err);
        });
    });
    return await promise;
}
export async function runCommand(options) {
    const child = spawn(options.command, options.args, {
        stdio: 'inherit',
        cwd: options.cwd,
        env: options.envs,
    });
    return new Promise((resolve, reject) => {
        child.on('close', (code) => {
            resolve(code ?? 0);
        });
        child.on('error', (err) => {
            reject(err);
        });
    });
}
