// vite-plus@0.1.12/dist/utils/terminal.js
import { styleText } from 'node:util';
export function log(message) {
    /* oxlint-disable-next-line no-console */
    console.log(message);
}
export function accent(text) {
    return styleText('blue', text);
}
export function muted(text) {
    return styleText('gray', text);
}
export function success(text) {
    return styleText('green', text);
}
export function error(text) {
    return styleText('red', text);
}
// Standard message prefix functions matching the Rust CLI convention.
// info/note go to stdout (normal output), warn/error go to stderr (diagnostics).
export function infoMsg(msg) {
    /* oxlint-disable-next-line no-console */
    console.log(styleText(['blue', 'bold'], 'info:'), msg);
}
export function warnMsg(msg) {
    /* oxlint-disable-next-line no-console */
    console.error(styleText(['yellow', 'bold'], 'warn:'), msg);
}
export function errorMsg(msg) {
    /* oxlint-disable-next-line no-console */
    console.error(styleText(['red', 'bold'], 'error:'), msg);
}
export function noteMsg(msg) {
    /* oxlint-disable-next-line no-console */
    console.log(styleText(['gray', 'bold'], 'note:'), msg);
}
