// @voidzero-dev/vite-plus-test@0.1.12/dist/browser.js
export { l as loadDiffConfig, a as loadSnapshotSerializers, s as setupCommonEnv, b as startCoverageInsideWorker, c as stopCoverageInsideWorker, t as takeCoverageInsideWorker } from './chunks/setup-common.B41N_kPE.js';
export { T as Traces } from './chunks/traces.CCmnQaNT.js';
export { collectTests, startTests } from './@vitest/runner/index.js';
import * as spyModule from './@vitest/spy/index.js';
export { spyModule as SpyModule };
export { browserFormat, format, inspect, stringify } from './@vitest/utils/display.js';
export { processError } from './@vitest/utils/error.js';
export { getType } from './@vitest/utils/helpers.js';
export { DecodedMap, getOriginalPosition } from './@vitest/utils/source-map.js';
export { getSafeTimers, setSafeTimers } from './@vitest/utils/timers.js';
import './chunks/coverage.D_JHT54q.js';
import './@vitest/snapshot/index.js';
import './chunks/utils.BX5Fg8C4.js';

/**
* @internal
*/
const __INTERNAL = { _extendedMethods: /* @__PURE__ */ new Set() };

export { __INTERNAL };
