// @voidzero-dev/vite-plus-test@0.1.12/dist/module-runner-stub.js
// Browser-safe stub for module-runner
// The real module-runner contains Node.js-only code that crashes browsers
// This stub provides placeholder exports for browser compatibility

// Stub class - browser doesn't actually use these
export class EvaluatedModules {
  constructor() {
    this.idToModuleMap = new Map();
    this.fileToModulesMap = new Map();
    this.urlToIdModuleMap = new Map();
  }
  getModuleById() { return undefined; }
  getModulesByFile() { return []; }
  getModuleByUrl() { return undefined; }
  ensureModule() { return {}; }
  invalidateModule() {}
  clear() {}
}

export class ModuleRunner {
  constructor() {}
  async import() { throw new Error('ModuleRunner is not available in browser'); }
  evaluatedModules = new EvaluatedModules();
}

export class ESModulesEvaluator {
  constructor() {}
  async runExternalModule() { return {}; }
  async runViteModule() { return {}; }
}

// Stub functions
export function createDefaultImportMeta() { return {}; }
export function createNodeImportMeta() { return {}; }
export function createWebSocketModuleRunnerTransport() { return {}; }
export function normalizeModuleId(id) { return id; }

// SSR-related constants (browser doesn't use these)
export const ssrDynamicImportKey = '__vite_ssr_dynamic_import__';
export const ssrExportAllKey = '__vite_ssr_exportAll__';
export const ssrExportNameKey = '__vite_ssr_export__';
export const ssrImportKey = '__vite_ssr_import__';
export const ssrImportMetaKey = '__vite_ssr_import_meta__';
export const ssrModuleExportsKey = '__vite_ssr_exports__';
