// @voidzero-dev/vite-plus-test@0.1.12/dist/reporters.js
export { A as AgentReporter, D as DefaultReporter, a as DotReporter, G as GithubActionsReporter, H as HangingProcessReporter, J as JUnitReporter, b as JsonReporter, R as ReportersMap, T as TapFlatReporter, c as TapReporter, V as VerboseReporter } from './chunks/index.DXMFO5MJ.js';
export { B as BenchmarkReporter, a as BenchmarkReportsMap, V as VerboseBenchmarkReporter } from './chunks/index.CEzQDJGb.js';
import 'node:fs';
import 'node:perf_hooks';
import './@vitest/runner/utils.js';
import './@vitest/utils/helpers.js';
import './@vitest/utils/source-map.js';
import './vendor/pathe.mjs';
import './vendor/tinyrainbow.mjs';
import './chunks/env.D4Lgay0q.js';
import './vendor/std-env.mjs';
import 'node:console';
import 'node:stream';
import 'node:util';
import './@vitest/utils/display.js';
import 'node:os';
import './vendor/tinyexec.mjs';
import './path.js';
import 'node:path';
import 'node:url';
import '@voidzero-dev/vite-plus-core';
import './@vitest/utils/offset.js';
import 'node:fs/promises';
import 'node:module';

console.warn("Importing from \"vitest/reporters\" is deprecated since Vitest 4.1. Please use \"vitest/node\" instead.");
