// vite-plus@0.1.12/dist/global/agent-l4VXJklB.js
import { t as __commonJSMin } from "./chunk-CgnkrU7a.js";
import { o as DEFAULT_ENVS, u as resolve$1 } from "./json-BRdVJ52a.js";
import { n as accent } from "./help-CbTzUdXc.js";
import path from "node:path";
import { stripVTControlCharacters, styleText } from "node:util";
import process$1, { stdin, stdout } from "node:process";
import * as k from "node:readline";
import ot from "node:readline";
import { ReadStream } from "node:tty";
import color from "picocolors";
import fs from "node:fs";
import { downloadPackageManager } from "../../binding/index.js";
import spawn from "cross-spawn";
import fsPromises from "node:fs/promises";
//#endregion
//#region ../../node_modules/.pnpm/@clack+core@1.1.0/node_modules/@clack/core/dist/index.mjs
var import_src = (/* @__PURE__ */ __commonJSMin(((exports, module) => {
	const ESC = "\x1B";
	const CSI = `${ESC}[`;
	const beep = "\x07";
	const cursor = {
		to(x, y) {
			if (!y) return `${CSI}${x + 1}G`;
			return `${CSI}${y + 1};${x + 1}H`;
		},
		move(x, y) {
			let ret = "";
			if (x < 0) ret += `${CSI}${-x}D`;
			else if (x > 0) ret += `${CSI}${x}C`;
			if (y < 0) ret += `${CSI}${-y}A`;
			else if (y > 0) ret += `${CSI}${y}B`;
			return ret;
		},
		up: (count = 1) => `${CSI}${count}A`,
		down: (count = 1) => `${CSI}${count}B`,
		forward: (count = 1) => `${CSI}${count}C`,
		backward: (count = 1) => `${CSI}${count}D`,
		nextLine: (count = 1) => `${CSI}E`.repeat(count),
		prevLine: (count = 1) => `${CSI}F`.repeat(count),
		left: `${CSI}G`,
		hide: `${CSI}?25l`,
		show: `${CSI}?25h`,
		save: `${ESC}7`,
		restore: `${ESC}8`
	};
	module.exports = {
		cursor,
		scroll: {
			up: (count = 1) => `${CSI}S`.repeat(count),
			down: (count = 1) => `${CSI}T`.repeat(count)
		},
		erase: {
			screen: `${CSI}2J`,
			up: (count = 1) => `${CSI}1J`.repeat(count),
			down: (count = 1) => `${CSI}J`.repeat(count),
			line: `${CSI}2K`,
			lineEnd: `${CSI}K`,
			lineStart: `${CSI}1K`,
			lines(count) {
				let clear = "";
				for (let i = 0; i < count; i++) clear += this.line + (i < count - 1 ? cursor.up() : "");
				if (count) clear += cursor.left;
				return clear;
			}
		},
		beep
	};
})))();
function x(t, e, s) {
	if (!s.some((u) => !u.disabled)) return t;
	const i = t + e, r = Math.max(s.length - 1, 0), n = i < 0 ? r : i > r ? 0 : i;
	return s[n].disabled ? x(n, e < 0 ? -1 : 1, s) : n;
}
const at = (t) => t === 161 || t === 164 || t === 167 || t === 168 || t === 170 || t === 173 || t === 174 || t >= 176 && t <= 180 || t >= 182 && t <= 186 || t >= 188 && t <= 191 || t === 198 || t === 208 || t === 215 || t === 216 || t >= 222 && t <= 225 || t === 230 || t >= 232 && t <= 234 || t === 236 || t === 237 || t === 240 || t === 242 || t === 243 || t >= 247 && t <= 250 || t === 252 || t === 254 || t === 257 || t === 273 || t === 275 || t === 283 || t === 294 || t === 295 || t === 299 || t >= 305 && t <= 307 || t === 312 || t >= 319 && t <= 322 || t === 324 || t >= 328 && t <= 331 || t === 333 || t === 338 || t === 339 || t === 358 || t === 359 || t === 363 || t === 462 || t === 464 || t === 466 || t === 468 || t === 470 || t === 472 || t === 474 || t === 476 || t === 593 || t === 609 || t === 708 || t === 711 || t >= 713 && t <= 715 || t === 717 || t === 720 || t >= 728 && t <= 731 || t === 733 || t === 735 || t >= 768 && t <= 879 || t >= 913 && t <= 929 || t >= 931 && t <= 937 || t >= 945 && t <= 961 || t >= 963 && t <= 969 || t === 1025 || t >= 1040 && t <= 1103 || t === 1105 || t === 8208 || t >= 8211 && t <= 8214 || t === 8216 || t === 8217 || t === 8220 || t === 8221 || t >= 8224 && t <= 8226 || t >= 8228 && t <= 8231 || t === 8240 || t === 8242 || t === 8243 || t === 8245 || t === 8251 || t === 8254 || t === 8308 || t === 8319 || t >= 8321 && t <= 8324 || t === 8364 || t === 8451 || t === 8453 || t === 8457 || t === 8467 || t === 8470 || t === 8481 || t === 8482 || t === 8486 || t === 8491 || t === 8531 || t === 8532 || t >= 8539 && t <= 8542 || t >= 8544 && t <= 8555 || t >= 8560 && t <= 8569 || t === 8585 || t >= 8592 && t <= 8601 || t === 8632 || t === 8633 || t === 8658 || t === 8660 || t === 8679 || t === 8704 || t === 8706 || t === 8707 || t === 8711 || t === 8712 || t === 8715 || t === 8719 || t === 8721 || t === 8725 || t === 8730 || t >= 8733 && t <= 8736 || t === 8739 || t === 8741 || t >= 8743 && t <= 8748 || t === 8750 || t >= 8756 && t <= 8759 || t === 8764 || t === 8765 || t === 8776 || t === 8780 || t === 8786 || t === 8800 || t === 8801 || t >= 8804 && t <= 8807 || t === 8810 || t === 8811 || t === 8814 || t === 8815 || t === 8834 || t === 8835 || t === 8838 || t === 8839 || t === 8853 || t === 8857 || t === 8869 || t === 8895 || t === 8978 || t >= 9312 && t <= 9449 || t >= 9451 && t <= 9547 || t >= 9552 && t <= 9587 || t >= 9600 && t <= 9615 || t >= 9618 && t <= 9621 || t === 9632 || t === 9633 || t >= 9635 && t <= 9641 || t === 9650 || t === 9651 || t === 9654 || t === 9655 || t === 9660 || t === 9661 || t === 9664 || t === 9665 || t >= 9670 && t <= 9672 || t === 9675 || t >= 9678 && t <= 9681 || t >= 9698 && t <= 9701 || t === 9711 || t === 9733 || t === 9734 || t === 9737 || t === 9742 || t === 9743 || t === 9756 || t === 9758 || t === 9792 || t === 9794 || t === 9824 || t === 9825 || t >= 9827 && t <= 9829 || t >= 9831 && t <= 9834 || t === 9836 || t === 9837 || t === 9839 || t === 9886 || t === 9887 || t === 9919 || t >= 9926 && t <= 9933 || t >= 9935 && t <= 9939 || t >= 9941 && t <= 9953 || t === 9955 || t === 9960 || t === 9961 || t >= 9963 && t <= 9969 || t === 9972 || t >= 9974 && t <= 9977 || t === 9979 || t === 9980 || t === 9982 || t === 9983 || t === 10045 || t >= 10102 && t <= 10111 || t >= 11094 && t <= 11097 || t >= 12872 && t <= 12879 || t >= 57344 && t <= 63743 || t >= 65024 && t <= 65039 || t === 65533 || t >= 127232 && t <= 127242 || t >= 127248 && t <= 127277 || t >= 127280 && t <= 127337 || t >= 127344 && t <= 127373 || t === 127375 || t === 127376 || t >= 127387 && t <= 127404 || t >= 917760 && t <= 917999 || t >= 983040 && t <= 1048573 || t >= 1048576 && t <= 1114109, lt = (t) => t === 12288 || t >= 65281 && t <= 65376 || t >= 65504 && t <= 65510, ht = (t) => t >= 4352 && t <= 4447 || t === 8986 || t === 8987 || t === 9001 || t === 9002 || t >= 9193 && t <= 9196 || t === 9200 || t === 9203 || t === 9725 || t === 9726 || t === 9748 || t === 9749 || t >= 9800 && t <= 9811 || t === 9855 || t === 9875 || t === 9889 || t === 9898 || t === 9899 || t === 9917 || t === 9918 || t === 9924 || t === 9925 || t === 9934 || t === 9940 || t === 9962 || t === 9970 || t === 9971 || t === 9973 || t === 9978 || t === 9981 || t === 9989 || t === 9994 || t === 9995 || t === 10024 || t === 10060 || t === 10062 || t >= 10067 && t <= 10069 || t === 10071 || t >= 10133 && t <= 10135 || t === 10160 || t === 10175 || t === 11035 || t === 11036 || t === 11088 || t === 11093 || t >= 11904 && t <= 11929 || t >= 11931 && t <= 12019 || t >= 12032 && t <= 12245 || t >= 12272 && t <= 12287 || t >= 12289 && t <= 12350 || t >= 12353 && t <= 12438 || t >= 12441 && t <= 12543 || t >= 12549 && t <= 12591 || t >= 12593 && t <= 12686 || t >= 12688 && t <= 12771 || t >= 12783 && t <= 12830 || t >= 12832 && t <= 12871 || t >= 12880 && t <= 19903 || t >= 19968 && t <= 42124 || t >= 42128 && t <= 42182 || t >= 43360 && t <= 43388 || t >= 44032 && t <= 55203 || t >= 63744 && t <= 64255 || t >= 65040 && t <= 65049 || t >= 65072 && t <= 65106 || t >= 65108 && t <= 65126 || t >= 65128 && t <= 65131 || t >= 94176 && t <= 94180 || t === 94192 || t === 94193 || t >= 94208 && t <= 100343 || t >= 100352 && t <= 101589 || t >= 101632 && t <= 101640 || t >= 110576 && t <= 110579 || t >= 110581 && t <= 110587 || t === 110589 || t === 110590 || t >= 110592 && t <= 110882 || t === 110898 || t >= 110928 && t <= 110930 || t === 110933 || t >= 110948 && t <= 110951 || t >= 110960 && t <= 111355 || t === 126980 || t === 127183 || t === 127374 || t >= 127377 && t <= 127386 || t >= 127488 && t <= 127490 || t >= 127504 && t <= 127547 || t >= 127552 && t <= 127560 || t === 127568 || t === 127569 || t >= 127584 && t <= 127589 || t >= 127744 && t <= 127776 || t >= 127789 && t <= 127797 || t >= 127799 && t <= 127868 || t >= 127870 && t <= 127891 || t >= 127904 && t <= 127946 || t >= 127951 && t <= 127955 || t >= 127968 && t <= 127984 || t === 127988 || t >= 127992 && t <= 128062 || t === 128064 || t >= 128066 && t <= 128252 || t >= 128255 && t <= 128317 || t >= 128331 && t <= 128334 || t >= 128336 && t <= 128359 || t === 128378 || t === 128405 || t === 128406 || t === 128420 || t >= 128507 && t <= 128591 || t >= 128640 && t <= 128709 || t === 128716 || t >= 128720 && t <= 128722 || t >= 128725 && t <= 128727 || t >= 128732 && t <= 128735 || t === 128747 || t === 128748 || t >= 128756 && t <= 128764 || t >= 128992 && t <= 129003 || t === 129008 || t >= 129292 && t <= 129338 || t >= 129340 && t <= 129349 || t >= 129351 && t <= 129535 || t >= 129648 && t <= 129660 || t >= 129664 && t <= 129672 || t >= 129680 && t <= 129725 || t >= 129727 && t <= 129733 || t >= 129742 && t <= 129755 || t >= 129760 && t <= 129768 || t >= 129776 && t <= 129784 || t >= 131072 && t <= 196605 || t >= 196608 && t <= 262141, O = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/y, y = /[\x00-\x08\x0A-\x1F\x7F-\x9F]{1,1000}/y, L = /\t{1,1000}/y, P = /[\u{1F1E6}-\u{1F1FF}]{2}|\u{1F3F4}[\u{E0061}-\u{E007A}]{2}[\u{E0030}-\u{E0039}\u{E0061}-\u{E007A}]{1,3}\u{E007F}|(?:\p{Emoji}\uFE0F\u20E3?|\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation})(?:\u200D(?:\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation}|\p{Emoji}\uFE0F\u20E3?))*/uy, M = /(?:[\x20-\x7E\xA0-\xFF](?!\uFE0F)){1,1000}/y, ct = /\p{M}+/gu, ft = {
	limit: Infinity,
	ellipsis: ""
}, X = (t, e = {}, s = {}) => {
	const i = e.limit ?? Infinity, r = e.ellipsis ?? "", n = e?.ellipsisWidth ?? (r ? X(r, ft, s).width : 0), u = s.ansiWidth ?? 0, a = s.controlWidth ?? 0, l = s.tabWidth ?? 8, E = s.ambiguousWidth ?? 1, g = s.emojiWidth ?? 2, m = s.fullWidthWidth ?? 2, A = s.regularWidth ?? 1, V = s.wideWidth ?? 2;
	let h = 0, o = 0, p = t.length, v = 0, F = !1, d = p, b = Math.max(0, i - n), C = 0, w = 0, c = 0, f = 0;
	t: for (;;) {
		if (w > C || o >= p && o > h) {
			const ut = t.slice(C, w) || t.slice(h, o);
			v = 0;
			for (const Y of ut.replaceAll(ct, "")) {
				const $ = Y.codePointAt(0) || 0;
				if (lt($) ? f = m : ht($) ? f = V : E !== A && at($) ? f = E : f = A, c + f > b && (d = Math.min(d, Math.max(C, h) + v)), c + f > i) {
					F = !0;
					break t;
				}
				v += Y.length, c += f;
			}
			C = w = 0;
		}
		if (o >= p) break;
		if (M.lastIndex = o, M.test(t)) {
			if (v = M.lastIndex - o, f = v * A, c + f > b && (d = Math.min(d, o + Math.floor((b - c) / A))), c + f > i) {
				F = !0;
				break;
			}
			c += f, C = h, w = o, o = h = M.lastIndex;
			continue;
		}
		if (O.lastIndex = o, O.test(t)) {
			if (c + u > b && (d = Math.min(d, o)), c + u > i) {
				F = !0;
				break;
			}
			c += u, C = h, w = o, o = h = O.lastIndex;
			continue;
		}
		if (y.lastIndex = o, y.test(t)) {
			if (v = y.lastIndex - o, f = v * a, c + f > b && (d = Math.min(d, o + Math.floor((b - c) / a))), c + f > i) {
				F = !0;
				break;
			}
			c += f, C = h, w = o, o = h = y.lastIndex;
			continue;
		}
		if (L.lastIndex = o, L.test(t)) {
			if (v = L.lastIndex - o, f = v * l, c + f > b && (d = Math.min(d, o + Math.floor((b - c) / l))), c + f > i) {
				F = !0;
				break;
			}
			c += f, C = h, w = o, o = h = L.lastIndex;
			continue;
		}
		if (P.lastIndex = o, P.test(t)) {
			if (c + g > b && (d = Math.min(d, o)), c + g > i) {
				F = !0;
				break;
			}
			c += g, C = h, w = o, o = h = P.lastIndex;
			continue;
		}
		o += 1;
	}
	return {
		width: F ? b : c,
		index: F ? d : p,
		truncated: F,
		ellipsed: F && i >= n
	};
}, pt = {
	limit: Infinity,
	ellipsis: "",
	ellipsisWidth: 0
}, S = (t, e = {}) => X(t, pt, e).width, T = "\x1B", Z = "", Ft = 39, j = "\x07", Q = "[", dt = "]", tt = "m", U = `${dt}8;;`, et = new RegExp(`(?:\\${Q}(?<code>\\d+)m|\\${U}(?<uri>.*)${j})`, "y"), mt = (t) => {
	if (t >= 30 && t <= 37 || t >= 90 && t <= 97) return 39;
	if (t >= 40 && t <= 47 || t >= 100 && t <= 107) return 49;
	if (t === 1 || t === 2) return 22;
	if (t === 3) return 23;
	if (t === 4) return 24;
	if (t === 7) return 27;
	if (t === 8) return 28;
	if (t === 9) return 29;
	if (t === 0) return 0;
}, st = (t) => `${T}${Q}${t}${tt}`, it = (t) => `${T}${U}${t}${j}`, gt = (t) => t.map((e) => S(e)), G = (t, e, s) => {
	const i = e[Symbol.iterator]();
	let r = !1, n = !1, u = t.at(-1), a = u === void 0 ? 0 : S(u), l = i.next(), E = i.next(), g = 0;
	for (; !l.done;) {
		const m = l.value, A = S(m);
		a + A <= s ? t[t.length - 1] += m : (t.push(m), a = 0), (m === T || m === Z) && (r = !0, n = e.startsWith(U, g + 1)), r ? n ? m === j && (r = !1, n = !1) : m === tt && (r = !1) : (a += A, a === s && !E.done && (t.push(""), a = 0)), l = E, E = i.next(), g += m.length;
	}
	u = t.at(-1), !a && u !== void 0 && u.length > 0 && t.length > 1 && (t[t.length - 2] += t.pop());
}, vt = (t) => {
	const e = t.split(" ");
	let s = e.length;
	for (; s > 0 && !(S(e[s - 1]) > 0);) s--;
	return s === e.length ? t : e.slice(0, s).join(" ") + e.slice(s).join("");
}, Et = (t, e, s = {}) => {
	if (s.trim !== !1 && t.trim() === "") return "";
	let i = "", r, n;
	const u = t.split(" "), a = gt(u);
	let l = [""];
	for (const [h, o] of u.entries()) {
		s.trim !== !1 && (l[l.length - 1] = (l.at(-1) ?? "").trimStart());
		let p = S(l.at(-1) ?? "");
		if (h !== 0 && (p >= e && (s.wordWrap === !1 || s.trim === !1) && (l.push(""), p = 0), (p > 0 || s.trim === !1) && (l[l.length - 1] += " ", p++)), s.hard && a[h] > e) {
			const v = e - p, F = 1 + Math.floor((a[h] - v - 1) / e);
			Math.floor((a[h] - 1) / e) < F && l.push(""), G(l, o, e);
			continue;
		}
		if (p + a[h] > e && p > 0 && a[h] > 0) {
			if (s.wordWrap === !1 && p < e) {
				G(l, o, e);
				continue;
			}
			l.push("");
		}
		if (p + a[h] > e && s.wordWrap === !1) {
			G(l, o, e);
			continue;
		}
		l[l.length - 1] += o;
	}
	s.trim !== !1 && (l = l.map((h) => vt(h)));
	const E = l.join(`
`), g = E[Symbol.iterator]();
	let m = g.next(), A = g.next(), V = 0;
	for (; !m.done;) {
		const h = m.value, o = A.value;
		if (i += h, h === T || h === Z) {
			et.lastIndex = V + 1;
			const F = et.exec(E)?.groups;
			if (F?.code !== void 0) {
				const d = Number.parseFloat(F.code);
				r = d === Ft ? void 0 : d;
			} else F?.uri !== void 0 && (n = F.uri.length === 0 ? void 0 : F.uri);
		}
		const p = r ? mt(r) : void 0;
		o === `
` ? (n && (i += it("")), r && p && (i += st(p))) : h === `
` && (r && p && (i += st(r)), n && (i += it(n))), V += h.length, m = A, A = g.next();
	}
	return i;
};
function K(t, e, s) {
	return String(t).normalize().replaceAll(`\r
`, `
`).split(`
`).map((i) => Et(i, e, s)).join(`
`);
}
const _ = {
	actions: new Set([
		"up",
		"down",
		"left",
		"right",
		"space",
		"enter",
		"cancel"
	]),
	aliases: new Map([
		["k", "up"],
		["j", "down"],
		["h", "left"],
		["l", "right"],
		["", "cancel"],
		["escape", "cancel"]
	]),
	messages: {
		cancel: "Canceled",
		error: "Something went wrong"
	},
	withGuide: !0
};
function H(t, e) {
	if (typeof t == "string") return _.aliases.get(t) === e;
	for (const s of t) if (s !== void 0 && H(s, e)) return !0;
	return !1;
}
function _t(t, e) {
	if (t === e) return;
	const s = t.split(`
`), i = e.split(`
`), r = Math.max(s.length, i.length), n = [];
	for (let u = 0; u < r; u++) s[u] !== i[u] && n.push(u);
	return {
		lines: n,
		numLinesBefore: s.length,
		numLinesAfter: i.length,
		numLines: r
	};
}
const bt = globalThis.process.platform.startsWith("win"), z = Symbol("clack:cancel");
function Ct(t) {
	return t === z;
}
function W(t, e) {
	const s = t;
	s.isTTY && s.setRawMode(e);
}
function xt({ input: t = stdin, output: e = stdout, overwrite: s = !0, hideCursor: i = !0 } = {}) {
	const r = k.createInterface({
		input: t,
		output: e,
		prompt: "",
		tabSize: 1
	});
	k.emitKeypressEvents(t, r), t instanceof ReadStream && t.isTTY && t.setRawMode(!0);
	const n = (u, { name: a, sequence: l }) => {
		if (H([
			String(u),
			a,
			l
		], "cancel")) {
			i && e.write(import_src.cursor.show), process.exit(0);
			return;
		}
		if (!s) return;
		const g = a === "return" ? 0 : -1, m = a === "return" ? -1 : 0;
		k.moveCursor(e, g, m, () => {
			k.clearLine(e, 1, () => {
				t.once("keypress", n);
			});
		});
	};
	return i && e.write(import_src.cursor.hide), t.once("keypress", n), () => {
		t.off("keypress", n), i && e.write(import_src.cursor.show), t instanceof ReadStream && t.isTTY && !bt && t.setRawMode(!1), r.terminal = !1, r.close();
	};
}
const rt = (t) => "columns" in t && typeof t.columns == "number" ? t.columns : 80, nt = (t) => "rows" in t && typeof t.rows == "number" ? t.rows : 20;
function Bt(t, e, s, i = s) {
	return K(e, rt(t ?? stdout) - s.length, {
		hard: !0,
		trim: !1
	}).split(`
`).map((n, u) => `${u === 0 ? i : s}${n}`).join(`
`);
}
var B = class {
	input;
	output;
	_abortSignal;
	rl;
	opts;
	_render;
	_track = !1;
	_prevFrame = "";
	_subscribers = /* @__PURE__ */ new Map();
	_cursor = 0;
	state = "initial";
	error = "";
	value;
	userInput = "";
	constructor(e, s = !0) {
		const { input: i = stdin, output: r = stdout, render: n, signal: u, ...a } = e;
		this.opts = a, this.onKeypress = this.onKeypress.bind(this), this.close = this.close.bind(this), this.render = this.render.bind(this), this._render = n.bind(this), this._track = s, this._abortSignal = u, this.input = i, this.output = r;
	}
	unsubscribe() {
		this._subscribers.clear();
	}
	setSubscriber(e, s) {
		const i = this._subscribers.get(e) ?? [];
		i.push(s), this._subscribers.set(e, i);
	}
	on(e, s) {
		this.setSubscriber(e, { cb: s });
	}
	once(e, s) {
		this.setSubscriber(e, {
			cb: s,
			once: !0
		});
	}
	emit(e, ...s) {
		const i = this._subscribers.get(e) ?? [], r = [];
		for (const n of i) n.cb(...s), n.once && r.push(() => i.splice(i.indexOf(n), 1));
		for (const n of r) n();
	}
	prompt() {
		return new Promise((e) => {
			if (this._abortSignal) {
				if (this._abortSignal.aborted) return this.state = "cancel", this.close(), e(z);
				this._abortSignal.addEventListener("abort", () => {
					this.state = "cancel", this.close();
				}, { once: !0 });
			}
			this.rl = ot.createInterface({
				input: this.input,
				tabSize: 2,
				prompt: "",
				escapeCodeTimeout: 50,
				terminal: !0
			}), this.rl.prompt(), this.opts.initialUserInput !== void 0 && this._setUserInput(this.opts.initialUserInput, !0), this.input.on("keypress", this.onKeypress), W(this.input, !0), this.output.on("resize", this.render), this.render(), this.once("submit", () => {
				this.output.write(import_src.cursor.show), this.output.off("resize", this.render), W(this.input, !1), e(this.value);
			}), this.once("cancel", () => {
				this.output.write(import_src.cursor.show), this.output.off("resize", this.render), W(this.input, !1), e(z);
			});
		});
	}
	_isActionKey(e, s) {
		return e === "	";
	}
	_setValue(e) {
		this.value = e, this.emit("value", this.value);
	}
	_setUserInput(e, s) {
		this.userInput = e ?? "", this.emit("userInput", this.userInput), s && this._track && this.rl && (this.rl.write(this.userInput), this._cursor = this.rl.cursor);
	}
	_clearUserInput() {
		this.rl?.write(null, {
			ctrl: !0,
			name: "u"
		}), this._setUserInput("");
	}
	onKeypress(e, s) {
		if (this._track && s.name !== "return" && (s.name && this._isActionKey(e, s) && this.rl?.write(null, {
			ctrl: !0,
			name: "h"
		}), this._cursor = this.rl?.cursor ?? 0, this._setUserInput(this.rl?.line)), this.state === "error" && (this.state = "active"), s?.name && (!this._track && _.aliases.has(s.name) && this.emit("cursor", _.aliases.get(s.name)), _.actions.has(s.name) && this.emit("cursor", s.name)), e && (e.toLowerCase() === "y" || e.toLowerCase() === "n") && this.emit("confirm", e.toLowerCase() === "y"), this.emit("key", e?.toLowerCase(), s), s?.name === "return") {
			if (this.opts.validate) {
				const i = this.opts.validate(this.value);
				i && (this.error = i instanceof Error ? i.message : i, this.state = "error", this.rl?.write(this.userInput));
			}
			this.state !== "error" && (this.state = "submit");
		}
		H([
			e,
			s?.name,
			s?.sequence
		], "cancel") && (this.state = "cancel"), (this.state === "submit" || this.state === "cancel") && this.emit("finalize"), this.render(), (this.state === "submit" || this.state === "cancel") && this.close();
	}
	close() {
		this.input.unpipe(), this.input.removeListener("keypress", this.onKeypress), this.output.write(`
`), W(this.input, !1), this.rl?.close(), this.rl = void 0, this.emit(`${this.state}`, this.value), this.unsubscribe();
	}
	restoreCursor() {
		const e = K(this._prevFrame, process.stdout.columns, {
			hard: !0,
			trim: !1
		}).split(`
`).length - 1;
		this.output.write(import_src.cursor.move(-999, e * -1));
	}
	render() {
		const e = K(this._render(this) ?? "", process.stdout.columns, {
			hard: !0,
			trim: !1
		});
		if (e !== this._prevFrame) {
			if (this.state === "initial") this.output.write(import_src.cursor.hide);
			else {
				const s = _t(this._prevFrame, e), i = nt(this.output);
				if (this.restoreCursor(), s) {
					const r = Math.max(0, s.numLinesAfter - i), n = Math.max(0, s.numLinesBefore - i);
					let u = s.lines.find((a) => a >= r);
					if (u === void 0) {
						this._prevFrame = e;
						return;
					}
					if (s.lines.length === 1) {
						this.output.write(import_src.cursor.move(0, u - n)), this.output.write(import_src.erase.lines(1));
						const a = e.split(`
`);
						this.output.write(a[u]), this._prevFrame = e, this.output.write(import_src.cursor.move(0, a.length - u - 1));
						return;
					} else if (s.lines.length > 1) {
						if (r < n) u = r;
						else {
							const l = u - n;
							l > 0 && this.output.write(import_src.cursor.move(0, l));
						}
						this.output.write(import_src.erase.down());
						const a = e.split(`
`).slice(u);
						this.output.write(a.join(`
`)), this._prevFrame = e;
						return;
					}
				}
				this.output.write(import_src.erase.down());
			}
			this.output.write(e), this.state === "initial" && (this.state = "active"), this._prevFrame = e;
		}
	}
};
var kt = class extends B {
	get cursor() {
		return this.value ? 0 : 1;
	}
	get _value() {
		return this.cursor === 0;
	}
	constructor(e) {
		super(e, !1), this.value = !!e.initialValue, this.on("userInput", () => {
			this.value = this._value;
		}), this.on("confirm", (s) => {
			this.output.write(import_src.cursor.move(0, -1)), this.value = s, this.state = "submit", this.close();
		}), this.on("cursor", () => {
			this.value = !this.value;
		});
	}
};
let Lt = class extends B {
	options;
	cursor = 0;
	get _value() {
		return this.options[this.cursor].value;
	}
	get _enabledOptions() {
		return this.options.filter((e) => e.disabled !== !0);
	}
	toggleAll() {
		const e = this._enabledOptions;
		this.value = this.value !== void 0 && this.value.length === e.length ? [] : e.map((i) => i.value);
	}
	toggleInvert() {
		const e = this.value;
		if (!e) return;
		this.value = this._enabledOptions.filter((i) => !e.includes(i.value)).map((i) => i.value);
	}
	toggleValue() {
		this.value === void 0 && (this.value = []);
		this.value = this.value.includes(this._value) ? this.value.filter((s) => s !== this._value) : [...this.value, this._value];
	}
	constructor(e) {
		super(e, !1), this.options = e.options, this.value = [...e.initialValues ?? []];
		const s = Math.max(this.options.findIndex(({ value: i }) => i === e.cursorAt), 0);
		this.cursor = this.options[s].disabled ? x(s, 1, this.options) : s, this.on("key", (i) => {
			i === "a" && this.toggleAll(), i === "i" && this.toggleInvert();
		}), this.on("cursor", (i) => {
			switch (i) {
				case "left":
				case "up":
					this.cursor = x(this.cursor, -1, this.options);
					break;
				case "down":
				case "right":
					this.cursor = x(this.cursor, 1, this.options);
					break;
				case "space":
					this.toggleValue();
					break;
			}
		});
	}
};
var Tt = class extends B {
	options;
	cursor = 0;
	get _selectedValue() {
		return this.options[this.cursor];
	}
	changeValue() {
		this.value = this._selectedValue.value;
	}
	constructor(e) {
		super(e, !1), this.options = e.options;
		const s = this.options.findIndex(({ value: r }) => r === e.initialValue), i = s === -1 ? 0 : s;
		this.cursor = this.options[i].disabled ? x(i, 1, this.options) : i, this.changeValue(), this.on("cursor", (r) => {
			switch (r) {
				case "left":
				case "up":
					this.cursor = x(this.cursor, -1, this.options);
					break;
				case "down":
				case "right":
					this.cursor = x(this.cursor, 1, this.options);
					break;
			}
			this.changeValue();
		});
	}
};
var $t = class extends B {
	get userInputWithCursor() {
		if (this.state === "submit") return this.userInput;
		const e = this.userInput;
		if (this.cursor >= e.length) return `${this.userInput}\u2588`;
		const s = e.slice(0, this.cursor), [i, ...r] = e.slice(this.cursor);
		return `${s}${styleText("inverse", i)}${r.join("")}`;
	}
	get cursor() {
		return this._cursor;
	}
	constructor(e) {
		super({
			...e,
			initialUserInput: e.initialUserInput ?? e.initialValue
		}), this.on("userInput", (s) => {
			this._setValue(s);
		}), this.on("finalize", () => {
			this.value || (this.value = e.defaultValue), this.value === void 0 && (this.value = "");
		});
	}
};
//#endregion
//#region ../prompts/dist/index.mjs
function isUnicodeSupported() {
	if (process$1.platform !== "win32") return process$1.env.TERM !== "linux";
	return Boolean(process$1.env.CI) || Boolean(process$1.env.WT_SESSION) || Boolean(process$1.env.TERMINUS_SUBLIME) || process$1.env.ConEmuTask === "{cmd::Cmder}" || process$1.env.TERM_PROGRAM === "Terminus-Sublime" || process$1.env.TERM_PROGRAM === "vscode" || process$1.env.TERM === "xterm-256color" || process$1.env.TERM === "alacritty" || process$1.env.TERMINAL_EMULATOR === "JetBrains-JediTerm";
}
const unicode = isUnicodeSupported();
const isCI = () => process.env.CI === "true";
const unicodeOr = (c, fallback) => unicode ? c : fallback;
const S_POINTER_ACTIVE = unicodeOr("›", ">");
const S_STEP_ACTIVE = S_POINTER_ACTIVE;
const S_STEP_CANCEL = unicodeOr("■", "x");
const S_STEP_ERROR = unicodeOr("▲", "x");
const S_STEP_SUBMIT = unicodeOr("◇", "o");
unicodeOr("┌", "T");
const S_BAR = unicodeOr("│", "|");
const S_BAR_END = unicodeOr("└", "—");
unicodeOr("┐", "T");
unicodeOr("┘", "—");
const S_CHECKBOX_ACTIVE = unicodeOr("◻", "[•]");
const S_CHECKBOX_SELECTED = unicodeOr("◼", "[+]");
const S_CHECKBOX_INACTIVE = unicodeOr("◻", "[ ]");
unicodeOr("▪", "•");
const S_BAR_H = unicodeOr("─", "-");
const S_CORNER_TOP_RIGHT = unicodeOr("╮", "+");
const S_CONNECT_LEFT = unicodeOr("├", "+");
const S_CORNER_BOTTOM_RIGHT = unicodeOr("╯", "+");
const S_CORNER_BOTTOM_LEFT = unicodeOr("╰", "+");
unicodeOr("╭", "+");
const S_INFO = unicodeOr("●", "•");
const S_SUCCESS = unicodeOr("◆", "*");
const S_WARN = unicodeOr("▲", "!");
const S_ERROR = unicodeOr("■", "x");
const completeColor = (value) => color.gray(value);
const symbol = (state) => {
	switch (state) {
		case "initial":
		case "active": return color.blue(S_STEP_ACTIVE);
		case "cancel": return color.red(S_STEP_CANCEL);
		case "error": return color.yellow(S_STEP_ERROR);
		case "submit": return completeColor(S_STEP_SUBMIT);
	}
};
const symbolBar = (state) => {
	switch (state) {
		case "initial":
		case "active": return color.blue(S_BAR);
		case "cancel": return color.red(S_BAR);
		case "error": return color.yellow(S_BAR);
		case "submit": return completeColor(S_BAR);
	}
};
const isAmbiguous = (x) => {
	return x === 161 || x === 164 || x === 167 || x === 168 || x === 170 || x === 173 || x === 174 || x >= 176 && x <= 180 || x >= 182 && x <= 186 || x >= 188 && x <= 191 || x === 198 || x === 208 || x === 215 || x === 216 || x >= 222 && x <= 225 || x === 230 || x >= 232 && x <= 234 || x === 236 || x === 237 || x === 240 || x === 242 || x === 243 || x >= 247 && x <= 250 || x === 252 || x === 254 || x === 257 || x === 273 || x === 275 || x === 283 || x === 294 || x === 295 || x === 299 || x >= 305 && x <= 307 || x === 312 || x >= 319 && x <= 322 || x === 324 || x >= 328 && x <= 331 || x === 333 || x === 338 || x === 339 || x === 358 || x === 359 || x === 363 || x === 462 || x === 464 || x === 466 || x === 468 || x === 470 || x === 472 || x === 474 || x === 476 || x === 593 || x === 609 || x === 708 || x === 711 || x >= 713 && x <= 715 || x === 717 || x === 720 || x >= 728 && x <= 731 || x === 733 || x === 735 || x >= 768 && x <= 879 || x >= 913 && x <= 929 || x >= 931 && x <= 937 || x >= 945 && x <= 961 || x >= 963 && x <= 969 || x === 1025 || x >= 1040 && x <= 1103 || x === 1105 || x === 8208 || x >= 8211 && x <= 8214 || x === 8216 || x === 8217 || x === 8220 || x === 8221 || x >= 8224 && x <= 8226 || x >= 8228 && x <= 8231 || x === 8240 || x === 8242 || x === 8243 || x === 8245 || x === 8251 || x === 8254 || x === 8308 || x === 8319 || x >= 8321 && x <= 8324 || x === 8364 || x === 8451 || x === 8453 || x === 8457 || x === 8467 || x === 8470 || x === 8481 || x === 8482 || x === 8486 || x === 8491 || x === 8531 || x === 8532 || x >= 8539 && x <= 8542 || x >= 8544 && x <= 8555 || x >= 8560 && x <= 8569 || x === 8585 || x >= 8592 && x <= 8601 || x === 8632 || x === 8633 || x === 8658 || x === 8660 || x === 8679 || x === 8704 || x === 8706 || x === 8707 || x === 8711 || x === 8712 || x === 8715 || x === 8719 || x === 8721 || x === 8725 || x === 8730 || x >= 8733 && x <= 8736 || x === 8739 || x === 8741 || x >= 8743 && x <= 8748 || x === 8750 || x >= 8756 && x <= 8759 || x === 8764 || x === 8765 || x === 8776 || x === 8780 || x === 8786 || x === 8800 || x === 8801 || x >= 8804 && x <= 8807 || x === 8810 || x === 8811 || x === 8814 || x === 8815 || x === 8834 || x === 8835 || x === 8838 || x === 8839 || x === 8853 || x === 8857 || x === 8869 || x === 8895 || x === 8978 || x >= 9312 && x <= 9449 || x >= 9451 && x <= 9547 || x >= 9552 && x <= 9587 || x >= 9600 && x <= 9615 || x >= 9618 && x <= 9621 || x === 9632 || x === 9633 || x >= 9635 && x <= 9641 || x === 9650 || x === 9651 || x === 9654 || x === 9655 || x === 9660 || x === 9661 || x === 9664 || x === 9665 || x >= 9670 && x <= 9672 || x === 9675 || x >= 9678 && x <= 9681 || x >= 9698 && x <= 9701 || x === 9711 || x === 9733 || x === 9734 || x === 9737 || x === 9742 || x === 9743 || x === 9756 || x === 9758 || x === 9792 || x === 9794 || x === 9824 || x === 9825 || x >= 9827 && x <= 9829 || x >= 9831 && x <= 9834 || x === 9836 || x === 9837 || x === 9839 || x === 9886 || x === 9887 || x === 9919 || x >= 9926 && x <= 9933 || x >= 9935 && x <= 9939 || x >= 9941 && x <= 9953 || x === 9955 || x === 9960 || x === 9961 || x >= 9963 && x <= 9969 || x === 9972 || x >= 9974 && x <= 9977 || x === 9979 || x === 9980 || x === 9982 || x === 9983 || x === 10045 || x >= 10102 && x <= 10111 || x >= 11094 && x <= 11097 || x >= 12872 && x <= 12879 || x >= 57344 && x <= 63743 || x >= 65024 && x <= 65039 || x === 65533 || x >= 127232 && x <= 127242 || x >= 127248 && x <= 127277 || x >= 127280 && x <= 127337 || x >= 127344 && x <= 127373 || x === 127375 || x === 127376 || x >= 127387 && x <= 127404 || x >= 917760 && x <= 917999 || x >= 983040 && x <= 1048573 || x >= 1048576 && x <= 1114109;
};
const isFullWidth = (x) => {
	return x === 12288 || x >= 65281 && x <= 65376 || x >= 65504 && x <= 65510;
};
const isWide = (x) => {
	return x >= 4352 && x <= 4447 || x === 8986 || x === 8987 || x === 9001 || x === 9002 || x >= 9193 && x <= 9196 || x === 9200 || x === 9203 || x === 9725 || x === 9726 || x === 9748 || x === 9749 || x >= 9800 && x <= 9811 || x === 9855 || x === 9875 || x === 9889 || x === 9898 || x === 9899 || x === 9917 || x === 9918 || x === 9924 || x === 9925 || x === 9934 || x === 9940 || x === 9962 || x === 9970 || x === 9971 || x === 9973 || x === 9978 || x === 9981 || x === 9989 || x === 9994 || x === 9995 || x === 10024 || x === 10060 || x === 10062 || x >= 10067 && x <= 10069 || x === 10071 || x >= 10133 && x <= 10135 || x === 10160 || x === 10175 || x === 11035 || x === 11036 || x === 11088 || x === 11093 || x >= 11904 && x <= 11929 || x >= 11931 && x <= 12019 || x >= 12032 && x <= 12245 || x >= 12272 && x <= 12287 || x >= 12289 && x <= 12350 || x >= 12353 && x <= 12438 || x >= 12441 && x <= 12543 || x >= 12549 && x <= 12591 || x >= 12593 && x <= 12686 || x >= 12688 && x <= 12771 || x >= 12783 && x <= 12830 || x >= 12832 && x <= 12871 || x >= 12880 && x <= 19903 || x >= 19968 && x <= 42124 || x >= 42128 && x <= 42182 || x >= 43360 && x <= 43388 || x >= 44032 && x <= 55203 || x >= 63744 && x <= 64255 || x >= 65040 && x <= 65049 || x >= 65072 && x <= 65106 || x >= 65108 && x <= 65126 || x >= 65128 && x <= 65131 || x >= 94176 && x <= 94180 || x === 94192 || x === 94193 || x >= 94208 && x <= 100343 || x >= 100352 && x <= 101589 || x >= 101632 && x <= 101640 || x >= 110576 && x <= 110579 || x >= 110581 && x <= 110587 || x === 110589 || x === 110590 || x >= 110592 && x <= 110882 || x === 110898 || x >= 110928 && x <= 110930 || x === 110933 || x >= 110948 && x <= 110951 || x >= 110960 && x <= 111355 || x === 126980 || x === 127183 || x === 127374 || x >= 127377 && x <= 127386 || x >= 127488 && x <= 127490 || x >= 127504 && x <= 127547 || x >= 127552 && x <= 127560 || x === 127568 || x === 127569 || x >= 127584 && x <= 127589 || x >= 127744 && x <= 127776 || x >= 127789 && x <= 127797 || x >= 127799 && x <= 127868 || x >= 127870 && x <= 127891 || x >= 127904 && x <= 127946 || x >= 127951 && x <= 127955 || x >= 127968 && x <= 127984 || x === 127988 || x >= 127992 && x <= 128062 || x === 128064 || x >= 128066 && x <= 128252 || x >= 128255 && x <= 128317 || x >= 128331 && x <= 128334 || x >= 128336 && x <= 128359 || x === 128378 || x === 128405 || x === 128406 || x === 128420 || x >= 128507 && x <= 128591 || x >= 128640 && x <= 128709 || x === 128716 || x >= 128720 && x <= 128722 || x >= 128725 && x <= 128727 || x >= 128732 && x <= 128735 || x === 128747 || x === 128748 || x >= 128756 && x <= 128764 || x >= 128992 && x <= 129003 || x === 129008 || x >= 129292 && x <= 129338 || x >= 129340 && x <= 129349 || x >= 129351 && x <= 129535 || x >= 129648 && x <= 129660 || x >= 129664 && x <= 129672 || x >= 129680 && x <= 129725 || x >= 129727 && x <= 129733 || x >= 129742 && x <= 129755 || x >= 129760 && x <= 129768 || x >= 129776 && x <= 129784 || x >= 131072 && x <= 196605 || x >= 196608 && x <= 262141;
};
const ANSI_RE = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/y;
const CONTROL_RE = /[\x00-\x08\x0A-\x1F\x7F-\x9F]{1,1000}/y;
const TAB_RE = /\t{1,1000}/y;
const EMOJI_RE = /[\u{1F1E6}-\u{1F1FF}]{2}|\u{1F3F4}[\u{E0061}-\u{E007A}]{2}[\u{E0030}-\u{E0039}\u{E0061}-\u{E007A}]{1,3}\u{E007F}|(?:\p{Emoji}\uFE0F\u20E3?|\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation})(?:\u200D(?:\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation}|\p{Emoji}\uFE0F\u20E3?))*/uy;
const LATIN_RE = /(?:[\x20-\x7E\xA0-\xFF](?!\uFE0F)){1,1000}/y;
const MODIFIER_RE = /\p{M}+/gu;
const NO_TRUNCATION$1 = {
	limit: Infinity,
	ellipsis: ""
};
const getStringTruncatedWidth = (input, truncationOptions = {}, widthOptions = {}) => {
	const LIMIT = truncationOptions.limit ?? Infinity;
	const ELLIPSIS = truncationOptions.ellipsis ?? "";
	const ELLIPSIS_WIDTH = truncationOptions?.ellipsisWidth ?? (ELLIPSIS ? getStringTruncatedWidth(ELLIPSIS, NO_TRUNCATION$1, widthOptions).width : 0);
	const ANSI_WIDTH = widthOptions.ansiWidth ?? 0;
	const CONTROL_WIDTH = widthOptions.controlWidth ?? 0;
	const TAB_WIDTH = widthOptions.tabWidth ?? 8;
	const AMBIGUOUS_WIDTH = widthOptions.ambiguousWidth ?? 1;
	const EMOJI_WIDTH = widthOptions.emojiWidth ?? 2;
	const FULL_WIDTH_WIDTH = widthOptions.fullWidthWidth ?? 2;
	const REGULAR_WIDTH = widthOptions.regularWidth ?? 1;
	const WIDE_WIDTH = widthOptions.wideWidth ?? 2;
	let indexPrev = 0;
	let index = 0;
	let length = input.length;
	let lengthExtra = 0;
	let truncationEnabled = false;
	let truncationIndex = length;
	let truncationLimit = Math.max(0, LIMIT - ELLIPSIS_WIDTH);
	let unmatchedStart = 0;
	let unmatchedEnd = 0;
	let width = 0;
	let widthExtra = 0;
	outer: while (true) {
		if (unmatchedEnd > unmatchedStart || index >= length && index > indexPrev) {
			const unmatched = input.slice(unmatchedStart, unmatchedEnd) || input.slice(indexPrev, index);
			lengthExtra = 0;
			for (const char of unmatched.replaceAll(MODIFIER_RE, "")) {
				const codePoint = char.codePointAt(0) || 0;
				if (isFullWidth(codePoint)) widthExtra = FULL_WIDTH_WIDTH;
				else if (isWide(codePoint)) widthExtra = WIDE_WIDTH;
				else if (AMBIGUOUS_WIDTH !== REGULAR_WIDTH && isAmbiguous(codePoint)) widthExtra = AMBIGUOUS_WIDTH;
				else widthExtra = REGULAR_WIDTH;
				if (width + widthExtra > truncationLimit) truncationIndex = Math.min(truncationIndex, Math.max(unmatchedStart, indexPrev) + lengthExtra);
				if (width + widthExtra > LIMIT) {
					truncationEnabled = true;
					break outer;
				}
				lengthExtra += char.length;
				width += widthExtra;
			}
			unmatchedStart = unmatchedEnd = 0;
		}
		if (index >= length) break;
		LATIN_RE.lastIndex = index;
		if (LATIN_RE.test(input)) {
			lengthExtra = LATIN_RE.lastIndex - index;
			widthExtra = lengthExtra * REGULAR_WIDTH;
			if (width + widthExtra > truncationLimit) truncationIndex = Math.min(truncationIndex, index + Math.floor((truncationLimit - width) / REGULAR_WIDTH));
			if (width + widthExtra > LIMIT) {
				truncationEnabled = true;
				break;
			}
			width += widthExtra;
			unmatchedStart = indexPrev;
			unmatchedEnd = index;
			index = indexPrev = LATIN_RE.lastIndex;
			continue;
		}
		ANSI_RE.lastIndex = index;
		if (ANSI_RE.test(input)) {
			if (width + ANSI_WIDTH > truncationLimit) truncationIndex = Math.min(truncationIndex, index);
			if (width + ANSI_WIDTH > LIMIT) {
				truncationEnabled = true;
				break;
			}
			width += ANSI_WIDTH;
			unmatchedStart = indexPrev;
			unmatchedEnd = index;
			index = indexPrev = ANSI_RE.lastIndex;
			continue;
		}
		CONTROL_RE.lastIndex = index;
		if (CONTROL_RE.test(input)) {
			lengthExtra = CONTROL_RE.lastIndex - index;
			widthExtra = lengthExtra * CONTROL_WIDTH;
			if (width + widthExtra > truncationLimit) truncationIndex = Math.min(truncationIndex, index + Math.floor((truncationLimit - width) / CONTROL_WIDTH));
			if (width + widthExtra > LIMIT) {
				truncationEnabled = true;
				break;
			}
			width += widthExtra;
			unmatchedStart = indexPrev;
			unmatchedEnd = index;
			index = indexPrev = CONTROL_RE.lastIndex;
			continue;
		}
		TAB_RE.lastIndex = index;
		if (TAB_RE.test(input)) {
			lengthExtra = TAB_RE.lastIndex - index;
			widthExtra = lengthExtra * TAB_WIDTH;
			if (width + widthExtra > truncationLimit) truncationIndex = Math.min(truncationIndex, index + Math.floor((truncationLimit - width) / TAB_WIDTH));
			if (width + widthExtra > LIMIT) {
				truncationEnabled = true;
				break;
			}
			width += widthExtra;
			unmatchedStart = indexPrev;
			unmatchedEnd = index;
			index = indexPrev = TAB_RE.lastIndex;
			continue;
		}
		EMOJI_RE.lastIndex = index;
		if (EMOJI_RE.test(input)) {
			if (width + EMOJI_WIDTH > truncationLimit) truncationIndex = Math.min(truncationIndex, index);
			if (width + EMOJI_WIDTH > LIMIT) {
				truncationEnabled = true;
				break;
			}
			width += EMOJI_WIDTH;
			unmatchedStart = indexPrev;
			unmatchedEnd = index;
			index = indexPrev = EMOJI_RE.lastIndex;
			continue;
		}
		index += 1;
	}
	return {
		width: truncationEnabled ? truncationLimit : width,
		index: truncationEnabled ? truncationIndex : length,
		truncated: truncationEnabled,
		ellipsed: truncationEnabled && LIMIT >= ELLIPSIS_WIDTH
	};
};
const NO_TRUNCATION = {
	limit: Infinity,
	ellipsis: "",
	ellipsisWidth: 0
};
const fastStringWidth = (input, options = {}) => {
	return getStringTruncatedWidth(input, NO_TRUNCATION, options).width;
};
const ESC = "\x1B";
const CSI = "";
const END_CODE = 39;
const ANSI_ESCAPE_BELL = "\x07";
const ANSI_CSI = "[";
const ANSI_OSC = "]";
const ANSI_SGR_TERMINATOR = "m";
const ANSI_ESCAPE_LINK = `${ANSI_OSC}8;;`;
const GROUP_REGEX = new RegExp(`(?:\\${ANSI_CSI}(?<code>\\d+)m|\\${ANSI_ESCAPE_LINK}(?<uri>.*)${ANSI_ESCAPE_BELL})`, "y");
const getClosingCode = (openingCode) => {
	if (openingCode >= 30 && openingCode <= 37) return 39;
	if (openingCode >= 90 && openingCode <= 97) return 39;
	if (openingCode >= 40 && openingCode <= 47) return 49;
	if (openingCode >= 100 && openingCode <= 107) return 49;
	if (openingCode === 1 || openingCode === 2) return 22;
	if (openingCode === 3) return 23;
	if (openingCode === 4) return 24;
	if (openingCode === 7) return 27;
	if (openingCode === 8) return 28;
	if (openingCode === 9) return 29;
	if (openingCode === 0) return 0;
};
const wrapAnsiCode = (code) => `${ESC}${ANSI_CSI}${code}${ANSI_SGR_TERMINATOR}`;
const wrapAnsiHyperlink = (url) => `${ESC}${ANSI_ESCAPE_LINK}${url}${ANSI_ESCAPE_BELL}`;
const wrapWord = (rows, word, columns) => {
	const characters = word[Symbol.iterator]();
	let isInsideEscape = false;
	let isInsideLinkEscape = false;
	let lastRow = rows.at(-1);
	let visible = lastRow === void 0 ? 0 : fastStringWidth(lastRow);
	let currentCharacter = characters.next();
	let nextCharacter = characters.next();
	let rawCharacterIndex = 0;
	while (!currentCharacter.done) {
		const character = currentCharacter.value;
		const characterLength = fastStringWidth(character);
		if (visible + characterLength <= columns) rows[rows.length - 1] += character;
		else {
			rows.push(character);
			visible = 0;
		}
		if (character === ESC || character === CSI) {
			isInsideEscape = true;
			isInsideLinkEscape = word.startsWith(ANSI_ESCAPE_LINK, rawCharacterIndex + 1);
		}
		if (isInsideEscape) {
			if (isInsideLinkEscape) {
				if (character === ANSI_ESCAPE_BELL) {
					isInsideEscape = false;
					isInsideLinkEscape = false;
				}
			} else if (character === ANSI_SGR_TERMINATOR) isInsideEscape = false;
		} else {
			visible += characterLength;
			if (visible === columns && !nextCharacter.done) {
				rows.push("");
				visible = 0;
			}
		}
		currentCharacter = nextCharacter;
		nextCharacter = characters.next();
		rawCharacterIndex += character.length;
	}
	lastRow = rows.at(-1);
	if (!visible && lastRow !== void 0 && lastRow.length && rows.length > 1) rows[rows.length - 2] += rows.pop();
};
const stringVisibleTrimSpacesRight = (string) => {
	const words = string.split(" ");
	let last = words.length;
	while (last) {
		if (fastStringWidth(words[last - 1])) break;
		last--;
	}
	if (last === words.length) return string;
	return words.slice(0, last).join(" ") + words.slice(last).join("");
};
const exec = (string, columns, options = {}) => {
	if (options.trim !== false && string.trim() === "") return "";
	let returnValue = "";
	let escapeCode;
	let escapeUrl;
	const words = string.split(" ");
	let rows = [""];
	let rowLength = 0;
	for (let index = 0; index < words.length; index++) {
		const word = words[index];
		if (options.trim !== false) {
			const row = rows.at(-1) ?? "";
			const trimmed = row.trimStart();
			if (row.length !== trimmed.length) {
				rows[rows.length - 1] = trimmed;
				rowLength = fastStringWidth(trimmed);
			}
		}
		if (index !== 0) {
			if (rowLength >= columns && (options.wordWrap === false || options.trim === false)) {
				rows.push("");
				rowLength = 0;
			}
			if (rowLength || options.trim === false) {
				rows[rows.length - 1] += " ";
				rowLength++;
			}
		}
		const wordLength = fastStringWidth(word);
		if (options.hard && wordLength > columns) {
			const remainingColumns = columns - rowLength;
			const breaksStartingThisLine = 1 + Math.floor((wordLength - remainingColumns - 1) / columns);
			if (Math.floor((wordLength - 1) / columns) < breaksStartingThisLine) rows.push("");
			wrapWord(rows, word, columns);
			rowLength = fastStringWidth(rows.at(-1) ?? "");
			continue;
		}
		if (rowLength + wordLength > columns && rowLength && wordLength) {
			if (options.wordWrap === false && rowLength < columns) {
				wrapWord(rows, word, columns);
				rowLength = fastStringWidth(rows.at(-1) ?? "");
				continue;
			}
			rows.push("");
			rowLength = 0;
		}
		if (rowLength + wordLength > columns && options.wordWrap === false) {
			wrapWord(rows, word, columns);
			rowLength = fastStringWidth(rows.at(-1) ?? "");
			continue;
		}
		rows[rows.length - 1] += word;
		rowLength += wordLength;
	}
	if (options.trim !== false) rows = rows.map((row) => stringVisibleTrimSpacesRight(row));
	const preString = rows.join("\n");
	let inSurrogate = false;
	for (let i = 0; i < preString.length; i++) {
		const character = preString[i];
		returnValue += character;
		if (!inSurrogate) inSurrogate = character >= "\ud800" && character <= "\udbff";
		else continue;
		if (character === ESC || character === CSI) {
			GROUP_REGEX.lastIndex = i + 1;
			const groups = GROUP_REGEX.exec(preString)?.groups;
			if (groups?.code !== void 0) {
				const code = Number.parseFloat(groups.code);
				escapeCode = code === END_CODE ? void 0 : code;
			} else if (groups?.uri !== void 0) escapeUrl = groups.uri.length === 0 ? void 0 : groups.uri;
		}
		if (preString[i + 1] === "\n") {
			if (escapeUrl) returnValue += wrapAnsiHyperlink("");
			const closingCode = escapeCode ? getClosingCode(escapeCode) : void 0;
			if (escapeCode && closingCode) returnValue += wrapAnsiCode(closingCode);
		} else if (character === "\n") {
			if (escapeCode && getClosingCode(escapeCode)) returnValue += wrapAnsiCode(escapeCode);
			if (escapeUrl) returnValue += wrapAnsiHyperlink(escapeUrl);
		}
	}
	return returnValue;
};
const CRLF_OR_LF = /\r?\n/;
function wrapAnsi(string, columns, options) {
	return String(string).normalize().split(CRLF_OR_LF).map((line) => exec(line, columns, options)).join("\n");
}
const trimLines = (groups, initialLineCount, startIndex, endIndex, maxLines) => {
	let lineCount = initialLineCount;
	let removals = 0;
	for (let i = startIndex; i < endIndex; i++) {
		const group = groups[i];
		lineCount = lineCount - group.length;
		removals++;
		if (lineCount <= maxLines) break;
	}
	return {
		lineCount,
		removals
	};
};
const limitOptions = (params) => {
	const { cursor, options, style } = params;
	const output = params.output ?? process.stdout;
	const columns = rt(output);
	const columnPadding = params.columnPadding ?? 0;
	const rowPadding = params.rowPadding ?? 4;
	const maxWidth = columns - columnPadding;
	const rows = nt(output);
	const overflowFormat = color.dim("...");
	const paramMaxItems = params.maxItems ?? Number.POSITIVE_INFINITY;
	const outputMaxItems = Math.max(rows - rowPadding, 0);
	const maxItems = Math.max(Math.min(paramMaxItems, outputMaxItems), 5);
	let slidingWindowLocation = 0;
	if (cursor >= maxItems - 3) slidingWindowLocation = Math.max(Math.min(cursor - maxItems + 3, options.length - maxItems), 0);
	let shouldRenderTopEllipsis = maxItems < options.length && slidingWindowLocation > 0;
	let shouldRenderBottomEllipsis = maxItems < options.length && slidingWindowLocation + maxItems < options.length;
	const slidingWindowLocationEnd = Math.min(slidingWindowLocation + maxItems, options.length);
	const lineGroups = [];
	let lineCount = 0;
	if (shouldRenderTopEllipsis) lineCount++;
	if (shouldRenderBottomEllipsis) lineCount++;
	const slidingWindowLocationWithEllipsis = slidingWindowLocation + (shouldRenderTopEllipsis ? 1 : 0);
	const slidingWindowLocationEndWithEllipsis = slidingWindowLocationEnd - (shouldRenderBottomEllipsis ? 1 : 0);
	for (let i = slidingWindowLocationWithEllipsis; i < slidingWindowLocationEndWithEllipsis; i++) {
		const wrappedLines = wrapAnsi(style(options[i], i === cursor), maxWidth, {
			hard: true,
			trim: false
		}).split("\n");
		lineGroups.push(wrappedLines);
		lineCount += wrappedLines.length;
	}
	if (lineCount > outputMaxItems) {
		let precedingRemovals = 0;
		let followingRemovals = 0;
		let newLineCount = lineCount;
		const cursorGroupIndex = cursor - slidingWindowLocationWithEllipsis;
		const trimLinesLocal = (startIndex, endIndex) => trimLines(lineGroups, newLineCount, startIndex, endIndex, outputMaxItems);
		if (shouldRenderTopEllipsis) {
			({lineCount: newLineCount, removals: precedingRemovals} = trimLinesLocal(0, cursorGroupIndex));
			if (newLineCount > outputMaxItems) ({lineCount: newLineCount, removals: followingRemovals} = trimLinesLocal(cursorGroupIndex + 1, lineGroups.length));
		} else {
			({lineCount: newLineCount, removals: followingRemovals} = trimLinesLocal(cursorGroupIndex + 1, lineGroups.length));
			if (newLineCount > outputMaxItems) ({lineCount: newLineCount, removals: precedingRemovals} = trimLinesLocal(0, cursorGroupIndex));
		}
		if (precedingRemovals > 0) {
			shouldRenderTopEllipsis = true;
			lineGroups.splice(0, precedingRemovals);
		}
		if (followingRemovals > 0) {
			shouldRenderBottomEllipsis = true;
			lineGroups.splice(lineGroups.length - followingRemovals, followingRemovals);
		}
	}
	const result = [];
	if (shouldRenderTopEllipsis) result.push(overflowFormat);
	for (const lineGroup of lineGroups) for (const line of lineGroup) result.push(line);
	if (shouldRenderBottomEllipsis) result.push(overflowFormat);
	return result;
};
const confirm = (opts) => {
	const active = opts.active ?? "Yes";
	const inactive = opts.inactive ?? "No";
	return new kt({
		active,
		inactive,
		signal: opts.signal,
		input: opts.input,
		output: opts.output,
		initialValue: opts.initialValue ?? true,
		render() {
			const hasGuide = opts.withGuide ?? false;
			const nestedPrefix = "  ";
			const title = `${hasGuide ? `${color.gray(S_BAR)}\n` : ""}${symbol(this.state)} ${opts.message}\n`;
			const value = this.value ? active : inactive;
			switch (this.state) {
				case "submit": return `${title}${hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix}${color.dim(value)}\n`;
				case "cancel": return `${title}${hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix}${color.strikethrough(color.dim(value))}${hasGuide ? `\n${color.gray(S_BAR)}` : ""}\n`;
				default: {
					const defaultPrefix = hasGuide ? `${color.blue(S_BAR)} ` : nestedPrefix;
					const defaultPrefixEnd = hasGuide ? color.blue(S_BAR_END) : "";
					return `${title}${defaultPrefix}${this.value ? `${color.blue(S_POINTER_ACTIVE)} ${color.bold(active)}` : `${color.dim(" ")} ${color.dim(active)}`}${opts.vertical ? hasGuide ? `\n${color.blue(S_BAR)} ` : `\n${nestedPrefix}` : ` ${color.dim("/")} `}${!this.value ? `${color.blue(S_POINTER_ACTIVE)} ${color.bold(inactive)}` : `${color.dim(" ")} ${color.dim(inactive)}`}\n${defaultPrefixEnd}\n`;
				}
			}
		}
	}).prompt();
};
const log = {
	message: (message = [], { symbol = color.gray(S_BAR), secondarySymbol = color.gray(S_BAR), output = process.stdout, spacing = 1, withGuide } = {}) => {
		const parts = [];
		const hasGuide = withGuide ?? false;
		const spacingString = !hasGuide ? "" : secondarySymbol;
		const prefix = !hasGuide ? "" : `${symbol}  `;
		const secondaryPrefix = !hasGuide ? "" : `${secondarySymbol}  `;
		for (let i = 0; i < spacing; i++) parts.push(spacingString);
		const messageParts = Array.isArray(message) ? message : message.split("\n");
		if (messageParts.length > 0) {
			const [firstLine, ...lines] = messageParts;
			if (firstLine.length > 0) parts.push(`${prefix}${firstLine}`);
			else parts.push(hasGuide ? symbol : "");
			for (const ln of lines) if (ln.length > 0) parts.push(`${secondaryPrefix}${ln}`);
			else parts.push(hasGuide ? secondarySymbol : "");
		}
		output.write(`${parts.join("\n")}\n`);
	},
	info: (message, opts) => {
		log.message(message, {
			...opts,
			symbol: color.blue(S_INFO)
		});
	},
	success: (message, opts) => {
		log.message(message, {
			...opts,
			symbol: completeColor(S_SUCCESS)
		});
	},
	step: (message, opts) => {
		log.message(message, {
			...opts,
			symbol: completeColor(S_STEP_SUBMIT)
		});
	},
	warn: (message, opts) => {
		log.message(message, {
			...opts,
			symbol: color.yellow(S_WARN)
		});
	},
	warning: (message, opts) => {
		log.warn(message, opts);
	},
	error: (message, opts) => {
		log.message(message, {
			...opts,
			symbol: color.red(S_ERROR)
		});
	}
};
const cancel = (message = "", opts) => {
	(opts?.output ?? process.stdout).write(`${color.red(message)}\n\n`);
};
const intro = (title = "", opts) => {
	(opts?.output ?? process.stdout).write(`${title}\n\n`);
};
const outro = (message = "", opts) => {
	(opts?.output ?? process.stdout).write(`${message}\n\n`);
};
const computeLabel$1 = (label, format) => {
	return label.split("\n").map((line) => format(line)).join("\n");
};
const withMarkerAndCheckbox = (marker, checkbox, checkboxWidth, label, format, firstLineSuffix = "") => {
	const lines = label.split("\n");
	const continuationPrefix = `  ${" ".repeat(checkboxWidth)} `;
	if (lines.length === 1) return `${marker} ${checkbox} ${format(lines[0])}${firstLineSuffix}`;
	const [firstLine, ...rest] = lines;
	return [`${marker} ${checkbox} ${format(firstLine)}${firstLineSuffix}`, ...rest.map((line) => `${continuationPrefix}${format(line)}`)].join("\n");
};
const multiselect = (opts) => {
	const opt = (option, state) => {
		const label = option.label ?? String(option.value);
		const hint = option.hint ? ` ${color.gray(`(${option.hint})`)}` : "";
		if (state === "disabled") return withMarkerAndCheckbox(color.gray(" "), color.gray(S_CHECKBOX_INACTIVE), S_CHECKBOX_INACTIVE.length, label, (str) => color.strikethrough(color.gray(str)), option.hint ? ` ${color.dim(`(${option.hint ?? "disabled"})`)}` : "");
		if (state === "active") return withMarkerAndCheckbox(color.blue(S_POINTER_ACTIVE), color.blue(S_CHECKBOX_ACTIVE), S_CHECKBOX_ACTIVE.length, label, (text) => color.blue(color.bold(text)), hint);
		if (state === "selected") return withMarkerAndCheckbox(color.dim(" "), color.blue(S_CHECKBOX_SELECTED), S_CHECKBOX_SELECTED.length, label, color.dim, hint);
		if (state === "cancelled") return computeLabel$1(label, (text) => color.strikethrough(color.dim(text)));
		if (state === "active-selected") return withMarkerAndCheckbox(color.blue(S_POINTER_ACTIVE), color.blue(S_CHECKBOX_SELECTED), S_CHECKBOX_SELECTED.length, label, (text) => color.blue(color.bold(text)), hint);
		if (state === "submitted") return computeLabel$1(label, color.dim);
		return withMarkerAndCheckbox(color.dim(" "), color.dim(S_CHECKBOX_INACTIVE), S_CHECKBOX_INACTIVE.length, label, color.dim);
	};
	const required = opts.required ?? true;
	return new Lt({
		options: opts.options,
		signal: opts.signal,
		input: opts.input,
		output: opts.output,
		initialValues: opts.initialValues,
		required,
		cursorAt: opts.cursorAt,
		validate(selected) {
			if (required && (selected === void 0 || selected.length === 0)) return `Please select at least one option.\n${color.reset(color.dim(`Press ${color.gray(color.bgWhite(color.inverse(" space ")))} to select, ${color.gray(color.bgWhite(color.inverse(" enter ")))} to submit`))}`;
		},
		render() {
			const hasGuide = opts.withGuide ?? false;
			const nestedPrefix = "  ";
			const formatMessageLines = (message) => {
				return message.split("\n").map((line, index) => `${index === 0 ? `${symbol(this.state)} ` : nestedPrefix}${line}`).join("\n");
			};
			const wrappedMessage = hasGuide ? Bt(opts.output, opts.message, `${symbolBar(this.state)} `, `${symbol(this.state)} `) : formatMessageLines(opts.message);
			const title = `${hasGuide ? `${color.gray(S_BAR)}\n` : ""}${wrappedMessage}\n`;
			const value = this.value ?? [];
			const styleOption = (option, active) => {
				if (option.disabled) return opt(option, "disabled");
				const selected = value.includes(option.value);
				if (active && selected) return opt(option, "active-selected");
				if (selected) return opt(option, "selected");
				return opt(option, active ? "active" : "inactive");
			};
			switch (this.state) {
				case "submit": {
					const submitText = this.options.filter(({ value: optionValue }) => value.includes(optionValue)).map((option) => opt(option, "submitted")).join(color.dim(", ")) || color.dim("none");
					const submitPrefix = hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix;
					return `${title}${Bt(opts.output, submitText, submitPrefix)}\n`;
				}
				case "cancel": {
					const label = this.options.filter(({ value: optionValue }) => value.includes(optionValue)).map((option) => opt(option, "cancelled")).join(color.dim(", "));
					if (label.trim() === "") return hasGuide ? `${title}${color.gray(S_BAR)}\n` : `${title.trimEnd()}\n`;
					const cancelPrefix = hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix;
					const wrappedLabel = Bt(opts.output, label, cancelPrefix);
					return hasGuide ? `${title}${wrappedLabel}\n${color.gray(S_BAR)}\n` : `${title}${wrappedLabel}\n`;
				}
				case "error": {
					const prefix = hasGuide ? `${color.yellow(S_BAR)} ` : nestedPrefix;
					const footer = hasGuide ? this.error.split("\n").map((ln, i) => i === 0 ? `${color.yellow(S_BAR_END)} ${color.yellow(ln)}` : `  ${ln}`).join("\n") : `${nestedPrefix}${color.yellow(this.error)}`;
					const titleLineCount = title.split("\n").length;
					const footerLineCount = footer.split("\n").length + 1;
					return `${title}${prefix}${limitOptions({
						output: opts.output,
						options: this.options,
						cursor: this.cursor,
						maxItems: opts.maxItems,
						columnPadding: prefix.length,
						rowPadding: titleLineCount + footerLineCount,
						style: styleOption
					}).join(`\n${prefix}`)}\n${footer}\n`;
				}
				default: {
					const prefix = hasGuide ? `${color.blue(S_BAR)} ` : nestedPrefix;
					const titleLineCount = title.split("\n").length;
					const footerLineCount = hasGuide ? 2 : 1;
					return `${title}${prefix}${limitOptions({
						output: opts.output,
						options: this.options,
						cursor: this.cursor,
						maxItems: opts.maxItems,
						columnPadding: prefix.length,
						rowPadding: titleLineCount + footerLineCount,
						style: styleOption
					}).join(`\n${prefix}`)}\n${hasGuide ? color.blue(S_BAR_END) : ""}\n`;
				}
			}
		}
	}).prompt();
};
const defaultNoteFormatter = (line) => color.dim(line);
const wrapWithFormat = (message, width, format) => {
	const opts = {
		hard: true,
		trim: false
	};
	const wrapMsg = wrapAnsi(message, width, opts).split("\n");
	const maxWidthNormal = wrapMsg.reduce((sum, ln) => Math.max(fastStringWidth(ln), sum), 0);
	return wrapAnsi(message, width - (wrapMsg.map(format).reduce((sum, ln) => Math.max(fastStringWidth(ln), sum), 0) - maxWidthNormal), opts);
};
const note = (message = "", title = "", opts) => {
	const output = opts?.output ?? process$1.stdout;
	const hasGuide = opts?.withGuide ?? false;
	const format = opts?.format ?? defaultNoteFormatter;
	const lines = [
		"",
		...wrapWithFormat(message, rt(output) - 6, format).split("\n").map(format),
		""
	];
	const titleLen = fastStringWidth(title);
	const len = Math.max(lines.reduce((sum, ln) => {
		const width = fastStringWidth(ln);
		return width > sum ? width : sum;
	}, 0), titleLen) + 2;
	const lineSymbol = hasGuide ? color.gray(S_BAR) : " ";
	const msg = lines.map((ln) => `${lineSymbol}  ${ln}${" ".repeat(len - fastStringWidth(ln))}${lineSymbol}`).join("\n");
	const leadingBorder = hasGuide ? `${color.gray(S_BAR)}\n` : "";
	const bottomLeft = hasGuide ? S_CONNECT_LEFT : S_CORNER_BOTTOM_LEFT;
	output.write(`${leadingBorder}${completeColor(S_STEP_SUBMIT)}  ${color.reset(title)} ${color.gray(S_BAR_H.repeat(Math.max(len - titleLen - 1, 1)) + S_CORNER_TOP_RIGHT)}\n${msg}\n${color.gray(bottomLeft + S_BAR_H.repeat(len + 2) + S_CORNER_BOTTOM_RIGHT)}\n`);
};
const defaultStyleFn = color.magenta;
const removeTrailingDots = (msg) => {
	return msg.replace(/\.+$/, "");
};
const formatTimer = (durationMs) => {
	const duration = durationMs / 1e3;
	const min = Math.floor(duration / 60);
	const secs = Math.floor(duration % 60);
	return color.gray(min > 0 ? `(${min}m ${secs}s)` : `(${secs}s)`);
};
const spinner = ({ indicator = "dots", onCancel, output = process.stdout, cancelMessage, errorMessage, frames = unicode ? [
	"◒",
	"◐",
	"◓",
	"◑"
] : [
	"•",
	"o",
	"O",
	"0"
], delay = unicode ? 80 : 120, signal, ...opts } = {}) => {
	const isCI$1 = isCI();
	let unblock;
	let loop;
	let isSpinnerActive = false;
	let isCancelled = false;
	let _message = "";
	let _prevMessage;
	let _origin = performance.now();
	let _elapsedMs = 0;
	const columns = rt(output);
	const styleFn = opts?.styleFrame ?? defaultStyleFn;
	const getElapsedMs = () => {
		if (!isSpinnerActive) return _elapsedMs;
		return _elapsedMs + (performance.now() - _origin);
	};
	const handleExit = (code) => {
		const msg = code > 1 ? errorMessage ?? _.messages.error : cancelMessage ?? _.messages.cancel;
		isCancelled = code === 1;
		if (isSpinnerActive) {
			_stop(msg, code);
			if (isCancelled && typeof onCancel === "function") onCancel();
		}
	};
	const errorEventHandler = () => handleExit(2);
	const signalEventHandler = () => handleExit(1);
	const registerHooks = () => {
		process.on("uncaughtExceptionMonitor", errorEventHandler);
		process.on("unhandledRejection", errorEventHandler);
		process.on("SIGINT", signalEventHandler);
		process.on("SIGTERM", signalEventHandler);
		process.on("exit", handleExit);
		if (signal) signal.addEventListener("abort", signalEventHandler);
	};
	const clearHooks = () => {
		process.removeListener("uncaughtExceptionMonitor", errorEventHandler);
		process.removeListener("unhandledRejection", errorEventHandler);
		process.removeListener("SIGINT", signalEventHandler);
		process.removeListener("SIGTERM", signalEventHandler);
		process.removeListener("exit", handleExit);
		if (signal) signal.removeEventListener("abort", signalEventHandler);
	};
	const clearPrevMessage = () => {
		if (_prevMessage === void 0) return;
		if (isCI$1) output.write("\n");
		const prevLines = wrapAnsi(_prevMessage, columns, {
			hard: true,
			trim: false
		}).split("\n");
		if (prevLines.length > 1) output.write(import_src.cursor.up(prevLines.length - 1));
		output.write(import_src.cursor.to(0));
		output.write(import_src.erase.down());
	};
	const hasGuide = opts.withGuide ?? false;
	const startLoop = () => {
		isSpinnerActive = true;
		unblock = xt({ output });
		_origin = performance.now();
		_prevMessage = void 0;
		if (hasGuide) output.write(`${color.gray(S_BAR)}\n`);
		let frameIndex = 0;
		let indicatorTimer = 0;
		registerHooks();
		loop = setInterval(() => {
			if (isCI$1 && _message === _prevMessage) return;
			clearPrevMessage();
			_prevMessage = _message;
			const frame = styleFn(frames[frameIndex]);
			let outputMessage;
			if (isCI$1) outputMessage = `${frame}  ${_message}...`;
			else if (indicator === "timer") outputMessage = `${frame}  ${_message} ${formatTimer(getElapsedMs())}`;
			else {
				const loadingDots = ".".repeat(Math.floor(indicatorTimer)).slice(0, 3);
				outputMessage = `${frame}  ${_message}${loadingDots}`;
			}
			const wrapped = wrapAnsi(outputMessage, columns, {
				hard: true,
				trim: false
			});
			output.write(wrapped);
			frameIndex = frameIndex + 1 < frames.length ? frameIndex + 1 : 0;
			indicatorTimer = indicatorTimer < 4 ? indicatorTimer + .125 : 0;
		}, delay);
	};
	const start = (msg = "") => {
		_elapsedMs = 0;
		_message = removeTrailingDots(msg);
		startLoop();
	};
	const _stop = (msg = "", code = 0, silent = false, preserveElapsed = false) => {
		if (!isSpinnerActive) return;
		isSpinnerActive = false;
		clearInterval(loop);
		clearPrevMessage();
		const elapsedMs = getElapsedMs();
		const step = code === 0 ? completeColor(S_STEP_SUBMIT) : code === 1 ? color.red(S_STEP_CANCEL) : color.red(S_STEP_ERROR);
		_message = msg ?? _message;
		if (!silent) if (indicator === "timer") output.write(`${step} ${_message} ${formatTimer(elapsedMs)}\n\n`);
		else output.write(`${step} ${_message}\n\n`);
		if (!preserveElapsed) _elapsedMs = 0;
		_prevMessage = void 0;
		clearHooks();
		unblock();
	};
	const pause = () => {
		if (!isSpinnerActive) return;
		_elapsedMs = getElapsedMs();
		_stop(_message, 0, true, true);
	};
	const resume = (msg = _message) => {
		if (isSpinnerActive) return;
		_message = removeTrailingDots(msg);
		startLoop();
	};
	const stop = (msg = "") => _stop(msg, 0);
	const cancel = (msg = "") => _stop(msg, 1);
	const error = (msg = "") => _stop(msg, 2);
	const clear = () => _stop("", 0, true);
	const message = (msg = "") => {
		_message = removeTrailingDots(msg ?? _message);
	};
	return {
		start,
		pause,
		resume,
		stop,
		message,
		cancel,
		error,
		clear,
		get isCancelled() {
			return isCancelled;
		}
	};
};
unicodeOr("─", "-"), unicodeOr("━", "="), unicodeOr("█", "#");
const computeLabel = (label, format) => {
	if (!label.includes("\n")) return format(label);
	return label.split("\n").map((line) => format(line)).join("\n");
};
const withMarker = (marker, label, format, firstLineSuffix = "") => {
	const lines = label.split("\n");
	if (lines.length === 1) return `${marker} ${format(lines[0])}${firstLineSuffix}`;
	const [firstLine, ...rest] = lines;
	return [`${marker} ${format(firstLine)}${firstLineSuffix}`, ...rest.map((line) => `  ${format(line)}`)].join("\n");
};
const select = (opts) => {
	const opt = (option, state) => {
		const label = option.label ?? String(option.value);
		const hint = option.hint ? `: ${color.gray(option.hint)}` : "";
		switch (state) {
			case "disabled": return withMarker(color.gray(" "), label, (text) => color.strikethrough(color.gray(text)), option.hint ? `: ${color.gray(option.hint ?? "disabled")}` : "");
			case "selected": return computeLabel(label, color.dim);
			case "active": return withMarker(color.blue(S_POINTER_ACTIVE), label, (text) => color.blue(color.bold(text)), hint);
			case "cancelled": return computeLabel(label, (str) => color.strikethrough(color.dim(str)));
			default: return withMarker(color.dim(" "), label, (text) => text, hint);
		}
	};
	return new Tt({
		options: opts.options,
		signal: opts.signal,
		input: opts.input,
		output: opts.output,
		initialValue: opts.initialValue,
		render() {
			const hasGuide = opts.withGuide ?? false;
			const nestedPrefix = "  ";
			const formatMessageLines = (message) => {
				return message.split("\n").map((line, index) => `${index === 0 ? `${symbol(this.state)} ` : nestedPrefix}${line}`).join("\n");
			};
			const hasMessage = opts.message.trim().length > 0;
			const messageLines = !hasMessage ? "" : hasGuide ? Bt(opts.output, opts.message, `${symbolBar(this.state)} `, `${symbol(this.state)} `) : formatMessageLines(opts.message);
			const title = hasMessage ? `${hasGuide ? `${color.gray(S_BAR)}\n` : ""}${messageLines}\n` : "";
			switch (this.state) {
				case "submit": {
					const submitPrefix = hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix;
					return `${title}${Bt(opts.output, opt(this.options[this.cursor], "selected"), submitPrefix)}\n`;
				}
				case "cancel": {
					const cancelPrefix = hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix;
					return `${title}${Bt(opts.output, opt(this.options[this.cursor], "cancelled"), cancelPrefix)}${hasGuide ? `\n${color.gray(S_BAR)}` : ""}\n`;
				}
				default: {
					const prefix = hasGuide ? `${color.blue(S_BAR)} ` : nestedPrefix;
					const prefixEnd = hasGuide ? color.blue(S_BAR_END) : "";
					const titleLineCount = title ? title.split("\n").length : 0;
					const footerLineCount = hasGuide ? 2 : 1;
					return `${title}${prefix}${limitOptions({
						output: opts.output,
						cursor: this.cursor,
						options: this.options,
						maxItems: opts.maxItems,
						columnPadding: prefix.length,
						rowPadding: titleLineCount + footerLineCount,
						style: (item, active) => opt(item, item.disabled ? "disabled" : active ? "active" : "inactive")
					}).join(`\n${prefix}`)}\n${prefixEnd}\n`;
				}
			}
		}
	}).prompt();
};
const prefix = "   ";
const stream = {
	message: async (iterable, { symbol = "" } = {}) => {
		process.stdout.write(symbol ? `${symbol}  ` : "");
		const initialWidth = symbol ? 3 : 0;
		let lineWidth = initialWidth;
		for await (let chunk of iterable) {
			chunk = chunk.replace(/\n/g, `\n${prefix}`);
			if (chunk.includes("\n")) lineWidth = initialWidth + stripVTControlCharacters(chunk.slice(chunk.lastIndexOf("\n"))).length;
			const chunkLen = stripVTControlCharacters(chunk).length;
			if (lineWidth + chunkLen < process.stdout.columns) {
				lineWidth += chunkLen;
				process.stdout.write(chunk);
			} else {
				process.stdout.write(`\n${prefix}${chunk.trimStart()}`);
				lineWidth = initialWidth + stripVTControlCharacters(chunk.trimStart()).length;
			}
		}
		process.stdout.write("\n");
	},
	info: (iterable) => {
		return stream.message(iterable, { symbol: color.blue(S_INFO) });
	},
	success: (iterable) => {
		return stream.message(iterable, { symbol: completeColor(S_SUCCESS) });
	},
	step: (iterable) => {
		return stream.message(iterable, { symbol: completeColor(S_STEP_SUBMIT) });
	},
	warn: (iterable) => {
		return stream.message(iterable, { symbol: color.yellow(S_WARN) });
	},
	warning: (iterable) => {
		return stream.warn(iterable);
	},
	error: (iterable) => {
		return stream.message(iterable, { symbol: color.red(S_ERROR) });
	}
};
const text = (opts) => {
	return new $t({
		validate: opts.validate,
		placeholder: opts.placeholder,
		defaultValue: opts.defaultValue,
		initialValue: opts.initialValue,
		output: opts.output,
		signal: opts.signal,
		input: opts.input,
		render() {
			const hasGuide = opts?.withGuide ?? false;
			const nestedPrefix = "  ";
			const title = `${hasGuide ? `${color.gray(S_BAR)}\n` : ""}${symbol(this.state)} ${opts.message}\n`;
			const placeholder = opts.placeholder ? color.inverse(opts.placeholder[0]) + color.dim(opts.placeholder.slice(1)) : color.inverse(color.hidden("_"));
			const userInput = !this.userInput ? placeholder : this.userInputWithCursor;
			const value = this.value ?? "";
			switch (this.state) {
				case "error": {
					const errorText = this.error ? ` ${color.yellow(this.error)}` : "";
					const errorPrefix = hasGuide ? `${color.yellow(S_BAR)} ` : nestedPrefix;
					const errorPrefixEnd = hasGuide ? color.yellow(S_BAR_END) : "";
					return `${title.trim()}\n${errorPrefix}${userInput}\n${errorPrefixEnd}${errorText}\n`;
				}
				case "submit": {
					const valueText = value ? color.dim(value) : "";
					return `${title}${hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix}${valueText}\n`;
				}
				case "cancel": {
					const valueText = value ? color.strikethrough(color.dim(value)) : "";
					const cancelPrefix = hasGuide ? `${color.gray(S_BAR)} ` : nestedPrefix;
					return `${title}${cancelPrefix}${valueText}${value.trim() ? `\n${cancelPrefix}` : ""}\n`;
				}
				default: return `${title}${hasGuide ? `${color.blue(S_BAR)} ` : nestedPrefix}${userInput}\n${hasGuide ? color.blue(S_BAR_END) : ""}\n`;
			}
		}
	}).prompt();
};
//#endregion
//#region src/types/package.ts
const PackageManager = {
	pnpm: "pnpm",
	npm: "npm",
	yarn: "yarn"
};
const DependencyType = {
	dependencies: "dependencies",
	devDependencies: "devDependencies",
	peerDependencies: "peerDependencies",
	optionalDependencies: "optionalDependencies"
};
//#endregion
//#region src/utils/command.ts
async function runCommandSilently(options) {
	const child = spawn(options.command, options.args, {
		stdio: "pipe",
		cwd: options.cwd,
		env: options.envs
	});
	return await new Promise((resolve, reject) => {
		const stdout = [];
		const stderr = [];
		child.stdout?.on("data", (data) => {
			stdout.push(data);
		});
		child.stderr?.on("data", (data) => {
			stderr.push(data);
		});
		child.on("close", (code) => {
			resolve({
				exitCode: code ?? 0,
				stdout: Buffer.concat(stdout),
				stderr: Buffer.concat(stderr)
			});
		});
		child.on("error", (err) => {
			reject(err);
		});
	});
}
//#endregion
//#region src/utils/path.ts
function findPkgRoot() {
	let dir = import.meta.dirname;
	while (dir !== path.dirname(dir)) {
		if (fs.existsSync(path.join(dir, "package.json"))) return dir;
		dir = path.dirname(dir);
	}
	return dir;
}
const pkgRoot = findPkgRoot();
const templatesDir = path.join(pkgRoot, "templates");
const rulesDir = path.join(pkgRoot, "rules");
function displayRelative(to, from = process.cwd()) {
	return path.relative(from, to).replaceAll("\\", "/");
}
//#endregion
//#region src/resolve-fmt.ts
/**
* Oxfmt tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the oxfmt binary path
* using Node.js module resolution. The resolved path is passed back
* to the Rust core, which then executes oxfmt for code formatting.
*
* Used for: `vite-plus fmt` command
*
* Oxfmt is a fast JavaScript/TypeScript formatter written in Rust that
* provides high-performance code formatting capabilities.
*/
/**
* Resolves the oxfmt binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the oxfmt binary
*   - envs: Environment variables to set when executing oxfmt
*
* The environment variables provide runtime context to oxfmt,
* including Node.js version information and package manager details.
*/
async function fmt() {
	return {
		binPath: resolve$1("oxfmt/bin/oxfmt"),
		envs: { ...DEFAULT_ENVS }
	};
}
//#endregion
//#region src/utils/prompts.ts
function cancelAndExit(message = "Operation cancelled", exitCode = 0) {
	cancel(message);
	process.exit(exitCode);
}
async function selectPackageManager(interactive, silent = false) {
	if (interactive) {
		const selected = await select({
			message: "Which package manager would you like to use?",
			options: [
				{
					value: PackageManager.pnpm,
					hint: "recommended"
				},
				{ value: PackageManager.yarn },
				{ value: PackageManager.npm }
			],
			initialValue: PackageManager.pnpm
		});
		if (Ct(selected)) cancelAndExit();
		return selected;
	} else {
		if (!silent) log.info(`Using default package manager: ${accent(PackageManager.pnpm)}`);
		return PackageManager.pnpm;
	}
}
async function downloadPackageManager$1(packageManager, version, interactive, silent = false) {
	const spinner = silent ? getSilentSpinner() : getSpinner(interactive);
	spinner.start(`${packageManager}@${version} installing...`);
	const downloadResult = await downloadPackageManager({
		name: packageManager,
		version
	});
	spinner.stop(`${packageManager}@${downloadResult.version} installed`);
	return downloadResult;
}
async function runViteInstall(cwd, interactive, extraArgs, options) {
	if (process.env.VITE_PLUS_SKIP_INSTALL) return {
		durationMs: 0,
		status: "skipped"
	};
	const spinner = options?.silent ? getSilentSpinner() : getSpinner(interactive);
	const startTime = Date.now();
	spinner.start(`Installing dependencies...`);
	const { exitCode, stderr, stdout } = await runCommandSilently({
		command: process.env.VITE_PLUS_CLI_BIN ?? "vp",
		args: ["install", ...extraArgs ?? []],
		cwd,
		envs: process.env
	});
	if (exitCode === 0) {
		spinner.stop(`Dependencies installed`);
		return {
			durationMs: Date.now() - startTime,
			exitCode,
			status: "installed"
		};
	} else {
		spinner.stop(`Install failed`);
		log.info(stdout.toString());
		log.error(stderr.toString());
		log.info(`You may need to run "vp install" manually in ${cwd}`);
		return {
			durationMs: Date.now() - startTime,
			exitCode,
			status: "failed"
		};
	}
}
async function runViteFmt(cwd, interactive, paths, options) {
	const spinner = options?.silent ? getSilentSpinner() : getSpinner(interactive);
	const startTime = Date.now();
	spinner.start(`Formatting code...`);
	const { binPath, envs } = await fmt();
	const { exitCode, stderr, stdout } = await runCommandSilently({
		command: binPath,
		args: ["--write", ...paths ?? []],
		cwd,
		envs: {
			...process.env,
			...envs
		}
	});
	if (exitCode === 0) {
		spinner.stop(`Code formatted`);
		return {
			durationMs: Date.now() - startTime,
			exitCode,
			status: "formatted"
		};
	} else {
		spinner.stop(`Format failed`);
		log.info(stdout.toString());
		log.error(stderr.toString());
		const relativePaths = (paths ?? []).length > 0 ? ` ${(paths ?? []).join(" ")}` : "";
		log.info(`You may need to run "vp fmt --write${relativePaths}" manually in ${cwd}`);
		return {
			durationMs: Date.now() - startTime,
			exitCode,
			status: "failed"
		};
	}
}
async function upgradeYarn(cwd, interactive, silent = false) {
	const spinner = silent ? getSilentSpinner() : getSpinner(interactive);
	spinner.start(`Running yarn set version stable...`);
	const { exitCode, stderr, stdout } = await runCommandSilently({
		command: "yarn",
		args: [
			"set",
			"version",
			"stable"
		],
		cwd,
		envs: process.env
	});
	if (exitCode === 0) spinner.stop(`Yarn upgraded to stable version`);
	else {
		spinner.stop(`yarn upgrade failed`);
		log.info(stdout.toString());
		log.error(stderr.toString());
	}
}
async function promptGitHooks(options) {
	if (options.hooks === false) return false;
	if (options.hooks === true) return true;
	if (options.interactive) {
		const selected = await confirm({
			message: "Set up pre-commit hooks to run formatting, linting, and type checking with auto-fixes?",
			initialValue: true
		});
		if (Ct(selected)) {
			cancelAndExit();
			return false;
		}
		return selected;
	}
	return true;
}
function defaultInteractive() {
	return !process.env.CI && process.stdin.isTTY;
}
function getSpinner(interactive) {
	if (interactive) return spinner();
	return {
		start: (msg) => {
			if (msg) log.info(msg);
		},
		stop: (msg) => {
			if (msg) log.info(msg);
		},
		message: (msg) => {
			if (msg) log.info(msg);
		}
	};
}
function getSilentSpinner() {
	return {
		start: () => {},
		stop: () => {},
		message: () => {}
	};
}
//#endregion
//#region src/utils/agent.ts
const DEFAULT_MCP_HINT = "Run `npx vp mcp` — this starts a stdio MCP server. See your agent's docs for how to add a local MCP server.";
const agents = {
	"claude-code": {
		displayName: "Claude Code",
		skillsDir: ".claude/skills",
		detect: (root) => fs.existsSync(path.join(root, ".claude")) || fs.existsSync(path.join(root, "CLAUDE.md")),
		mcpConfig: [{
			filePath: ".claude/settings.json",
			rootKey: "mcpServers"
		}, {
			filePath: ".claude/settings.local.json",
			rootKey: "mcpServers"
		}]
	},
	amp: {
		displayName: "Amp",
		skillsDir: ".agents/skills",
		detect: (root) => fs.existsSync(path.join(root, ".amp")),
		mcpHint: DEFAULT_MCP_HINT
	},
	codex: {
		displayName: "Codex",
		skillsDir: ".agents/skills",
		detect: (root) => fs.existsSync(path.join(root, ".codex")),
		mcpHint: "codex mcp add vite-plus -- npx vp mcp"
	},
	cursor: {
		displayName: "Cursor",
		skillsDir: ".agents/skills",
		detect: (root) => fs.existsSync(path.join(root, ".cursor")),
		mcpConfig: [{
			filePath: ".cursor/mcp.json",
			rootKey: "mcpServers"
		}]
	},
	windsurf: {
		displayName: "Windsurf",
		skillsDir: ".windsurf/skills",
		detect: (root) => fs.existsSync(path.join(root, ".windsurf")),
		mcpConfig: [{
			filePath: ".windsurf/mcp.json",
			rootKey: "mcpServers"
		}]
	},
	"gemini-cli": {
		displayName: "Gemini CLI",
		skillsDir: ".agents/skills",
		detect: (root) => fs.existsSync(path.join(root, ".gemini")),
		mcpHint: "gemini mcp add vite-plus -- npx vp mcp"
	},
	"github-copilot": {
		displayName: "GitHub Copilot",
		skillsDir: ".agents/skills",
		detect: (root) => fs.existsSync(path.join(root, ".github", "copilot-instructions.md")) || fs.existsSync(path.join(root, ".vscode", "mcp.json")),
		mcpConfig: [{
			filePath: ".vscode/mcp.json",
			rootKey: "servers",
			extraFields: { type: "stdio" }
		}]
	},
	cline: {
		displayName: "Cline",
		skillsDir: ".cline/skills",
		detect: (root) => fs.existsSync(path.join(root, ".cline")),
		mcpHint: DEFAULT_MCP_HINT
	},
	roo: {
		displayName: "Roo Code",
		skillsDir: ".roo/skills",
		detect: (root) => fs.existsSync(path.join(root, ".roo")),
		mcpConfig: [{
			filePath: ".roo/mcp.json",
			rootKey: "mcpServers"
		}]
	},
	kilo: {
		displayName: "Kilo Code",
		skillsDir: ".kilocode/skills",
		detect: (root) => fs.existsSync(path.join(root, ".kilocode")),
		mcpHint: DEFAULT_MCP_HINT
	},
	continue: {
		displayName: "Continue",
		skillsDir: ".continue/skills",
		detect: (root) => fs.existsSync(path.join(root, ".continue")),
		mcpHint: DEFAULT_MCP_HINT
	},
	goose: {
		displayName: "Goose",
		skillsDir: ".goose/skills",
		detect: (root) => fs.existsSync(path.join(root, ".goose")),
		mcpHint: DEFAULT_MCP_HINT
	},
	opencode: {
		displayName: "OpenCode",
		skillsDir: ".agents/skills",
		detect: (root) => fs.existsSync(path.join(root, ".opencode")),
		mcpHint: DEFAULT_MCP_HINT
	},
	trae: {
		displayName: "Trae",
		skillsDir: ".trae/skills",
		detect: (root) => fs.existsSync(path.join(root, ".trae")),
		mcpHint: DEFAULT_MCP_HINT
	},
	junie: {
		displayName: "Junie",
		skillsDir: ".junie/skills",
		detect: (root) => fs.existsSync(path.join(root, ".junie")),
		mcpHint: DEFAULT_MCP_HINT
	},
	"kiro-cli": {
		displayName: "Kiro CLI",
		skillsDir: ".kiro/skills",
		detect: (root) => fs.existsSync(path.join(root, ".kiro")),
		mcpHint: DEFAULT_MCP_HINT
	},
	zencoder: {
		displayName: "Zencoder",
		skillsDir: ".zencoder/skills",
		detect: (root) => fs.existsSync(path.join(root, ".zencoder")),
		mcpHint: DEFAULT_MCP_HINT
	},
	"qwen-code": {
		displayName: "Qwen Code",
		skillsDir: ".qwen/skills",
		detect: (root) => fs.existsSync(path.join(root, ".qwen")),
		mcpHint: DEFAULT_MCP_HINT
	}
};
function getAgentById(id) {
	return agents[id];
}
function detectAgents(root) {
	return Object.values(agents).filter((a) => a.detect(root));
}
const AGENT_ALIASES = {
	chatgpt: "chatgpt-codex",
	codex: "chatgpt-codex"
};
const AGENTS = [
	{
		id: "chatgpt-codex",
		label: "ChatGPT (Codex)",
		targetPath: "AGENTS.md"
	},
	{
		id: "claude",
		label: "Claude Code",
		targetPath: "CLAUDE.md"
	},
	{
		id: "gemini",
		label: "Gemini CLI",
		targetPath: "GEMINI.md"
	},
	{
		id: "copilot",
		label: "GitHub Copilot",
		targetPath: ".github/copilot-instructions.md"
	},
	{
		id: "cursor",
		label: "Cursor",
		targetPath: ".cursor/rules/viteplus.mdc"
	},
	{
		id: "jetbrains",
		label: "JetBrains AI Assistant",
		targetPath: ".aiassistant/rules/viteplus.md"
	},
	{
		id: "amp",
		label: "Amp",
		targetPath: "AGENTS.md"
	},
	{
		id: "opencode",
		label: "OpenCode",
		targetPath: "AGENTS.md"
	},
	{
		id: "other",
		label: "Other",
		targetPath: "AGENTS.md"
	}
];
const AGENT_STANDARD_PATH = "AGENTS.md";
const AGENT_INSTRUCTIONS_START_MARKER = "<!--VITE PLUS START-->";
const AGENT_INSTRUCTIONS_END_MARKER = "<!--VITE PLUS END-->";
async function selectAgentTargetPaths({ interactive, agent, onCancel }) {
	if (agent === false) return;
	if (interactive && !agent) {
		const selectedAgents = await multiselect({
			message: "Which agents are you using?",
			options: AGENTS.map((option) => ({
				label: option.label,
				value: option.id,
				hint: option.targetPath
			})),
			initialValues: ["chatgpt-codex"],
			required: false
		});
		if (Ct(selectedAgents)) {
			onCancel();
			return;
		}
		if (selectedAgents.length === 0) return;
		return resolveAgentTargetPaths(selectedAgents);
	}
	return resolveAgentTargetPaths(agent ?? "other");
}
function detectExistingAgentTargetPaths(projectRoot) {
	const detectedPaths = [];
	const seenTargetPaths = /* @__PURE__ */ new Set();
	for (const option of AGENTS) {
		if (seenTargetPaths.has(option.targetPath)) continue;
		seenTargetPaths.add(option.targetPath);
		const targetPath = path.join(projectRoot, option.targetPath);
		if (fs.existsSync(targetPath) && !fs.lstatSync(targetPath).isSymbolicLink()) detectedPaths.push(option.targetPath);
	}
	return detectedPaths.length > 0 ? detectedPaths : void 0;
}
function hasExistingAgentInstructions(projectRoot) {
	const targetPaths = detectExistingAgentTargetPaths(projectRoot);
	if (!targetPaths) return false;
	for (const targetPath of targetPaths) if (fs.readFileSync(path.join(projectRoot, targetPath), "utf-8").includes(AGENT_INSTRUCTIONS_START_MARKER)) return true;
	return false;
}
function resolveAgentTargetPaths(agent) {
	const agentNames = parseAgentNames(agent);
	const resolvedAgentNames = agentNames.length > 0 ? agentNames : ["other"];
	const dedupedTargetPaths = [];
	const seenTargetPaths = /* @__PURE__ */ new Set();
	for (const name of resolvedAgentNames) {
		const targetPath = resolveSingleAgentTargetPath(name);
		if (seenTargetPaths.has(targetPath)) continue;
		seenTargetPaths.add(targetPath);
		dedupedTargetPaths.push(targetPath);
	}
	return dedupedTargetPaths;
}
function parseAgentNames(agent) {
	if (!agent) return [];
	return (Array.isArray(agent) ? agent : [agent]).filter((value) => typeof value === "string").flatMap((value) => value.split(",")).map((value) => value.trim()).filter((value) => value.length > 0);
}
function resolveSingleAgentTargetPath(agent) {
	const normalized = normalizeAgentName(agent);
	const alias = AGENT_ALIASES[normalized];
	const resolved = alias ? normalizeAgentName(alias) : normalized;
	return AGENTS.find((option) => normalizeAgentName(option.id) === resolved || normalizeAgentName(option.label) === resolved)?.targetPath ?? AGENTS[AGENTS.length - 1].targetPath;
}
/**
* Detect agent instruction files that would conflict (exist without markers).
* Returns only files that need a user decision (append or skip).
* Read-only — does not write or modify any files.
*/
async function detectAgentConflicts({ projectRoot, targetPaths }) {
	if (!targetPaths || targetPaths.length === 0) return [];
	const sourcePath = path.join(pkgRoot, "AGENTS.md");
	if (!fs.existsSync(sourcePath)) return [];
	const incomingContent = await fsPromises.readFile(sourcePath, "utf-8");
	const shouldLinkToAgents = targetPaths.includes(AGENT_STANDARD_PATH);
	const orderedPaths = shouldLinkToAgents ? [AGENT_STANDARD_PATH, ...targetPaths.filter((p) => p !== AGENT_STANDARD_PATH)] : targetPaths;
	const conflicts = [];
	const seenDestinationPaths = /* @__PURE__ */ new Set();
	const seenRealPaths = /* @__PURE__ */ new Set();
	for (const targetPathToCheck of orderedPaths) {
		const destinationPath = path.join(projectRoot, targetPathToCheck);
		const destinationKey = path.resolve(destinationPath);
		if (seenDestinationPaths.has(destinationKey)) continue;
		seenDestinationPaths.add(destinationKey);
		if (shouldLinkToAgents && targetPathToCheck !== AGENT_STANDARD_PATH) {
			if (await getExistingPathKind(destinationPath) !== "file") continue;
		}
		if (fs.existsSync(destinationPath)) {
			if (fs.lstatSync(destinationPath).isSymbolicLink()) continue;
			const destinationRealPath = await fsPromises.realpath(destinationPath);
			if (seenRealPaths.has(destinationRealPath)) continue;
			if (replaceMarkedAgentInstructionsSection(await fsPromises.readFile(destinationPath, "utf-8"), incomingContent) !== void 0) {
				seenRealPaths.add(destinationRealPath);
				continue;
			}
			conflicts.push({ targetPath: targetPathToCheck });
			seenRealPaths.add(destinationRealPath);
		}
	}
	return conflicts;
}
async function writeAgentInstructions({ projectRoot, targetPath, targetPaths, interactive, conflictDecisions, silent = false }) {
	const paths = [...targetPaths ?? [], ...targetPath ? [targetPath] : []];
	if (paths.length === 0) return;
	const sourcePath = path.join(pkgRoot, "AGENTS.md");
	if (!fs.existsSync(sourcePath)) {
		if (!silent) log.warn("Agent instructions template not found; skipping.");
		return;
	}
	const seenDestinationPaths = /* @__PURE__ */ new Set();
	const seenRealPaths = /* @__PURE__ */ new Set();
	const incomingContent = await fsPromises.readFile(sourcePath, "utf-8");
	const shouldLinkToAgents = paths.includes(AGENT_STANDARD_PATH);
	const orderedPaths = shouldLinkToAgents ? [AGENT_STANDARD_PATH, ...paths.filter((p) => p !== AGENT_STANDARD_PATH)] : paths;
	for (const targetPathToWrite of orderedPaths) {
		const destinationPath = path.join(projectRoot, targetPathToWrite);
		const destinationKey = path.resolve(destinationPath);
		if (seenDestinationPaths.has(destinationKey)) continue;
		seenDestinationPaths.add(destinationKey);
		await fsPromises.mkdir(path.dirname(destinationPath), { recursive: true });
		if (shouldLinkToAgents && targetPathToWrite !== AGENT_STANDARD_PATH) {
			if (await tryLinkTargetToAgents(projectRoot, targetPathToWrite, silent)) continue;
		}
		if (fs.existsSync(destinationPath)) {
			if (fs.lstatSync(destinationPath).isSymbolicLink()) {
				if (!silent) log.info(`Skipped writing ${targetPathToWrite} (symlink)`);
				continue;
			}
			const destinationRealPath = await fsPromises.realpath(destinationPath);
			if (seenRealPaths.has(destinationRealPath)) {
				if (!silent) log.info(`Skipped writing ${targetPathToWrite} (duplicate target)`);
				continue;
			}
			const existingContent = await fsPromises.readFile(destinationPath, "utf-8");
			const updatedContent = replaceMarkedAgentInstructionsSection(existingContent, incomingContent);
			if (updatedContent !== void 0) {
				if (updatedContent !== existingContent) await fsPromises.writeFile(destinationPath, updatedContent);
				seenRealPaths.add(destinationRealPath);
				continue;
			}
			let conflictAction;
			const preResolved = conflictDecisions?.get(targetPathToWrite);
			if (preResolved) conflictAction = preResolved;
			else if (interactive) {
				const action = await select({
					message: `Agent instructions already exist at ${targetPathToWrite}.`,
					options: [{
						label: "Append",
						value: "append",
						hint: "Add template content to the end"
					}, {
						label: "Skip",
						value: "skip",
						hint: "Leave existing file unchanged"
					}],
					initialValue: "skip"
				});
				conflictAction = Ct(action) || action === "skip" ? "skip" : "append";
			} else conflictAction = "skip";
			if (conflictAction === "append") await appendAgentContent(destinationPath, targetPathToWrite, existingContent, incomingContent, silent);
			else {
				const suffix = !preResolved && !interactive ? " (already exists)" : "";
				if (!silent) log.info(`Skipped writing ${targetPathToWrite}${suffix}`);
			}
			seenRealPaths.add(destinationRealPath);
			continue;
		}
		await fsPromises.writeFile(destinationPath, incomingContent);
		if (!silent) log.success(`Wrote agent instructions to ${targetPathToWrite}`);
		seenRealPaths.add(await fsPromises.realpath(destinationPath));
	}
}
async function appendAgentContent(destinationPath, targetPath, existingContent, incomingContent, silent = false) {
	const separator = existingContent.endsWith("\n") ? "" : "\n";
	await fsPromises.appendFile(destinationPath, `${separator}\n${incomingContent}`);
	if (!silent) log.success(`Appended agent instructions to ${targetPath}`);
}
function normalizeAgentName(value) {
	return value.trim().toLowerCase().replace(/[^a-z0-9]+/g, "");
}
function replaceMarkedAgentInstructionsSection(existing, incoming) {
	const existingRange = getMarkedRange(existing, AGENT_INSTRUCTIONS_START_MARKER, AGENT_INSTRUCTIONS_END_MARKER);
	if (!existingRange) return;
	const incomingRange = getMarkedRange(incoming, AGENT_INSTRUCTIONS_START_MARKER, AGENT_INSTRUCTIONS_END_MARKER);
	if (!incomingRange) return;
	return `${existing.slice(0, existingRange.start)}${incoming.slice(incomingRange.start, incomingRange.end)}${existing.slice(existingRange.end)}`;
}
async function tryLinkTargetToAgents(projectRoot, targetPath, silent = false) {
	const destinationPath = path.join(projectRoot, targetPath);
	const agentsPath = path.join(projectRoot, AGENT_STANDARD_PATH);
	const symlinkTarget = path.relative(path.dirname(destinationPath), agentsPath);
	const existing = await getExistingPathKind(destinationPath);
	if (existing === "file") return false;
	if (existing === "symlink") {
		const currentLink = await fsPromises.readlink(destinationPath);
		if (path.resolve(path.dirname(destinationPath), currentLink) === agentsPath) {
			if (!silent) log.info(`Skipped linking ${targetPath} (already linked to ${AGENT_STANDARD_PATH})`);
			return true;
		}
		await fsPromises.unlink(destinationPath);
	}
	try {
		await fsPromises.symlink(symlinkTarget, destinationPath);
	} catch (err) {
		if (err.code === "EPERM") {
			await fsPromises.copyFile(agentsPath, destinationPath);
			if (!silent) log.success(`Copied ${AGENT_STANDARD_PATH} to ${targetPath}`);
			return true;
		}
		throw err;
	}
	if (!silent) log.success(`Linked ${targetPath} to ${AGENT_STANDARD_PATH}`);
	return true;
}
async function getExistingPathKind(filePath) {
	if (!fs.existsSync(filePath)) return "missing";
	return (await fsPromises.lstat(filePath)).isSymbolicLink() ? "symlink" : "file";
}
function getMarkedRange(content, startMarker, endMarker) {
	const start = content.indexOf(startMarker);
	if (start === -1) return;
	const endMarkerIndex = content.indexOf(endMarker, start + startMarker.length);
	if (endMarkerIndex === -1) return;
	return {
		start,
		end: endMarkerIndex + endMarker.length
	};
}
//#endregion
export { note as A, DependencyType as C, intro as D, confirm as E, Ct as F, select as M, spinner as N, log as O, text as P, runCommandSilently as S, cancel as T, upgradeYarn as _, hasExistingAgentInstructions as a, rulesDir as b, writeAgentInstructions as c, downloadPackageManager$1 as d, getSpinner as f, selectPackageManager as g, runViteInstall as h, getAgentById as i, outro as j, multiselect as k, cancelAndExit as l, runViteFmt as m, detectAgents as n, replaceMarkedAgentInstructionsSection as o, promptGitHooks as p, detectExistingAgentTargetPaths as r, selectAgentTargetPaths as s, detectAgentConflicts as t, defaultInteractive as u, displayRelative as v, PackageManager as w, templatesDir as x, pkgRoot as y };
