// vite-plus@0.1.12/dist/global/config.js
import { A as note, F as Ct, M as select, O as log, T as cancel, a as hasExistingAgentInstructions, i as getAgentById, n as detectAgents, o as replaceMarkedAgentInstructionsSection, p as promptGitHooks, u as defaultInteractive, y as pkgRoot } from "./agent-l4VXJklB.js";
import { t as lib_default } from "./lib-DxappLRQ.js";
import { i as writeJsonFile, r as readJsonFile, s as VITE_PLUS_NAME } from "./json-BRdVJ52a.js";
import { i as log$1, t as renderCliDoc } from "./help-CbTzUdXc.js";
import { dirname, join, relative } from "node:path";
import { existsSync, lstatSync, mkdirSync, readFileSync, readdirSync, readlinkSync, realpathSync, rmSync, symlinkSync, writeFileSync } from "node:fs";
import { vitePlusHeader } from "../../binding/index.js";
import { spawnSync } from "node:child_process";
//#region src/utils/skills.ts
function parseSkills(skillsDir) {
	if (!existsSync(skillsDir)) return [];
	const entries = readdirSync(skillsDir, { withFileTypes: true });
	const skills = [];
	for (const entry of entries) {
		if (!entry.isDirectory()) continue;
		const skillMd = join(skillsDir, entry.name, "SKILL.md");
		if (!existsSync(skillMd)) continue;
		const frontmatter = readFileSync(skillMd, "utf-8").match(/^---\n([\s\S]*?)\n---/);
		if (!frontmatter) {
			log.warn(`  Skipping ${entry.name}: SKILL.md is missing valid frontmatter`);
			continue;
		}
		const nameMatch = frontmatter[1].match(/^name:\s*(.+)$/m);
		const descMatch = frontmatter[1].match(/^description:\s*(.+)$/m);
		skills.push({
			dirName: entry.name,
			name: nameMatch ? nameMatch[1].trim() : entry.name,
			description: descMatch ? descMatch[1].trim() : ""
		});
	}
	return skills;
}
/** Check if a path exists on disk, including broken symlinks that existsSync misses. */
function pathExists(p) {
	try {
		lstatSync(p);
		return true;
	} catch {
		return false;
	}
}
function linkSkills(root, skillsDir, skills, agentSkillsDir) {
	const targetDir = join(root, agentSkillsDir);
	if (!existsSync(targetDir)) mkdirSync(targetDir, { recursive: true });
	const isWindows = process.platform === "win32";
	const symlinkType = isWindows ? "junction" : "dir";
	let linked = 0;
	for (const skill of skills) {
		const linkPath = join(targetDir, skill.dirName);
		const sourcePath = join(skillsDir, skill.dirName);
		const relativeTarget = relative(targetDir, sourcePath);
		const symlinkTarget = isWindows ? sourcePath : relativeTarget;
		if (pathExists(linkPath)) {
			try {
				if (readlinkSync(linkPath) === symlinkTarget) {
					log.info(`  ${skill.name} — already linked`);
					continue;
				}
			} catch (err) {
				if (err.code !== "EINVAL") {
					log.warn(`  ${skill.name} — failed to read existing path: ${err.message}`);
					continue;
				}
			}
			log.warn(`  ${skill.name} — path exists but is not the expected symlink, skipping`);
			continue;
		}
		try {
			symlinkSync(symlinkTarget, linkPath, symlinkType);
		} catch (err) {
			log.warn(`  ${skill.name} — failed to create symlink: ${err.message}`);
			continue;
		}
		log.success(`  ${skill.name} — linked`);
		linked++;
	}
	return linked;
}
function getStableSkillsDir(root) {
	const resolvedSkillsDir = join(pkgRoot, "skills");
	const logicalSkillsDir = join(root, "node_modules", VITE_PLUS_NAME, "skills");
	try {
		if (realpathSync(logicalSkillsDir) === realpathSync(resolvedSkillsDir)) return logicalSkillsDir;
	} catch {}
	return resolvedSkillsDir;
}
function linkSkillsForSpecificAgents(root, agentConfigs) {
	const skillsDir = getStableSkillsDir(root);
	const skills = parseSkills(skillsDir);
	if (skills.length === 0) return 0;
	if (agentConfigs.length === 0) return 0;
	let totalLinked = 0;
	for (const agent of agentConfigs) {
		log.info(`${agent.displayName} → ${agent.skillsDir}`);
		totalLinked += linkSkills(root, skillsDir, skills, agent.skillsDir);
	}
	return totalLinked;
}
//#endregion
//#region src/config/agent.ts
function detectInstructionFilePath(root, agentConfigs) {
	if (agentConfigs.some((a) => a.skillsDir === ".claude/skills")) return "CLAUDE.md";
	if (existsSync(join(root, "CLAUDE.md"))) return "CLAUDE.md";
	return "AGENTS.md";
}
async function pickAgentWhenUndetected() {
	const choice = await select({
		message: "Could not detect your coding agent. Which one are you using?",
		options: [
			{
				value: "claude-code",
				label: "Claude Code"
			},
			{
				value: "cursor",
				label: "Cursor"
			},
			{
				value: "codex",
				label: "Codex"
			},
			{
				value: "gemini-cli",
				label: "Gemini CLI"
			},
			{
				value: "generic",
				label: "Generic"
			}
		]
	});
	if (Ct(choice)) {
		cancel("Setup cancelled.");
		process.exit(0);
	}
	if (choice === "generic") return {
		instructionFilePath: "AGENTS.md",
		agents: []
	};
	const selected = getAgentById(choice);
	if (!selected) return {
		instructionFilePath: "AGENTS.md",
		agents: []
	};
	return {
		instructionFilePath: choice === "claude-code" ? "CLAUDE.md" : "AGENTS.md",
		agents: [selected]
	};
}
async function resolveAgentSetup(root, interactive) {
	const detected = detectAgents(root);
	if (detected.length > 0 || !interactive) return {
		instructionFilePath: detectInstructionFilePath(root, detected),
		agents: detected
	};
	return pickAgentWhenUndetected();
}
function readAgentPrompt() {
	return readFileSync(join(pkgRoot, "AGENTS.md"), "utf-8");
}
function injectAgentBlock(root, filePath) {
	const fullPath = join(root, filePath);
	const template = readAgentPrompt();
	if (existsSync(fullPath)) {
		const existing = readFileSync(fullPath, "utf-8");
		const updated = replaceMarkedAgentInstructionsSection(existing, template);
		if (updated !== void 0) if (updated !== existing) {
			writeFileSync(fullPath, updated);
			log.success(`Updated Vite+ instructions in ${filePath}`);
		} else log.info(`${filePath} already has up-to-date Vite+ instructions`);
		else {
			writeFileSync(fullPath, existing + (existing.endsWith("\n") ? "\n" : "\n\n") + template);
			log.success(`Added Vite+ instructions to ${filePath}`);
		}
	} else {
		writeFileSync(fullPath, template);
		log.success(`Created ${filePath} with Vite+ instructions`);
	}
}
function writeMcpConfigForTarget(root, target) {
	const fullPath = join(root, target.filePath);
	let existing = {};
	if (existsSync(fullPath)) try {
		existing = readJsonFile(fullPath);
	} catch {
		log.warn(`Could not parse ${target.filePath} — skipping MCP config. Please add the config manually.`);
		return;
	}
	if (!existing[target.rootKey]) existing[target.rootKey] = {};
	if (existing[target.rootKey]["vite-plus"]) {
		log.info(`${target.filePath} already has vite-plus MCP config`);
		return;
	}
	existing[target.rootKey]["vite-plus"] = {
		command: "npx",
		args: ["vp", "mcp"],
		...target.extraFields
	};
	mkdirSync(dirname(fullPath), { recursive: true });
	writeJsonFile(fullPath, existing);
	log.success(`Added vite-plus MCP server to ${target.filePath}`);
}
function pickMcpTarget(root, targets) {
	if (targets.length === 1) return targets[0];
	return targets.find((t) => existsSync(join(root, t.filePath))) ?? targets[0];
}
function setupMcpConfig(root, selectedAgents) {
	if (selectedAgents.length === 0) {
		note(JSON.stringify({ "vite-plus": {
			command: "npx",
			args: ["vp", "mcp"]
		} }, null, 2), "Add this MCP server config to your agent");
		return;
	}
	const mcpAgents = [];
	const hintAgents = [];
	for (const agent of selectedAgents) if (agent.mcpConfig) mcpAgents.push({
		agent,
		targets: agent.mcpConfig
	});
	else if (agent.mcpHint) hintAgents.push({
		agent,
		hint: agent.mcpHint
	});
	for (const { agent, hint } of hintAgents) log.info(`${agent.displayName}: ${hint}`);
	for (const { agent, targets } of mcpAgents) {
		const target = pickMcpTarget(root, targets);
		log.info(`${agent.displayName} MCP target: ${target.filePath}`);
		writeMcpConfigForTarget(root, target);
	}
}
//#endregion
//#region src/config/hooks.ts
const HOOKS = [
	"pre-commit",
	"pre-merge-commit",
	"prepare-commit-msg",
	"commit-msg",
	"post-commit",
	"applypatch-msg",
	"pre-applypatch",
	"post-applypatch",
	"pre-rebase",
	"post-rewrite",
	"post-checkout",
	"post-merge",
	"pre-push",
	"pre-auto-gc"
];
function nestedDirname(depth) {
	let expr = "\"$0\"";
	for (let i = 0; i < depth; i++) expr = `"$(dirname ${expr})"`;
	return expr;
}
function hookScript(dir) {
	return `#!/usr/bin/env sh
{ [ "$HUSKY" = "2" ] || [ "$VITE_GIT_HOOKS" = "2" ]; } && set -x
n=$(basename "$0")
s=$(dirname "$(dirname "$0")")/$n

[ ! -f "$s" ] && exit 0

i="\${XDG_CONFIG_HOME:-$HOME/.config}/vite-plus/hooks-init.sh"
[ ! -f "$i" ] && i="\${XDG_CONFIG_HOME:-$HOME/.config}/husky/init.sh"
[ -f "$i" ] && . "$i"

{ [ "\${HUSKY-}" = "0" ] || [ "\${VITE_GIT_HOOKS-}" = "0" ]; } && exit 0

d=${nestedDirname(dir.split("/").filter((s) => s !== "" && s !== ".").length + 2)}
export PATH="$d/node_modules/.bin:$PATH"
sh -e "$s" "$@"
c=$?

[ $c != 0 ] && echo "VITE+ - $n script failed (code $c)"
[ $c = 127 ] && echo "VITE+ - command not found in PATH=$PATH"
exit $c`;
}
function install(dir = ".vite-hooks") {
	if (process.env.HUSKY === "0" || process.env.VITE_GIT_HOOKS === "0") return {
		message: "skip install (git hooks disabled)",
		isError: false
	};
	if (dir.includes("..")) return {
		message: ".. not allowed",
		isError: false
	};
	const prefixResult = spawnSync("git", ["rev-parse", "--show-prefix"]);
	if (prefixResult.status == null) return {
		message: "git command not found",
		isError: true
	};
	if (prefixResult.status !== 0) return {
		message: ".git can't be found",
		isError: false
	};
	const internal = (x = "") => join(dir, "_", x);
	const rel = prefixResult.stdout.toString().trim().replace(/\/$/, "");
	const target = rel ? `${rel}/${dir}/_` : `${dir}/_`;
	const checkResult = spawnSync("git", [
		"config",
		"--local",
		"core.hooksPath"
	]);
	const existingHooksPath = checkResult.status === 0 ? checkResult.stdout?.toString().trim() : "";
	if (existingHooksPath && existingHooksPath !== target && existingHooksPath !== ".husky" && !existingHooksPath.startsWith(".husky/")) return {
		message: `core.hooksPath is already set to "${existingHooksPath}", skipping`,
		isError: false
	};
	const { status, stderr } = spawnSync("git", [
		"config",
		"core.hooksPath",
		target
	]);
	if (status == null) return {
		message: "git command not found",
		isError: true
	};
	if (status) return {
		message: "" + stderr,
		isError: true
	};
	rmSync(internal("husky.sh"), { force: true });
	mkdirSync(internal(), { recursive: true });
	writeFileSync(internal(".gitignore"), "*");
	writeFileSync(internal("h"), hookScript(dir), { mode: 493 });
	for (const hook of HOOKS) writeFileSync(internal(hook), `#!/usr/bin/env sh\n. "$(dirname "$0")/h"`, { mode: 493 });
	return {
		message: "",
		isError: false
	};
}
//#endregion
//#region src/config/bin.ts
async function main() {
	const args = lib_default(process.argv.slice(3), {
		boolean: ["help", "hooks-only"],
		string: ["hooks-dir"],
		alias: { h: "help" }
	});
	if (args.help) {
		const helpMessage = renderCliDoc({
			usage: "vp config [OPTIONS]",
			summary: "Configure Vite+ for the current project (hooks + agent integration).",
			documentationUrl: "https://viteplus.dev/guide/commit-hooks",
			sections: [{
				title: "Options",
				rows: [{
					label: "--hooks-dir <path>",
					description: "Custom hooks directory (default: .vite-hooks)"
				}, {
					label: "-h, --help",
					description: "Show this help message"
				}]
			}, {
				title: "Environment",
				rows: [{
					label: "VITE_GIT_HOOKS=0",
					description: "Skip hook installation"
				}]
			}]
		});
		log$1(vitePlusHeader() + "\n");
		log$1(helpMessage);
		return;
	}
	const dir = args["hooks-dir"];
	const hooksOnly = args["hooks-only"];
	const interactive = defaultInteractive();
	const root = process.cwd();
	const isFirstHooksRun = !existsSync(join(root, dir ?? ".vite-hooks", "_", "pre-commit"));
	let shouldSetupHooks = true;
	if (interactive && isFirstHooksRun && !dir) shouldSetupHooks = await promptGitHooks({ interactive });
	if (shouldSetupHooks) {
		const { message, isError } = install(dir);
		if (message) {
			log$1(message);
			if (isError) process.exit(1);
		}
	}
	if (!hooksOnly && process.env.npm_lifecycle_event !== "prepare") {
		const isFirstAgentRun = !hasExistingAgentInstructions(root);
		const agentSetup = await resolveAgentSetup(root, interactive && isFirstAgentRun);
		injectAgentBlock(root, agentSetup.instructionFilePath);
		setupMcpConfig(root, agentSetup.agents);
		if (agentSetup.agents.length > 0) linkSkillsForSpecificAgents(root, agentSetup.agents);
	}
}
main();
//#endregion
