// vite-plus@0.1.12/dist/global/create.js
import { i as __toESM, t as __commonJSMin } from "./chunk-CgnkrU7a.js";
import { C as DependencyType, D as intro, E as confirm, F as Ct, M as select, N as spinner, O as log, P as text, T as cancel, c as writeAgentInstructions, d as downloadPackageManager$1, g as selectPackageManager, h as runViteInstall, k as multiselect, m as runViteFmt, p as promptGitHooks, r as detectExistingAgentTargetPaths, s as selectAgentTargetPaths, u as defaultInteractive, v as displayRelative, w as PackageManager, x as templatesDir } from "./agent-l4VXJklB.js";
import { t as lib_default } from "./lib-DxappLRQ.js";
import { _ as rewriteMonorepo, a as detectExistingEditor, f as installGitHooks, n as updatePackageJsonWithDeps, o as selectEditor, r as updateWorkspaceConfig, s as writeEditorConfigs, t as detectWorkspace$1, v as rewriteMonorepoProject, y as rewriteStandaloneProject } from "./workspace-CshY0VeI.js";
import "./browser-CBapUTD0.js";
import { r as readJsonFile, t as editJsonFile } from "./json-BRdVJ52a.js";
import "./package-YAMvX5PJ.js";
import { a as muted, i as log$1, n as accent, o as success, t as renderCliDoc } from "./help-CbTzUdXc.js";
import path from "node:path";
import { styleText } from "node:util";
import color from "picocolors";
import fs from "node:fs";
import { runCommand, vitePlusHeader } from "../../binding/index.js";
import spawn from "cross-spawn";
import fsPromises from "node:fs/promises";
import assert from "node:assert";
//#region src/create/command.ts
async function runCommandAndDetectProjectDir(options, parentDir) {
	const cwd = parentDir ? path.join(options.cwd, parentDir) : options.cwd;
	const existingDirs = /* @__PURE__ */ new Set();
	if (parentDir) {
		await fsPromises.mkdir(cwd, { recursive: true });
		const entries = await fsPromises.readdir(cwd, { withFileTypes: true });
		for (const entry of entries) if (entry.isDirectory()) existingDirs.add(entry.name);
	}
	const result = await runCommand({
		binName: options.command,
		args: options.args,
		envs: options.envs,
		cwd
	});
	let projectDir;
	let minDepth = Infinity;
	for (const [filePath, pathAccess] of Object.entries(result.pathAccesses)) if (pathAccess.write && filePath.endsWith("package.json") && !filePath.includes("node_modules")) {
		const dir = path.dirname(filePath);
		if (dir === "." || dir === "") continue;
		if (existingDirs.has(dir)) continue;
		const depth = dir.split(path.sep).length;
		if (depth < minDepth) {
			minDepth = depth;
			projectDir = dir;
		}
	}
	if (parentDir && projectDir) projectDir = path.join(parentDir, projectDir);
	return {
		exitCode: result.exitCode,
		projectDir
	};
}
async function runCommandSilently(options) {
	const child = spawn(options.command, options.args, {
		stdio: "pipe",
		cwd: options.cwd,
		env: options.envs
	});
	return await new Promise((resolve, reject) => {
		const stdout = [];
		const stderr = [];
		child.stdout?.on("data", (data) => {
			stdout.push(data);
		});
		child.stderr?.on("data", (data) => {
			stderr.push(data);
		});
		child.on("close", (code) => {
			resolve({
				exitCode: code ?? 0,
				stdout: Buffer.concat(stdout),
				stderr: Buffer.concat(stderr)
			});
		});
		child.on("error", (err) => {
			reject(err);
		});
	});
}
async function runCommand$1(options) {
	const child = spawn(options.command, options.args, {
		stdio: "inherit",
		cwd: options.cwd,
		env: options.envs
	});
	return await new Promise((resolve, reject) => {
		child.on("close", (code) => {
			resolve({ exitCode: code ?? 0 });
		});
		child.on("error", (err) => {
			reject(err);
		});
	});
}
function getPackageRunner(workspaceInfo) {
	switch (workspaceInfo.packageManager) {
		case "pnpm": return {
			command: "pnpm",
			args: ["dlx"]
		};
		case "yarn": return {
			command: "yarn",
			args: ["dlx"]
		};
		default: return {
			command: "npx",
			args: []
		};
	}
}
function formatDlxCommand(packageName, args, workspaceInfo) {
	const runner = getPackageRunner(workspaceInfo);
	const dlxArgs = runner.command === "npm" ? ["--", ...args] : args;
	return {
		command: runner.command,
		args: [
			...runner.args,
			packageName,
			...dlxArgs
		]
	};
}
function prependToPathToEnvs(extraPath, envs) {
	const delimiter = path.delimiter;
	const pathKey = Object.keys(envs).find((key) => key.toLowerCase() === "path") ?? "PATH";
	const current = envs[pathKey] ?? "";
	if (!current.split(delimiter).filter(Boolean).includes(extraPath)) envs[pathKey] = extraPath + (current ? delimiter + current : "");
	return envs;
}
//#endregion
//#region src/create/templates/types.ts
const LibraryTemplateRepo = "github:sxzz/tsdown-templates/vite-plus";
const BuiltinTemplate = {
	generator: "vite:generator",
	monorepo: "vite:monorepo",
	application: "vite:application",
	library: "vite:library"
};
const TemplateType = {
	builtin: "builtin",
	bingo: "bingo",
	remote: "remote"
};
//#endregion
//#region src/create/discovery.ts
function isGitHubUrl(templateName) {
	return templateName.startsWith("https://github.com/") || templateName.startsWith("github:") || templateName.includes("github.com/");
}
function parseGitHubUrl(url) {
	if (url.startsWith("github:")) return url.slice(7);
	const match = url.match(/github\.com\/([^/]+\/[^/]+)/);
	if (match) return match[1].replace(/\.git$/, "");
	return null;
}
function inferGitHubRepoName(templateName) {
	const degitPath = parseGitHubUrl(templateName);
	if (!degitPath) return null;
	return degitPath.split("/").pop() || null;
}
function discoverTemplate(templateName, templateArgs, workspaceInfo, interactive) {
	const envs = prependToPathToEnvs(workspaceInfo.downloadPackageManager.binPrefix, { ...process.env });
	const parentDir = inferParentDir(templateName, workspaceInfo);
	if (templateName.startsWith("vite:")) return {
		command: templateName,
		args: [...templateArgs],
		envs,
		type: TemplateType.builtin,
		parentDir,
		interactive
	};
	if (isGitHubUrl(templateName)) {
		const degitPath = parseGitHubUrl(templateName);
		if (degitPath) return {
			command: "degit",
			args: [degitPath, ...templateArgs],
			envs,
			type: TemplateType.remote,
			parentDir,
			interactive
		};
	}
	const localPackage = workspaceInfo.packages.find((pkg) => pkg.name === templateName);
	if (localPackage) {
		const localPackagePath = path.join(workspaceInfo.rootDir, localPackage.path);
		const pkg = readJsonFile(path.join(localPackagePath, "package.json"));
		let binPath = "";
		if (pkg.bin) if (typeof pkg.bin === "string") binPath = path.join(localPackagePath, pkg.bin);
		else {
			const binName = Object.keys(pkg.bin)[0];
			binPath = path.join(localPackagePath, pkg.bin[binName]);
		}
		const args = [binPath, ...templateArgs];
		let type = TemplateType.remote;
		if (pkg.keywords?.includes("bingo-template") || !!pkg.dependencies?.bingo) {
			type = TemplateType.bingo;
			args.push("--skip-requests");
		}
		if (binPath) return {
			command: "node",
			args,
			envs,
			type,
			parentDir,
			interactive
		};
	}
	return {
		command: expandCreateShorthand(templateName),
		args: [...templateArgs],
		envs,
		type: TemplateType.remote,
		parentDir,
		interactive
	};
}
/**
* Expand shorthand template names to their full `create-*` package names.
*
* This follows the same convention as `npm create` / `pnpm create`:
* - `vite` → `create-vite`
* - `vite@latest` → `create-vite@latest`
* - `@tanstack/start` → `@tanstack/create-start`
* - `@tanstack/start@latest` → `@tanstack/create-start@latest`
*
* Special cases for packages where the convention doesn't work:
* - `nitro` → `create-nitro-app` (create-nitro is abandoned)
*
* Skips expansion for:
* - Builtin templates (`vite:*`)
* - GitHub URLs
* - Local paths (`./`, `../`, `/`)
* - Names already starting with `create-` (or `@scope/create-`)
*/
function expandCreateShorthand(templateName) {
	if (templateName.includes(":")) return templateName;
	if (isGitHubUrl(templateName)) return templateName;
	if (templateName.startsWith("./") || templateName.startsWith("../") || templateName.startsWith("/")) return templateName;
	if (templateName.startsWith("@")) {
		const slashIndex = templateName.indexOf("/");
		if (slashIndex === -1) return templateName;
		const scope = templateName.slice(0, slashIndex);
		const rest = templateName.slice(slashIndex + 1);
		const atIndex = rest.indexOf("@");
		const name = atIndex === -1 ? rest : rest.slice(0, atIndex);
		const version = atIndex === -1 ? "" : rest.slice(atIndex);
		if (name.startsWith("create-")) return templateName;
		return `${scope}/create-${name}${version}`;
	}
	const atIndex = templateName.indexOf("@");
	const name = atIndex === -1 ? templateName : templateName.slice(0, atIndex);
	const version = atIndex === -1 ? "" : templateName.slice(atIndex);
	if (name.startsWith("create-")) return templateName;
	if (name === "nitro") return `create-nitro-app${version}`;
	return `create-${name}${version}`;
}
function inferParentDir(templateName, workspaceInfo) {
	if (workspaceInfo.parentDirs.length === 0) return;
	let rule = /app/i;
	if (templateName === BuiltinTemplate.library) rule = /lib|component|package/i;
	else if (templateName === BuiltinTemplate.generator) rule = /generator|tool/i;
	for (const parentDir of workspaceInfo.parentDirs) if (rule.test(parentDir)) return parentDir;
}
//#endregion
//#region src/create/initial-template-options.ts
function getInitialTemplateOptions(isMonorepo) {
	return [
		...!isMonorepo ? [{
			label: "Vite+ Monorepo",
			value: BuiltinTemplate.monorepo,
			hint: "Create a new Vite+ monorepo project"
		}] : [],
		{
			label: "Vite+ Application",
			value: BuiltinTemplate.application,
			hint: "Create vite applications"
		},
		{
			label: "Vite+ Library",
			value: BuiltinTemplate.library,
			hint: "Create vite libraries"
		}
	];
}
//#endregion
//#region ../../node_modules/.pnpm/validate-npm-package-name@7.0.2/node_modules/validate-npm-package-name/lib/builtin-modules.json
var require_builtin_modules = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = [
		"_http_agent",
		"_http_client",
		"_http_common",
		"_http_incoming",
		"_http_outgoing",
		"_http_server",
		"_stream_duplex",
		"_stream_passthrough",
		"_stream_readable",
		"_stream_transform",
		"_stream_wrap",
		"_stream_writable",
		"_tls_common",
		"_tls_wrap",
		"assert",
		"assert/strict",
		"async_hooks",
		"buffer",
		"child_process",
		"cluster",
		"console",
		"constants",
		"crypto",
		"dgram",
		"diagnostics_channel",
		"dns",
		"dns/promises",
		"domain",
		"events",
		"fs",
		"fs/promises",
		"http",
		"http2",
		"https",
		"inspector",
		"inspector/promises",
		"module",
		"net",
		"os",
		"path",
		"path/posix",
		"path/win32",
		"perf_hooks",
		"process",
		"punycode",
		"querystring",
		"readline",
		"readline/promises",
		"repl",
		"stream",
		"stream/consumers",
		"stream/promises",
		"stream/web",
		"string_decoder",
		"sys",
		"timers",
		"timers/promises",
		"tls",
		"trace_events",
		"tty",
		"url",
		"util",
		"util/types",
		"v8",
		"vm",
		"wasi",
		"worker_threads",
		"zlib",
		"node:sea",
		"node:sqlite",
		"node:test",
		"node:test/reporters"
	];
}));
//#endregion
//#region ../../node_modules/.pnpm/@nkzw+safe-word-list@3.1.0/node_modules/@nkzw/safe-word-list/index.js
var import_lib = /* @__PURE__ */ __toESM((/* @__PURE__ */ __commonJSMin(((exports, module) => {
	const builtins = require_builtin_modules();
	var scopedPackagePattern = /* @__PURE__ */ new RegExp("^(?:@([^/]+?)[/])?([^/]+?)$");
	var exclusionList = ["node_modules", "favicon.ico"];
	function validate(name) {
		var warnings = [];
		var errors = [];
		if (name === null) {
			errors.push("name cannot be null");
			return done(warnings, errors);
		}
		if (name === void 0) {
			errors.push("name cannot be undefined");
			return done(warnings, errors);
		}
		if (typeof name !== "string") {
			errors.push("name must be a string");
			return done(warnings, errors);
		}
		if (!name.length) errors.push("name length must be greater than zero");
		if (name.startsWith(".")) errors.push("name cannot start with a period");
		if (name.startsWith("-")) errors.push("name cannot start with a hyphen");
		if (name.match(/^_/)) errors.push("name cannot start with an underscore");
		if (name.trim() !== name) errors.push("name cannot contain leading or trailing spaces");
		exclusionList.forEach(function(excludedName) {
			if (name.toLowerCase() === excludedName) errors.push(excludedName + " is not a valid package name");
		});
		if (builtins.includes(name.toLowerCase())) warnings.push(name + " is a core module name");
		if (name.length > 214) warnings.push("name can no longer contain more than 214 characters");
		if (name.toLowerCase() !== name) warnings.push("name can no longer contain capital letters");
		if (/[~'!()*]/.test(name.split("/").slice(-1)[0])) warnings.push("name can no longer contain special characters (\"~'!()*\")");
		if (encodeURIComponent(name) !== name) {
			var nameMatch = name.match(scopedPackagePattern);
			if (nameMatch) {
				var user = nameMatch[1];
				var pkg = nameMatch[2];
				if (pkg.startsWith(".")) errors.push("name cannot start with a period");
				if (encodeURIComponent(user) === user && encodeURIComponent(pkg) === pkg) return done(warnings, errors);
			}
			errors.push("name can only contain URL-friendly characters");
		}
		return done(warnings, errors);
	}
	var done = function(warnings, errors) {
		var result = {
			validForNewPackages: errors.length === 0 && warnings.length === 0,
			validForOldPackages: errors.length === 0,
			warnings,
			errors
		};
		if (!result.warnings.length) delete result.warnings;
		if (!result.errors.length) delete result.errors;
		return result;
	};
	module.exports = validate;
})))(), 1);
const words = [
	"ability",
	"able",
	"about",
	"above",
	"abroad",
	"absence",
	"absolute",
	"absolutely",
	"absorb",
	"academic",
	"accept",
	"access",
	"accompany",
	"accomplish",
	"according",
	"account",
	"accurate",
	"achieve",
	"achievement",
	"acid",
	"acknowledge",
	"acquire",
	"across",
	"act",
	"action",
	"active",
	"activist",
	"activity",
	"actor",
	"actress",
	"actual",
	"actually",
	"adapt",
	"add",
	"addition",
	"additional",
	"address",
	"adequate",
	"adjust",
	"adjustment",
	"administration",
	"administrator",
	"admire",
	"admission",
	"admit",
	"adolescent",
	"adopt",
	"adult",
	"advance",
	"advanced",
	"advantage",
	"adventure",
	"advertising",
	"advice",
	"advise",
	"adviser",
	"advocate",
	"affair",
	"affect",
	"afford",
	"after",
	"afternoon",
	"again",
	"against",
	"age",
	"agency",
	"agenda",
	"agent",
	"ago",
	"agree",
	"agreement",
	"agricultural",
	"ahead",
	"aid",
	"aide",
	"aim",
	"air",
	"aircraft",
	"airline",
	"airport",
	"album",
	"alive",
	"all",
	"alliance",
	"allow",
	"ally",
	"almost",
	"alone",
	"along",
	"already",
	"also",
	"alter",
	"alternative",
	"although",
	"always",
	"amazing",
	"among",
	"amount",
	"analysis",
	"analyst",
	"analyze",
	"ancient",
	"and",
	"angle",
	"animal",
	"anniversary",
	"announce",
	"annual",
	"another",
	"answer",
	"anticipate",
	"any",
	"anybody",
	"anymore",
	"anyone",
	"anything",
	"anyway",
	"anywhere",
	"apart",
	"apartment",
	"apparent",
	"apparently",
	"appeal",
	"appear",
	"appearance",
	"apple",
	"application",
	"apply",
	"appoint",
	"appointment",
	"appreciate",
	"approach",
	"appropriate",
	"approval",
	"approve",
	"approximately",
	"architect",
	"area",
	"argue",
	"argument",
	"arise",
	"arm",
	"armed",
	"around",
	"arrange",
	"arrangement",
	"arrival",
	"arrive",
	"art",
	"article",
	"artist",
	"artistic",
	"aside",
	"ask",
	"asleep",
	"aspect",
	"assert",
	"assess",
	"assessment",
	"asset",
	"assign",
	"assignment",
	"assist",
	"assistance",
	"assistant",
	"associate",
	"association",
	"assume",
	"assumption",
	"assure",
	"athlete",
	"athletic",
	"atmosphere",
	"attach",
	"attempt",
	"attend",
	"attention",
	"attitude",
	"attorney",
	"attract",
	"attractive",
	"attribute",
	"audience",
	"author",
	"authority",
	"auto",
	"available",
	"average",
	"avoid",
	"award",
	"aware",
	"awareness",
	"away",
	"awesome",
	"baby",
	"back",
	"background",
	"bag",
	"bake",
	"balance",
	"ball",
	"ban",
	"band",
	"bank",
	"bar",
	"barely",
	"barrel",
	"base",
	"baseball",
	"basic",
	"basically",
	"basis",
	"basket",
	"basketball",
	"bathroom",
	"battery",
	"beach",
	"bean",
	"bear",
	"beat",
	"beautiful",
	"beauty",
	"because",
	"become",
	"bed",
	"bedroom",
	"beer",
	"before",
	"begin",
	"beginning",
	"behavior",
	"behind",
	"being",
	"belief",
	"believe",
	"bell",
	"belong",
	"below",
	"belt",
	"bench",
	"bend",
	"beneath",
	"benefit",
	"beside",
	"besides",
	"best",
	"bet",
	"better",
	"between",
	"beyond",
	"big",
	"bike",
	"bill",
	"billion",
	"bind",
	"biological",
	"bird",
	"birth",
	"birthday",
	"bit",
	"bite",
	"black",
	"blade",
	"blanket",
	"blind",
	"block",
	"blow",
	"blue",
	"board",
	"boat",
	"body",
	"bond",
	"bone",
	"book",
	"boom",
	"boot",
	"border",
	"born",
	"borrow",
	"boss",
	"both",
	"bother",
	"bottle",
	"bottom",
	"boundary",
	"bowl",
	"box",
	"boy",
	"boyfriend",
	"brain",
	"branch",
	"brand",
	"bread",
	"break",
	"breakfast",
	"breast",
	"breath",
	"breathe",
	"brick",
	"bridge",
	"brief",
	"briefly",
	"bright",
	"brilliant",
	"bring",
	"broad",
	"brother",
	"brown",
	"brush",
	"buck",
	"budget",
	"build",
	"building",
	"bullet",
	"bunch",
	"bury",
	"bus",
	"business",
	"busy",
	"but",
	"butter",
	"button",
	"buy",
	"buyer",
	"cabin",
	"cabinet",
	"cable",
	"cake",
	"calculate",
	"call",
	"camera",
	"camp",
	"campaign",
	"campus",
	"can",
	"candidate",
	"cap",
	"capability",
	"capable",
	"capacity",
	"capital",
	"captain",
	"capture",
	"car",
	"carbon",
	"card",
	"care",
	"career",
	"careful",
	"carefully",
	"carrier",
	"carry",
	"case",
	"cash",
	"cast",
	"cat",
	"catch",
	"category",
	"cause",
	"ceiling",
	"celebrate",
	"celebration",
	"celebrity",
	"cell",
	"center",
	"central",
	"century",
	"ceremony",
	"certain",
	"certainly",
	"chain",
	"chair",
	"chairman",
	"challenge",
	"chamber",
	"champion",
	"championship",
	"chance",
	"change",
	"changing",
	"channel",
	"chapter",
	"character",
	"characteristic",
	"characterize",
	"charge",
	"charity",
	"chart",
	"chase",
	"cheap",
	"check",
	"cheek",
	"cheese",
	"chef",
	"chemical",
	"chest",
	"chicken",
	"chief",
	"child",
	"childhood",
	"chip",
	"chocolate",
	"choice",
	"cholesterol",
	"choose",
	"church",
	"cigarette",
	"circle",
	"circumstance",
	"cite",
	"citizen",
	"city",
	"civil",
	"civilian",
	"claim",
	"class",
	"classic",
	"classroom",
	"clean",
	"clear",
	"clearly",
	"client",
	"climate",
	"climb",
	"clinic",
	"clinical",
	"clock",
	"close",
	"closely",
	"closer",
	"clothes",
	"clothing",
	"cloud",
	"club",
	"clue",
	"cluster",
	"coach",
	"coal",
	"coalition",
	"coast",
	"coat",
	"code",
	"coffee",
	"cognitive",
	"cold",
	"colleague",
	"collect",
	"collection",
	"collective",
	"college",
	"colonial",
	"color",
	"column",
	"combination",
	"combine",
	"come",
	"comedy",
	"comfort",
	"comfortable",
	"command",
	"commander",
	"comment",
	"commercial",
	"commission",
	"commit",
	"commitment",
	"committee",
	"common",
	"communicate",
	"communication",
	"community",
	"company",
	"compare",
	"comparison",
	"compete",
	"competition",
	"competitive",
	"competitor",
	"complete",
	"completely",
	"complex",
	"complicated",
	"component",
	"compose",
	"composition",
	"comprehensive",
	"computer",
	"concentrate",
	"concentration",
	"concept",
	"concern",
	"concerned",
	"concert",
	"conclude",
	"conclusion",
	"concrete",
	"condition",
	"conduct",
	"conference",
	"confidence",
	"confident",
	"confirm",
	"confront",
	"confusion",
	"congressional",
	"connect",
	"connection",
	"consciousness",
	"consensus",
	"consequence",
	"conservative",
	"consider",
	"considerable",
	"consideration",
	"consist",
	"consistent",
	"constant",
	"constantly",
	"constitute",
	"constitutional",
	"construct",
	"construction",
	"consultant",
	"consume",
	"consumer",
	"consumption",
	"contact",
	"contain",
	"container",
	"contemporary",
	"content",
	"contest",
	"context",
	"continue",
	"continued",
	"contract",
	"contrast",
	"contribute",
	"contribution",
	"control",
	"convention",
	"conventional",
	"conversation",
	"convert",
	"conviction",
	"convince",
	"cook",
	"cookie",
	"cooking",
	"cool",
	"cooperation",
	"cop",
	"cope",
	"copy",
	"core",
	"corn",
	"corner",
	"corporate",
	"corporation",
	"correct",
	"correspondent",
	"cost",
	"cotton",
	"couch",
	"could",
	"council",
	"counselor",
	"count",
	"counter",
	"country",
	"county",
	"couple",
	"courage",
	"course",
	"court",
	"cousin",
	"cover",
	"coverage",
	"cow",
	"crack",
	"craft",
	"cream",
	"create",
	"creation",
	"creative",
	"creature",
	"credit",
	"crew",
	"criteria",
	"crop",
	"cross",
	"crowd",
	"crucial",
	"cultural",
	"culture",
	"cup",
	"curious",
	"current",
	"currently",
	"curriculum",
	"custom",
	"customer",
	"cut",
	"cycle",
	"dad",
	"daily",
	"dance",
	"dare",
	"dark",
	"darkness",
	"data",
	"date",
	"daughter",
	"day",
	"deal",
	"dealer",
	"dear",
	"debate",
	"decade",
	"decide",
	"decision",
	"deck",
	"declare",
	"decrease",
	"deep",
	"deeply",
	"deer",
	"defend",
	"defendant",
	"defense",
	"defensive",
	"define",
	"definitely",
	"definition",
	"degree",
	"delay",
	"deliver",
	"delivery",
	"demand",
	"democracy",
	"democratic",
	"demonstrate",
	"demonstration",
	"deny",
	"department",
	"depend",
	"dependent",
	"depending",
	"depict",
	"depth",
	"deputy",
	"derive",
	"describe",
	"description",
	"desert",
	"deserve",
	"design",
	"designer",
	"desire",
	"desk",
	"desperate",
	"despite",
	"detail",
	"detailed",
	"detect",
	"determine",
	"develop",
	"developing",
	"development",
	"device",
	"devote",
	"dialogue",
	"diet",
	"differ",
	"difference",
	"different",
	"differently",
	"difficult",
	"difficulty",
	"dig",
	"digital",
	"dimension",
	"dining",
	"dinner",
	"direct",
	"direction",
	"directly",
	"director",
	"disability",
	"disagree",
	"disappear",
	"discipline",
	"discourse",
	"discover",
	"discovery",
	"discuss",
	"discussion",
	"dish",
	"dismiss",
	"display",
	"distance",
	"distant",
	"distinct",
	"distinction",
	"distinguish",
	"distribute",
	"distribution",
	"district",
	"diverse",
	"diversity",
	"divide",
	"division",
	"divorce",
	"doctor",
	"document",
	"dog",
	"domestic",
	"dominant",
	"dominate",
	"door",
	"double",
	"down",
	"downtown",
	"dozen",
	"draft",
	"drag",
	"drama",
	"dramatic",
	"dramatically",
	"draw",
	"drawing",
	"dream",
	"dress",
	"drink",
	"drive",
	"driver",
	"drop",
	"dry",
	"due",
	"during",
	"dust",
	"duty",
	"each",
	"eager",
	"ear",
	"early",
	"earn",
	"earnings",
	"earth",
	"ease",
	"easily",
	"east",
	"eastern",
	"easy",
	"eat",
	"economic",
	"economics",
	"economist",
	"economy",
	"edge",
	"edition",
	"editor",
	"educate",
	"education",
	"educational",
	"educator",
	"effect",
	"effective",
	"effectively",
	"efficiency",
	"efficient",
	"effort",
	"egg",
	"eight",
	"either",
	"elderly",
	"elect",
	"election",
	"electric",
	"electricity",
	"electronic",
	"element",
	"elementary",
	"eliminate",
	"elite",
	"else",
	"elsewhere",
	"embrace",
	"emerge",
	"emission",
	"emotion",
	"emotional",
	"emphasis",
	"emphasize",
	"employ",
	"employee",
	"employer",
	"employment",
	"empty",
	"enable",
	"encounter",
	"encourage",
	"end",
	"energy",
	"enforcement",
	"engage",
	"engine",
	"engineer",
	"engineering",
	"enhance",
	"enjoy",
	"enormous",
	"enough",
	"ensure",
	"enter",
	"enterprise",
	"entertainment",
	"entire",
	"entirely",
	"entrance",
	"entry",
	"environment",
	"environmental",
	"episode",
	"equal",
	"equally",
	"equipment",
	"era",
	"escape",
	"especially",
	"essay",
	"essential",
	"essentially",
	"establish",
	"establishment",
	"estate",
	"estimate",
	"etc",
	"ethics",
	"ethnic",
	"evaluate",
	"evaluation",
	"even",
	"evening",
	"event",
	"eventually",
	"ever",
	"every",
	"everybody",
	"everyday",
	"everyone",
	"everything",
	"everywhere",
	"evidence",
	"evolution",
	"evolve",
	"exact",
	"exactly",
	"examination",
	"examine",
	"example",
	"exceed",
	"excellent",
	"except",
	"exception",
	"exchange",
	"exciting",
	"executive",
	"exercise",
	"exhibit",
	"exhibition",
	"exist",
	"existence",
	"existing",
	"expand",
	"expansion",
	"expect",
	"expectation",
	"expense",
	"expensive",
	"experience",
	"experiment",
	"expert",
	"explain",
	"explanation",
	"explore",
	"expose",
	"express",
	"expression",
	"extend",
	"extension",
	"extensive",
	"extent",
	"external",
	"extra",
	"extraordinary",
	"extreme",
	"extremely",
	"eye",
	"fabric",
	"face",
	"facility",
	"fact",
	"factor",
	"factory",
	"faculty",
	"fade",
	"fair",
	"fairly",
	"faith",
	"fall",
	"false",
	"familiar",
	"family",
	"famous",
	"fan",
	"fantasy",
	"far",
	"farm",
	"farmer",
	"fashion",
	"fast",
	"fate",
	"father",
	"favor",
	"favorite",
	"feature",
	"federal",
	"fee",
	"feed",
	"feel",
	"feeling",
	"fellow",
	"female",
	"fence",
	"few",
	"fewer",
	"fiber",
	"fiction",
	"field",
	"fifteen",
	"fifth",
	"fifty",
	"figure",
	"file",
	"fill",
	"film",
	"final",
	"finally",
	"finance",
	"financial",
	"find",
	"finding",
	"fine",
	"finger",
	"finish",
	"firm",
	"first",
	"fish",
	"fishing",
	"fit",
	"fitness",
	"five",
	"fix",
	"flag",
	"flame",
	"flat",
	"flavor",
	"flesh",
	"flight",
	"float",
	"floor",
	"flow",
	"flower",
	"fly",
	"focus",
	"folk",
	"follow",
	"following",
	"food",
	"foot",
	"football",
	"for",
	"force",
	"foreign",
	"forest",
	"forever",
	"forget",
	"form",
	"formal",
	"formation",
	"former",
	"formula",
	"forth",
	"fortune",
	"forward",
	"found",
	"foundation",
	"founder",
	"four",
	"fourth",
	"frame",
	"framework",
	"free",
	"freedom",
	"freeze",
	"frequency",
	"frequent",
	"frequently",
	"fresh",
	"friend",
	"friendly",
	"friendship",
	"from",
	"front",
	"fruit",
	"fuel",
	"full",
	"fully",
	"fun",
	"function",
	"fund",
	"fundamental",
	"funding",
	"funeral",
	"funny",
	"furniture",
	"furthermore",
	"future",
	"gain",
	"galaxy",
	"gallery",
	"game",
	"gap",
	"garage",
	"garden",
	"garlic",
	"gas",
	"gate",
	"gather",
	"gaze",
	"gear",
	"gender",
	"gene",
	"general",
	"generally",
	"generate",
	"generation",
	"genetic",
	"gentleman",
	"gently",
	"gesture",
	"get",
	"ghost",
	"giant",
	"gift",
	"gifted",
	"girl",
	"girlfriend",
	"give",
	"given",
	"glad",
	"glance",
	"glass",
	"global",
	"glove",
	"goal",
	"gold",
	"golden",
	"golf",
	"good",
	"government",
	"governor",
	"grab",
	"grade",
	"gradually",
	"graduate",
	"grain",
	"grand",
	"grandfather",
	"grandmother",
	"grant",
	"grass",
	"grave",
	"gray",
	"great",
	"greatest",
	"green",
	"grocery",
	"ground",
	"group",
	"grow",
	"growing",
	"growth",
	"guarantee",
	"guard",
	"guess",
	"guest",
	"guide",
	"guideline",
	"guy",
	"habit",
	"habitat",
	"hair",
	"half",
	"hall",
	"hand",
	"handful",
	"handle",
	"hang",
	"happen",
	"happy",
	"hard",
	"hardly",
	"hat",
	"have",
	"head",
	"headline",
	"headquarters",
	"health",
	"healthy",
	"hear",
	"hearing",
	"heart",
	"heat",
	"heaven",
	"heavily",
	"heavy",
	"heel",
	"height",
	"helicopter",
	"hell",
	"hello",
	"help",
	"helpful",
	"here",
	"heritage",
	"hero",
	"herself",
	"hey",
	"hide",
	"high",
	"highlight",
	"highly",
	"highway",
	"hill",
	"himself",
	"hip",
	"hire",
	"historian",
	"historic",
	"historical",
	"history",
	"hit",
	"hold",
	"hole",
	"holiday",
	"holy",
	"home",
	"honest",
	"honey",
	"honor",
	"hope",
	"horizon",
	"horse",
	"hospital",
	"host",
	"hot",
	"hotel",
	"hour",
	"house",
	"household",
	"housing",
	"how",
	"however",
	"huge",
	"human",
	"humor",
	"hundred",
	"hungry",
	"hunter",
	"hunting",
	"husband",
	"hypothesis",
	"ice",
	"idea",
	"ideal",
	"identification",
	"identify",
	"identity",
	"ignore",
	"illustrate",
	"image",
	"imagination",
	"imagine",
	"immediate",
	"immediately",
	"immigrant",
	"immigration",
	"impact",
	"implement",
	"implication",
	"imply",
	"importance",
	"important",
	"impose",
	"impossible",
	"impress",
	"impression",
	"impressive",
	"improve",
	"improvement",
	"incentive",
	"incident",
	"include",
	"including",
	"income",
	"incorporate",
	"increase",
	"increased",
	"increasing",
	"increasingly",
	"incredible",
	"indeed",
	"independence",
	"independent",
	"index",
	"indicate",
	"indication",
	"individual",
	"industrial",
	"industry",
	"infant",
	"inflation",
	"influence",
	"inform",
	"information",
	"ingredient",
	"initial",
	"initially",
	"initiative",
	"inner",
	"innocent",
	"inquiry",
	"inside",
	"insight",
	"insist",
	"inspire",
	"install",
	"instance",
	"instead",
	"institution",
	"institutional",
	"instruction",
	"instructor",
	"instrument",
	"insurance",
	"intellectual",
	"intelligence",
	"intend",
	"intense",
	"intensity",
	"intention",
	"interaction",
	"interest",
	"interested",
	"interesting",
	"internal",
	"international",
	"interpret",
	"interpretation",
	"intervention",
	"interview",
	"into",
	"introduce",
	"introduction",
	"invest",
	"investigate",
	"investigation",
	"investigator",
	"investment",
	"investor",
	"invite",
	"involve",
	"involved",
	"involvement",
	"iron",
	"island",
	"issue",
	"item",
	"its",
	"itself",
	"jacket",
	"jet",
	"job",
	"join",
	"joint",
	"joke",
	"journal",
	"journalist",
	"journey",
	"joy",
	"judge",
	"judgment",
	"juice",
	"jump",
	"junior",
	"jury",
	"just",
	"justice",
	"justify",
	"keep",
	"key",
	"kick",
	"kid",
	"kind",
	"king",
	"kiss",
	"kitchen",
	"knee",
	"knife",
	"knock",
	"know",
	"knowledge",
	"lab",
	"label",
	"labor",
	"laboratory",
	"lady",
	"lake",
	"land",
	"landscape",
	"language",
	"lap",
	"large",
	"largely",
	"last",
	"late",
	"later",
	"latter",
	"laugh",
	"launch",
	"law",
	"lawn",
	"lawsuit",
	"lawyer",
	"lay",
	"layer",
	"lead",
	"leader",
	"leadership",
	"leading",
	"leaf",
	"league",
	"lean",
	"learn",
	"learning",
	"least",
	"leather",
	"leave",
	"left",
	"leg",
	"legacy",
	"legal",
	"legend",
	"legislation",
	"legitimate",
	"lemon",
	"length",
	"less",
	"lesson",
	"let",
	"letter",
	"level",
	"liberal",
	"library",
	"license",
	"lie",
	"life",
	"lifestyle",
	"lifetime",
	"lift",
	"light",
	"like",
	"likely",
	"limit",
	"limitation",
	"limited",
	"line",
	"link",
	"lip",
	"list",
	"listen",
	"literally",
	"literary",
	"literature",
	"little",
	"live",
	"living",
	"load",
	"loan",
	"local",
	"locate",
	"location",
	"lock",
	"long",
	"look",
	"loose",
	"lose",
	"lost",
	"lot",
	"lots",
	"loud",
	"love",
	"lovely",
	"lover",
	"low",
	"lower",
	"luck",
	"lucky",
	"lunch",
	"lung",
	"machine",
	"magazine",
	"mail",
	"main",
	"mainly",
	"maintain",
	"maintenance",
	"major",
	"majority",
	"make",
	"maker",
	"makeup",
	"male",
	"mall",
	"man",
	"manage",
	"management",
	"manager",
	"manner",
	"manufacturer",
	"manufacturing",
	"many",
	"map",
	"margin",
	"mark",
	"market",
	"marketing",
	"marriage",
	"married",
	"marry",
	"mask",
	"mass",
	"massive",
	"master",
	"match",
	"material",
	"math",
	"matter",
	"may",
	"maybe",
	"mayor",
	"meal",
	"mean",
	"meaning",
	"meanwhile",
	"measure",
	"measurement",
	"meat",
	"mechanism",
	"media",
	"medical",
	"medication",
	"medicine",
	"medium",
	"meet",
	"meeting",
	"member",
	"membership",
	"memory",
	"mental",
	"mention",
	"menu",
	"mere",
	"merely",
	"message",
	"metal",
	"meter",
	"method",
	"middle",
	"might",
	"military",
	"milk",
	"million",
	"mind",
	"mine",
	"minister",
	"minor",
	"minority",
	"minute",
	"miracle",
	"mirror",
	"miss",
	"missile",
	"mission",
	"mix",
	"mixture",
	"mode",
	"model",
	"moderate",
	"modern",
	"modest",
	"mom",
	"moment",
	"money",
	"monitor",
	"month",
	"mood",
	"moon",
	"moral",
	"more",
	"moreover",
	"morning",
	"mortgage",
	"most",
	"mostly",
	"mother",
	"motion",
	"motivation",
	"motor",
	"mount",
	"mountain",
	"mouse",
	"mouth",
	"move",
	"movement",
	"movie",
	"much",
	"multiple",
	"muscle",
	"museum",
	"music",
	"musical",
	"musician",
	"must",
	"mutual",
	"myself",
	"mystery",
	"myth",
	"naked",
	"name",
	"narrative",
	"narrow",
	"nation",
	"national",
	"native",
	"natural",
	"naturally",
	"nature",
	"near",
	"nearby",
	"nearly",
	"necessarily",
	"necessary",
	"neck",
	"need",
	"negative",
	"negotiate",
	"negotiation",
	"neighbor",
	"neighborhood",
	"neither",
	"nerve",
	"net",
	"network",
	"never",
	"nevertheless",
	"new",
	"newly",
	"news",
	"newspaper",
	"next",
	"nice",
	"night",
	"nine",
	"nobody",
	"nod",
	"nomination",
	"none",
	"nonetheless",
	"nor",
	"normal",
	"normally",
	"north",
	"northern",
	"nose",
	"not",
	"note",
	"nothing",
	"notice",
	"notion",
	"novel",
	"now",
	"nowhere",
	"nuclear",
	"number",
	"numerous",
	"nurse",
	"nut",
	"object",
	"objective",
	"obligation",
	"observation",
	"observe",
	"observer",
	"obtain",
	"obvious",
	"obviously",
	"occasion",
	"occasionally",
	"occupation",
	"occupy",
	"occur",
	"ocean",
	"odd",
	"odds",
	"off",
	"offer",
	"office",
	"officer",
	"official",
	"often",
	"oil",
	"okay",
	"old",
	"once",
	"one",
	"ongoing",
	"onion",
	"online",
	"only",
	"onto",
	"open",
	"opening",
	"operate",
	"operating",
	"operation",
	"operator",
	"opinion",
	"opponent",
	"opportunity",
	"oppose",
	"opposite",
	"opposition",
	"option",
	"orange",
	"order",
	"ordinary",
	"organic",
	"organization",
	"organize",
	"orientation",
	"origin",
	"original",
	"originally",
	"other",
	"others",
	"otherwise",
	"ought",
	"our",
	"ourselves",
	"out",
	"outcome",
	"outside",
	"oven",
	"over",
	"overall",
	"overcome",
	"overlook",
	"owe",
	"own",
	"owner",
	"pace",
	"pack",
	"package",
	"page",
	"paint",
	"painter",
	"painting",
	"pair",
	"pale",
	"palm",
	"pan",
	"panel",
	"pant",
	"paper",
	"parent",
	"park",
	"parking",
	"part",
	"participant",
	"participate",
	"participation",
	"particular",
	"particularly",
	"partly",
	"partner",
	"partnership",
	"party",
	"pass",
	"passage",
	"passenger",
	"passion",
	"past",
	"patch",
	"path",
	"patient",
	"pattern",
	"pause",
	"pay",
	"payment",
	"peace",
	"peak",
	"peer",
	"people",
	"pepper",
	"per",
	"perceive",
	"percentage",
	"perception",
	"perfect",
	"perfectly",
	"perform",
	"performance",
	"perhaps",
	"period",
	"permanent",
	"permission",
	"permit",
	"person",
	"personal",
	"personality",
	"personally",
	"personnel",
	"perspective",
	"persuade",
	"pet",
	"phase",
	"phenomenon",
	"philosophy",
	"phone",
	"photo",
	"photograph",
	"photographer",
	"phrase",
	"physical",
	"physically",
	"physician",
	"piano",
	"pick",
	"picture",
	"pie",
	"piece",
	"pile",
	"pilot",
	"pine",
	"pink",
	"pipe",
	"pitch",
	"place",
	"plan",
	"plane",
	"planet",
	"planning",
	"plant",
	"plastic",
	"plate",
	"platform",
	"play",
	"player",
	"please",
	"pleasure",
	"plenty",
	"plot",
	"plus",
	"pocket",
	"poem",
	"poet",
	"poetry",
	"point",
	"pole",
	"police",
	"policy",
	"political",
	"politically",
	"politician",
	"politics",
	"poll",
	"pool",
	"pop",
	"popular",
	"population",
	"porch",
	"port",
	"portion",
	"portrait",
	"portray",
	"pose",
	"position",
	"positive",
	"possess",
	"possibility",
	"possible",
	"possibly",
	"post",
	"pot",
	"potato",
	"potential",
	"potentially",
	"pound",
	"pour",
	"powder",
	"power",
	"powerful",
	"practical",
	"practice",
	"pray",
	"prayer",
	"precisely",
	"predict",
	"prefer",
	"preference",
	"pregnancy",
	"pregnant",
	"preparation",
	"prepare",
	"prescription",
	"presence",
	"present",
	"presentation",
	"preserve",
	"president",
	"presidential",
	"press",
	"pretend",
	"pretty",
	"prevent",
	"previous",
	"previously",
	"price",
	"pride",
	"priest",
	"primarily",
	"primary",
	"prime",
	"principal",
	"principle",
	"print",
	"prior",
	"priority",
	"privacy",
	"private",
	"probably",
	"procedure",
	"proceed",
	"process",
	"produce",
	"producer",
	"product",
	"production",
	"profession",
	"professional",
	"professor",
	"profile",
	"profit",
	"program",
	"progress",
	"project",
	"prominent",
	"promise",
	"promote",
	"prompt",
	"proof",
	"proper",
	"properly",
	"property",
	"proportion",
	"proposal",
	"propose",
	"proposed",
	"prosecutor",
	"prospect",
	"protect",
	"protection",
	"protein",
	"protest",
	"proud",
	"prove",
	"provide",
	"provider",
	"province",
	"provision",
	"psychological",
	"psychologist",
	"psychology",
	"public",
	"publication",
	"publicly",
	"publish",
	"publisher",
	"pull",
	"purchase",
	"pure",
	"purpose",
	"pursue",
	"push",
	"put",
	"qualify",
	"quality",
	"quarter",
	"quarterback",
	"question",
	"quick",
	"quickly",
	"quiet",
	"quietly",
	"quit",
	"quite",
	"quote",
	"race",
	"racial",
	"radical",
	"radio",
	"rail",
	"rain",
	"raise",
	"range",
	"rank",
	"rapid",
	"rapidly",
	"rare",
	"rarely",
	"rate",
	"rather",
	"rating",
	"ratio",
	"raw",
	"reach",
	"react",
	"reaction",
	"read",
	"reader",
	"reading",
	"ready",
	"real",
	"reality",
	"realize",
	"really",
	"reason",
	"reasonable",
	"recall",
	"receive",
	"recent",
	"recently",
	"recipe",
	"recognition",
	"recognize",
	"recommend",
	"recommendation",
	"record",
	"recording",
	"recover",
	"recovery",
	"recruit",
	"red",
	"reduce",
	"reduction",
	"refer",
	"reference",
	"reflect",
	"reflection",
	"reform",
	"refugee",
	"refuse",
	"regard",
	"regarding",
	"regardless",
	"regime",
	"region",
	"regional",
	"register",
	"regular",
	"regularly",
	"regulate",
	"regulation",
	"reinforce",
	"relate",
	"relation",
	"relationship",
	"relative",
	"relatively",
	"relax",
	"release",
	"relevant",
	"relief",
	"religion",
	"religious",
	"rely",
	"remain",
	"remaining",
	"remarkable",
	"remember",
	"remind",
	"remote",
	"remove",
	"repeat",
	"repeatedly",
	"replace",
	"reply",
	"report",
	"reporter",
	"represent",
	"representation",
	"representative",
	"reputation",
	"request",
	"require",
	"requirement",
	"research",
	"researcher",
	"resemble",
	"reservation",
	"resident",
	"resist",
	"resolution",
	"resolve",
	"resort",
	"resource",
	"respect",
	"respond",
	"respondent",
	"response",
	"responsibility",
	"responsible",
	"rest",
	"restaurant",
	"restore",
	"restriction",
	"result",
	"retain",
	"retire",
	"retirement",
	"return",
	"reveal",
	"revenue",
	"review",
	"revolution",
	"rhythm",
	"rice",
	"rich",
	"rid",
	"ride",
	"rifle",
	"right",
	"ring",
	"rise",
	"river",
	"road",
	"rock",
	"role",
	"roll",
	"romantic",
	"roof",
	"room",
	"root",
	"rope",
	"rose",
	"roughly",
	"round",
	"route",
	"routine",
	"row",
	"rub",
	"rule",
	"run",
	"running",
	"rural",
	"rush",
	"sacred",
	"safe",
	"safety",
	"sake",
	"salad",
	"salary",
	"sale",
	"sales",
	"salt",
	"same",
	"sample",
	"sanction",
	"sand",
	"satellite",
	"satisfaction",
	"satisfy",
	"sauce",
	"save",
	"saving",
	"say",
	"scale",
	"scenario",
	"scene",
	"schedule",
	"scheme",
	"scholar",
	"scholarship",
	"school",
	"science",
	"scientific",
	"scientist",
	"scope",
	"score",
	"screen",
	"script",
	"sea",
	"search",
	"season",
	"seat",
	"second",
	"secret",
	"secretary",
	"section",
	"sector",
	"secure",
	"security",
	"see",
	"seed",
	"seek",
	"seem",
	"segment",
	"seize",
	"select",
	"selection",
	"self",
	"sell",
	"senator",
	"send",
	"senior",
	"sense",
	"sensitive",
	"sentence",
	"separate",
	"sequence",
	"series",
	"serious",
	"seriously",
	"serve",
	"service",
	"session",
	"set",
	"setting",
	"settle",
	"settlement",
	"seven",
	"several",
	"shade",
	"shadow",
	"shake",
	"shall",
	"shape",
	"share",
	"sharp",
	"sheet",
	"shelf",
	"shell",
	"shelter",
	"shift",
	"shine",
	"ship",
	"shirt",
	"shoe",
	"shop",
	"shopping",
	"shore",
	"short",
	"shortly",
	"shot",
	"should",
	"shoulder",
	"shout",
	"show",
	"shower",
	"shrug",
	"shut",
	"side",
	"sigh",
	"sight",
	"sign",
	"signal",
	"significance",
	"significant",
	"significantly",
	"silence",
	"silent",
	"silver",
	"similar",
	"similarly",
	"simple",
	"simply",
	"since",
	"sing",
	"singer",
	"single",
	"sink",
	"sir",
	"sister",
	"sit",
	"site",
	"situation",
	"six",
	"size",
	"ski",
	"skill",
	"skin",
	"sky",
	"sleep",
	"slice",
	"slide",
	"slight",
	"slightly",
	"slip",
	"slow",
	"slowly",
	"small",
	"smart",
	"smell",
	"smile",
	"smooth",
	"snap",
	"snow",
	"soccer",
	"social",
	"society",
	"soft",
	"software",
	"soil",
	"solar",
	"solid",
	"solution",
	"solve",
	"some",
	"somebody",
	"somehow",
	"someone",
	"something",
	"sometimes",
	"somewhat",
	"somewhere",
	"son",
	"song",
	"soon",
	"sophisticated",
	"sorry",
	"sort",
	"soul",
	"sound",
	"soup",
	"source",
	"south",
	"southern",
	"space",
	"speak",
	"speaker",
	"special",
	"specialist",
	"species",
	"specific",
	"specifically",
	"speech",
	"speed",
	"spend",
	"spending",
	"spin",
	"spirit",
	"spiritual",
	"split",
	"spokesman",
	"sport",
	"spot",
	"spread",
	"spring",
	"square",
	"squeeze",
	"stability",
	"stable",
	"staff",
	"stage",
	"stair",
	"stake",
	"stand",
	"standard",
	"standing",
	"star",
	"stare",
	"start",
	"state",
	"statement",
	"station",
	"statistics",
	"status",
	"stay",
	"steady",
	"steal",
	"steel",
	"step",
	"stick",
	"still",
	"stir",
	"stock",
	"stomach",
	"stone",
	"stop",
	"storage",
	"store",
	"storm",
	"story",
	"straight",
	"strange",
	"stranger",
	"strategic",
	"strategy",
	"stream",
	"street",
	"strength",
	"strengthen",
	"stretch",
	"string",
	"strip",
	"strong",
	"strongly",
	"structure",
	"student",
	"studio",
	"study",
	"stuff",
	"style",
	"subject",
	"submit",
	"subsequent",
	"substance",
	"substantial",
	"succeed",
	"success",
	"successful",
	"successfully",
	"such",
	"sudden",
	"suddenly",
	"sue",
	"sufficient",
	"sugar",
	"suggest",
	"suggestion",
	"suit",
	"summer",
	"summit",
	"sun",
	"super",
	"supply",
	"support",
	"supporter",
	"suppose",
	"supposed",
	"sure",
	"surely",
	"surface",
	"surgery",
	"surprise",
	"surprised",
	"surprising",
	"surprisingly",
	"surround",
	"survey",
	"survival",
	"survive",
	"survivor",
	"sustain",
	"swear",
	"sweep",
	"sweet",
	"swim",
	"swing",
	"switch",
	"symbol",
	"system",
	"table",
	"tablespoon",
	"tactic",
	"tail",
	"take",
	"tale",
	"talent",
	"talk",
	"tall",
	"tank",
	"tap",
	"tape",
	"target",
	"task",
	"taste",
	"tax",
	"taxpayer",
	"tea",
	"teach",
	"teacher",
	"teaching",
	"team",
	"tear",
	"teaspoon",
	"technical",
	"technique",
	"technology",
	"teen",
	"teenager",
	"telephone",
	"telescope",
	"television",
	"tell",
	"temperature",
	"temporary",
	"ten",
	"tend",
	"tendency",
	"tennis",
	"tent",
	"term",
	"terms",
	"territory",
	"test",
	"testify",
	"testimony",
	"testing",
	"text",
	"than",
	"thank",
	"thanks",
	"that",
	"the",
	"theater",
	"their",
	"them",
	"theme",
	"themselves",
	"then",
	"theory",
	"therapy",
	"there",
	"therefore",
	"thick",
	"thin",
	"thing",
	"think",
	"thinking",
	"third",
	"thirty",
	"though",
	"thought",
	"thousand",
	"three",
	"throat",
	"through",
	"throughout",
	"throw",
	"ticket",
	"tie",
	"tight",
	"time",
	"tiny",
	"tip",
	"tire",
	"tired",
	"tissue",
	"title",
	"today",
	"toe",
	"together",
	"tomato",
	"tomorrow",
	"tone",
	"tongue",
	"tonight",
	"too",
	"tool",
	"tooth",
	"top",
	"topic",
	"toss",
	"total",
	"totally",
	"touch",
	"tough",
	"tour",
	"tourist",
	"tournament",
	"toward",
	"towards",
	"tower",
	"town",
	"toy",
	"trace",
	"track",
	"trade",
	"tradition",
	"traditional",
	"traffic",
	"trail",
	"train",
	"training",
	"transfer",
	"transform",
	"transformation",
	"transition",
	"translate",
	"transportation",
	"travel",
	"treat",
	"treatment",
	"treaty",
	"tree",
	"tremendous",
	"trend",
	"trial",
	"tribe",
	"trip",
	"troop",
	"truck",
	"true",
	"truly",
	"trust",
	"truth",
	"try",
	"tube",
	"tunnel",
	"turn",
	"twelve",
	"twenty",
	"twice",
	"twin",
	"two",
	"type",
	"typical",
	"typically",
	"ultimate",
	"ultimately",
	"unable",
	"uncle",
	"under",
	"undergo",
	"understand",
	"understanding",
	"unfortunately",
	"uniform",
	"union",
	"unique",
	"unit",
	"universal",
	"universe",
	"university",
	"unknown",
	"unless",
	"unlike",
	"unlikely",
	"until",
	"unusual",
	"upon",
	"upper",
	"urban",
	"urge",
	"use",
	"used",
	"useful",
	"user",
	"usual",
	"usually",
	"utility",
	"vacation",
	"valley",
	"valuable",
	"value",
	"variable",
	"variation",
	"variety",
	"various",
	"vary",
	"vast",
	"vegetable",
	"vehicle",
	"venture",
	"version",
	"versus",
	"very",
	"vessel",
	"veteran",
	"via",
	"victory",
	"video",
	"view",
	"viewer",
	"village",
	"virtually",
	"virtue",
	"visible",
	"vision",
	"visit",
	"visitor",
	"visual",
	"vital",
	"voice",
	"volume",
	"volunteer",
	"vote",
	"voter",
	"wage",
	"wait",
	"wake",
	"walk",
	"wall",
	"wander",
	"want",
	"warm",
	"warn",
	"warning",
	"wash",
	"watch",
	"water",
	"wave",
	"way",
	"wealth",
	"wealthy",
	"wear",
	"weather",
	"wedding",
	"week",
	"weekend",
	"weekly",
	"weigh",
	"weight",
	"welcome",
	"welfare",
	"well",
	"west",
	"western",
	"wet",
	"what",
	"whatever",
	"wheel",
	"when",
	"whenever",
	"where",
	"whereas",
	"whether",
	"which",
	"while",
	"whisper",
	"white",
	"who",
	"whole",
	"whom",
	"whose",
	"why",
	"wide",
	"widely",
	"widespread",
	"wife",
	"wild",
	"will",
	"willing",
	"win",
	"wind",
	"window",
	"wine",
	"wing",
	"winner",
	"winter",
	"wipe",
	"wire",
	"wisdom",
	"wise",
	"wish",
	"with",
	"withdraw",
	"within",
	"without",
	"witness",
	"woman",
	"wonder",
	"wonderful",
	"wood",
	"wooden",
	"word",
	"work",
	"worker",
	"working",
	"works",
	"workshop",
	"world",
	"worried",
	"worth",
	"would",
	"wrap",
	"write",
	"writer",
	"writing",
	"yard",
	"yeah",
	"year",
	"yell",
	"yellow",
	"yes",
	"yesterday",
	"yet",
	"yield",
	"young",
	"your",
	"yours",
	"yourself",
	"youth",
	"zone"
];
const random = (min, max) => Math.floor(min + Math.random() * (max - min + 1));
function getRandomWord() {
	return words[random(0, words.length - 1)];
}
//#endregion
//#region src/create/random-name.ts
const isTest = process.env.VITE_PLUS_CLI_TEST === "1";
function getRandomWords() {
	const first = getRandomWord();
	let second;
	do
		second = getRandomWord();
	while (second === first);
	return [first, second];
}
function getRandomProjectName(options = {}) {
	const { scope, fallbackName } = options;
	const projectName = isTest && fallbackName ? fallbackName : getRandomWords().join("-");
	return scope ? `${scope}/${projectName}` : projectName;
}
//#endregion
//#region src/create/utils.ts
function copy(src, dest) {
	if (fs.statSync(src).isDirectory()) copyDir(src, dest);
	else fs.copyFileSync(src, dest);
}
function copyDir(srcDir, destDir) {
	fs.mkdirSync(destDir, { recursive: true });
	for (const file of fs.readdirSync(srcDir)) copy(path.resolve(srcDir, file), path.resolve(destDir, file));
}
/**
* Format the target directory into a valid directory name and package name
*
* Examples:
* ```
* # invalid target directories
* ./ -> { directory: '', packageName: '', error: 'Invalid target directory' }
* /foo/bar -> { directory: '', packageName: '', error: 'Absolute path is not allowed' }
* @scope/ -> { directory: '', packageName: '', error: 'Invalid target directory' }
* ../../foo/bar -> { directory: '', packageName: '', error: 'Invalid target directory' }
*
* # valid target directories
* ./my-package -> { directory: './my-package', packageName: 'my-package' }
* ./foo/bar-package -> { directory: './foo/bar-package', packageName: 'bar-package' }
* ./foo/bar-package/ -> { directory: './foo/bar-package', packageName: 'bar-package' }
* my-package -> { directory: 'my-package', packageName: 'my-package' }
* @my-scope/my-package -> { directory: 'my-package', packageName: '@my-scope/my-package' }
* foo/@my-scope/my-package -> { directory: 'foo/my-package', packageName: '@scope/my-package' }
* ./foo/@my-scope/my-package -> { directory: './foo/my-package', packageName: '@scope/my-package' }
* ./foo/bar/@scope/my-package -> { directory: './foo/bar/my-package', packageName: '@scope/my-package' }
* ```
*/
function formatTargetDir(input) {
	let targetDir = path.normalize(input.trim());
	const parsed = path.parse(targetDir);
	if (parsed.root || path.isAbsolute(targetDir)) return {
		directory: "",
		packageName: "",
		error: "Absolute path is not allowed"
	};
	if (targetDir.includes("..")) return {
		directory: "",
		packageName: "",
		error: "Relative path contains \"..\" which is not allowed"
	};
	let packageName = parsed.base;
	const parentName = path.basename(parsed.dir);
	if (parentName.startsWith("@")) {
		targetDir = path.join(path.dirname(parsed.dir), packageName);
		packageName = `${parentName}/${packageName}`;
	}
	const result = (0, import_lib.default)(packageName);
	if (!result.validForNewPackages) {
		const message = result.errors?.[0] ?? result.warnings?.[0] ?? "Invalid package name";
		return {
			directory: "",
			packageName: "",
			error: `Parsed package name "${packageName}" is invalid: ${message}`
		};
	}
	return {
		directory: targetDir,
		packageName
	};
}
function getProjectDirFromPackageName(packageName) {
	if (packageName.startsWith("@")) return packageName.split("/")[1];
	return packageName;
}
function setPackageName(projectDir, packageName) {
	editJsonFile(path.join(projectDir, "package.json"), (pkg) => {
		pkg.name = packageName;
		return pkg;
	});
}
function formatDisplayTargetDir(targetDir) {
	const normalized = targetDir.split(path.sep).join("/");
	if (normalized === "" || normalized === ".") return "./";
	if (normalized.startsWith("./") || normalized.startsWith("../") || normalized.startsWith("/") || normalized.startsWith("~")) return normalized;
	return `./${normalized}`;
}
//#endregion
//#region src/create/prompts.ts
async function promptPackageNameAndTargetDir(defaultPackageName, interactive) {
	let packageName;
	let targetDir;
	if (interactive) {
		const selected = await text({
			message: "Package name:",
			placeholder: defaultPackageName,
			defaultValue: defaultPackageName,
			validate: (value) => {
				if (value == null || value.length === 0) return;
				const result = value ? (0, import_lib.default)(value) : null;
				if (result?.validForNewPackages) return;
				return result?.errors?.[0] ?? result?.warnings?.[0] ?? "Invalid package name";
			}
		});
		if (Ct(selected)) cancelAndExit();
		packageName = selected;
		targetDir = getProjectDirFromPackageName(packageName);
	} else {
		packageName = defaultPackageName;
		targetDir = getProjectDirFromPackageName(packageName);
		log.info(`Using default package name: ${accent(packageName)}`);
	}
	return {
		packageName,
		targetDir
	};
}
async function promptTargetDir(defaultTargetDir, interactive, options) {
	let targetDir;
	if (interactive) {
		const selected = await text({
			message: "Target directory:",
			placeholder: defaultTargetDir,
			defaultValue: defaultTargetDir,
			validate: (value) => validateTargetDir(value ?? defaultTargetDir, options?.cwd).error
		});
		if (Ct(selected)) cancelAndExit();
		targetDir = validateTargetDir(selected ?? defaultTargetDir, options?.cwd).directory;
	} else {
		targetDir = validateTargetDir(defaultTargetDir, options?.cwd).directory;
		log.info(`Using default target directory: ${accent(targetDir)}`);
	}
	return targetDir;
}
function suggestAvailableTargetDir(defaultTargetDir, cwd) {
	let suggestedTargetDir = defaultTargetDir;
	let attempt = 1;
	while (!isTargetDirAvailable(path.join(cwd, suggestedTargetDir))) {
		suggestedTargetDir = getRandomProjectName({ fallbackName: `${defaultTargetDir}-${attempt}` });
		attempt++;
	}
	return suggestedTargetDir;
}
async function checkProjectDirExists(projectDirFullPath, interactive) {
	if (isTargetDirAvailable(projectDirFullPath)) return;
	if (!interactive) {
		log.info("Use --directory to specify a different location or remove the directory first");
		cancelAndExit(`Target directory "${projectDirFullPath}" is not empty`, 1);
	}
	const overwrite = await select({
		message: `Target directory "${projectDirFullPath}" is not empty. Please choose how to proceed:`,
		options: [{
			label: "Cancel operation",
			value: "no"
		}, {
			label: "Remove existing files and continue",
			value: "yes"
		}]
	});
	if (Ct(overwrite)) cancelAndExit();
	switch (overwrite) {
		case "yes":
			emptyDir(projectDirFullPath);
			break;
		case "no": cancelAndExit();
	}
}
function cancelAndExit(message = "Operation cancelled", exitCode = 0) {
	cancel(message);
	process.exit(exitCode);
}
function isEmpty(path) {
	const files = fs.readdirSync(path);
	return files.length === 0 || files.length === 1 && files[0] === ".git";
}
function emptyDir(dir) {
	if (!fs.existsSync(dir)) return;
	for (const file of fs.readdirSync(dir)) {
		if (file === ".git") continue;
		fs.rmSync(path.resolve(dir, file), {
			recursive: true,
			force: true
		});
	}
}
function isTargetDirAvailable(projectDirFullPath) {
	return !fs.existsSync(projectDirFullPath) || isEmpty(projectDirFullPath);
}
function validateTargetDir(input, cwd) {
	const value = input?.trim() ?? "";
	if (!value) return {
		directory: "",
		error: "Target directory is required"
	};
	const targetDir = path.normalize(value);
	if (!targetDir || targetDir === ".") return {
		directory: "",
		error: "Target directory is required"
	};
	if (path.isAbsolute(targetDir)) return {
		directory: "",
		error: "Absolute path is not allowed"
	};
	if (targetDir.includes("..")) return {
		directory: "",
		error: "Relative path contains \"..\" which is not allowed"
	};
	if (cwd && !isTargetDirAvailable(path.join(cwd, targetDir))) return {
		directory: "",
		error: `Target directory "${targetDir}" already exists`
	};
	return { directory: targetDir };
}
//#endregion
//#region src/create/templates/generator.ts
async function executeGeneratorScaffold(workspaceInfo, templateInfo, options) {
	if (!options?.silent) log.step("Creating generator scaffold...");
	let description;
	if (templateInfo.interactive) {
		const defaultDescription = "Generate new components for our monorepo";
		const descPrompt = await text({
			message: "Description:",
			placeholder: defaultDescription,
			defaultValue: defaultDescription
		});
		if (!Ct(descPrompt)) description = descPrompt;
	}
	const fullPath = path.join(workspaceInfo.rootDir, templateInfo.targetDir);
	copyDir(path.join(templatesDir, "generator"), fullPath);
	fs.chmodSync(path.join(fullPath, "bin/index.ts"), "755");
	editJsonFile(path.join(fullPath, "package.json"), (pkg) => {
		pkg.name = templateInfo.packageName;
		if (description) pkg.description = description;
		return pkg;
	});
	if (!options?.silent) log.success("Generator scaffold created");
	return {
		exitCode: 0,
		projectDir: templateInfo.targetDir
	};
}
//#endregion
//#region src/create/templates/remote.ts
const { gray, yellow } = color;
async function executeRemoteTemplate(workspaceInfo, templateInfo, options) {
	const silent = options?.silent ?? false;
	if (!silent) log.step("Generating project…");
	let isGitHubTemplate = templateInfo.command === "degit";
	let result;
	if (templateInfo.command === "node") {
		const command = templateInfo.command;
		const args = templateInfo.args;
		const envs = templateInfo.envs;
		if (!silent) log.info(`Running: ${gray(`${command} ${args.join(" ")}`)}`);
		result = await runCommandAndDetectProjectDir({
			command,
			args,
			cwd: workspaceInfo.rootDir,
			envs
		}, templateInfo.parentDir);
	} else result = await runRemoteTemplateCommand(workspaceInfo, workspaceInfo.rootDir, templateInfo, true, silent);
	const exitCode = result.exitCode;
	if (exitCode === 127) {
		log.info(yellow("\nTroubleshooting:"));
		log.info(`  ${gray("•")} Command not found. Make sure Node.js is installed`);
	} else if (isGitHubTemplate && exitCode !== 0) {
		log.info(yellow("\nTroubleshooting:"));
		log.info(`  ${gray("•")} Make sure the GitHub repository exists`);
		log.info(`  ${gray("•")} Check your internet connection`);
		log.info(`  ${gray("•")} Repository might be private (requires authentication)`);
	}
	return result;
}
async function runRemoteTemplateCommand(workspaceInfo, cwd, templateInfo, detectCreatedProjectDir, silent = false) {
	autoFixRemoteTemplateCommand(templateInfo, workspaceInfo);
	const remotePackageName = templateInfo.command;
	const execArgs = [...templateInfo.args];
	const envs = templateInfo.envs;
	const { command, args } = formatDlxCommand(remotePackageName, execArgs, workspaceInfo);
	if (!silent) log.info(`Running: ${gray(`${command} ${args.join(" ")}`)}`);
	if (detectCreatedProjectDir) return await runCommandAndDetectProjectDir({
		command,
		args,
		cwd,
		envs
	}, templateInfo.parentDir);
	if (silent) return await runCommandSilently({
		command,
		args,
		cwd,
		envs
	});
	return await runCommand$1({
		command,
		args,
		cwd,
		envs
	});
}
function autoFixRemoteTemplateCommand(templateInfo, workspaceInfo) {
	let packageName = templateInfo.command;
	const indexOfAt = packageName.indexOf("@", 2);
	if (indexOfAt !== -1) packageName = packageName.substring(0, indexOfAt);
	if (packageName === "create-vite") {
		templateInfo.args.push("--no-immediate");
		templateInfo.args.push("--no-rolldown");
	} else if (packageName === "@tanstack/create-start") {
		templateInfo.args.push("--no-install");
		templateInfo.args.push("--no-toolchain");
	}
	if (workspaceInfo.isMonorepo) {
		if (packageName === "create-nuxt") templateInfo.args.push("--no-gitInit");
		else if (packageName === "@tanstack/create-start") templateInfo.args.push("--no-git");
	}
}
//#endregion
//#region src/create/templates/builtin.ts
async function executeBuiltinTemplate(workspaceInfo, templateInfo, options) {
	assert(templateInfo.targetDir, "targetDir is required");
	assert(templateInfo.packageName, "packageName is required");
	if (templateInfo.command === BuiltinTemplate.generator) return await executeGeneratorScaffold(workspaceInfo, templateInfo, options);
	if (templateInfo.command === BuiltinTemplate.application) {
		templateInfo.command = "create-vite@latest";
		if (!templateInfo.interactive) templateInfo.args.push("--no-interactive");
		templateInfo.args.unshift(templateInfo.targetDir);
	} else if (templateInfo.command === BuiltinTemplate.library) {
		const libraryTemplateInfo = discoverTemplate(LibraryTemplateRepo, [templateInfo.targetDir], workspaceInfo);
		const result = await runRemoteTemplateCommand(workspaceInfo, workspaceInfo.rootDir, libraryTemplateInfo, false, options?.silent ?? false);
		setPackageName(path.join(workspaceInfo.rootDir, templateInfo.targetDir), templateInfo.packageName);
		return {
			...result,
			projectDir: templateInfo.targetDir
		};
	}
	const result = await runRemoteTemplateCommand(workspaceInfo, workspaceInfo.rootDir, templateInfo, false, options?.silent ?? false);
	setPackageName(path.join(workspaceInfo.rootDir, templateInfo.targetDir), templateInfo.packageName);
	return {
		...result,
		projectDir: templateInfo.targetDir
	};
}
//#endregion
//#region src/create/templates/monorepo.ts
const InitialMonorepoAppDir = "apps/website";
async function executeMonorepoTemplate(workspaceInfo, templateInfo, interactive, options) {
	assert(templateInfo.packageName, "packageName is required");
	assert(templateInfo.targetDir, "targetDir is required");
	workspaceInfo.monorepoScope = getScopeFromPackageName(templateInfo.packageName);
	const fullPath = path.join(workspaceInfo.rootDir, templateInfo.targetDir);
	let initGit = true;
	if (interactive && !options?.silent) {
		const selected = await confirm({
			message: `Initialize git repository:`,
			initialValue: true
		});
		if (Ct(selected)) {
			log.info("Operation cancelled. Skipping git initialization");
			initGit = false;
		} else initGit = selected;
	} else if (!options?.silent) log.info(`Initializing git repository (default: yes)`);
	if (!options?.silent) {
		log.info(`Target directory: ${formatDisplayTargetDir(templateInfo.targetDir)}`);
		log.step("Creating Vite+ monorepo...");
	}
	copyDir(path.join(templatesDir, "monorepo"), fullPath);
	renameFiles(fullPath);
	editJsonFile(path.join(fullPath, "package.json"), (pkg) => {
		pkg.name = templateInfo.packageName;
		return pkg;
	});
	if (workspaceInfo.packageManager === PackageManager.pnpm) {
		editJsonFile(path.join(fullPath, "package.json"), (pkg) => {
			pkg.workspaces = void 0;
			pkg.resolutions = void 0;
			return pkg;
		});
		const yarnrcPath = path.join(fullPath, ".yarnrc.yml");
		if (fs.existsSync(yarnrcPath)) fs.unlinkSync(yarnrcPath);
	} else if (workspaceInfo.packageManager === PackageManager.yarn) {
		editJsonFile(path.join(fullPath, "package.json"), (pkg) => {
			pkg.pnpm = void 0;
			return pkg;
		});
		const pnpmWorkspacePath = path.join(fullPath, "pnpm-workspace.yaml");
		if (fs.existsSync(pnpmWorkspacePath)) fs.unlinkSync(pnpmWorkspacePath);
	} else {
		editJsonFile(path.join(fullPath, "package.json"), (pkg) => {
			pkg.pnpm = void 0;
			return pkg;
		});
		const pnpmWorkspacePath = path.join(fullPath, "pnpm-workspace.yaml");
		if (fs.existsSync(pnpmWorkspacePath)) fs.unlinkSync(pnpmWorkspacePath);
		const yarnrcPath = path.join(fullPath, ".yarnrc.yml");
		if (fs.existsSync(yarnrcPath)) fs.unlinkSync(yarnrcPath);
	}
	if (!options?.silent) log.success("Monorepo template created");
	if (initGit) {
		const gitResult = spawn.sync("git", ["init"], {
			stdio: "pipe",
			cwd: fullPath
		});
		if (gitResult.status === 0) {
			if (!options?.silent) log.success("Git repository initialized");
		} else {
			log.warn("Failed to initialize git repository");
			if (gitResult.stderr) log.info(gitResult.stderr.toString());
		}
	}
	if (!options?.silent) log.step("Creating default application in apps/website...");
	const appResult = await runRemoteTemplateCommand(workspaceInfo, fullPath, discoverTemplate("create-vite@latest", [
		InitialMonorepoAppDir,
		"--template",
		"vanilla-ts",
		"--no-interactive"
	], workspaceInfo), false, options?.silent ?? false);
	if (appResult.exitCode !== 0) {
		log.error(`Failed to create default application: ${appResult.exitCode}`);
		return appResult;
	}
	const appPackageName = workspaceInfo.monorepoScope ? `${workspaceInfo.monorepoScope}/website` : "website";
	const appProjectPath = path.join(fullPath, InitialMonorepoAppDir);
	setPackageName(appProjectPath, appPackageName);
	rewriteMonorepoProject(appProjectPath, workspaceInfo.packageManager, void 0, options?.silent ?? false);
	if (!options?.silent) log.step("Creating default library in packages/utils...");
	const libraryDir = "packages/utils";
	const libraryResult = await runRemoteTemplateCommand(workspaceInfo, fullPath, discoverTemplate(LibraryTemplateRepo, [libraryDir], workspaceInfo), false, options?.silent ?? false);
	if (libraryResult.exitCode !== 0) {
		log.error(`Failed to create default library, exit code: ${libraryResult.exitCode}`);
		return libraryResult;
	}
	const libraryPackageName = workspaceInfo.monorepoScope ? `${workspaceInfo.monorepoScope}/utils` : "utils";
	const libraryProjectPath = path.join(fullPath, libraryDir);
	setPackageName(libraryProjectPath, libraryPackageName);
	rewriteMonorepoProject(libraryProjectPath, workspaceInfo.packageManager, void 0, options?.silent ?? false);
	return {
		exitCode: 0,
		projectDir: templateInfo.targetDir
	};
}
const RENAME_FILES = {
	_gitignore: ".gitignore",
	_npmrc: ".npmrc",
	"_yarnrc.yml": ".yarnrc.yml"
};
function renameFiles(projectDir) {
	for (const [from, to] of Object.entries(RENAME_FILES)) {
		const fromPath = path.join(projectDir, from);
		if (fs.existsSync(fromPath)) fs.renameSync(fromPath, path.join(projectDir, to));
	}
}
function getScopeFromPackageName(packageName) {
	if (packageName.startsWith("@")) return packageName.split("/")[0];
	return "";
}
//#endregion
//#region src/create/bin.ts
const helpMessage = renderCliDoc({
	usage: "vp create [TEMPLATE] [OPTIONS] [-- TEMPLATE_OPTIONS]",
	summary: "Use any builtin, local or remote template with Vite+.",
	documentationUrl: "https://viteplus.dev/guide/create",
	sections: [
		{
			title: "Arguments",
			rows: [{
				label: "TEMPLATE",
				description: [
					`Template name. Run \`${accent("vp create --list")}\` to see available templates.`,
					`- Default: ${accent("vite:monorepo")}, ${accent("vite:application")}, ${accent("vite:library")}, ${accent("vite:generator")}`,
					"- Remote: vite, @tanstack/start, create-next-app,",
					"  create-nuxt, github:user/repo, https://github.com/user/template-repo, etc.",
					"- Local: @company/generator-*, ./tools/create-ui-component"
				]
			}]
		},
		{
			title: "Options",
			rows: [
				{
					label: "--directory DIR",
					description: "Target directory for the generated project."
				},
				{
					label: "--agent NAME",
					description: "Create an agent instructions file for the specified agent."
				},
				{
					label: "--editor NAME",
					description: "Write editor config files for the specified editor."
				},
				{
					label: "--hooks",
					description: "Set up pre-commit hooks (default in non-interactive mode)"
				},
				{
					label: "--no-hooks",
					description: "Skip pre-commit hooks setup"
				},
				{
					label: "--verbose",
					description: "Show detailed scaffolding output"
				},
				{
					label: "--no-interactive",
					description: "Run in non-interactive mode"
				},
				{
					label: "--list",
					description: "List all available templates"
				},
				{
					label: "-h, --help",
					description: "Show this help message"
				}
			]
		},
		{
			title: "Template Options",
			lines: ["  Any arguments after -- are passed directly to the template."]
		},
		{
			title: "Examples",
			lines: [
				`  ${muted("# Interactive mode")}`,
				`  ${accent("vp create")}`,
				"",
				`  ${muted("# Use existing templates (shorthand expands to create-* packages)")}`,
				`  ${accent("vp create vite")}`,
				`  ${accent("vp create @tanstack/start")}`,
				`  ${accent("vp create vite -- --template react-ts")}`,
				"",
				`  ${muted("# Full package names also work")}`,
				`  ${accent("vp create create-vite")}`,
				`  ${accent("vp create create-next-app")}`,
				"",
				`  ${muted("# Create Vite+ monorepo, application, library, or generator scaffolds")}`,
				`  ${accent("vp create vite:monorepo")}`,
				`  ${accent("vp create vite:application")}`,
				`  ${accent("vp create vite:library")}`,
				`  ${accent("vp create vite:generator")}`,
				"",
				`  ${muted("# Use templates from GitHub (via degit)")}`,
				`  ${accent("vp create github:user/repo")}`,
				`  ${accent("vp create https://github.com/user/template-repo")}`
			]
		}
	]
});
const listTemplatesMessage = renderCliDoc({
	usage: "vp create --list",
	summary: "List available builtin and popular project templates.",
	documentationUrl: "https://viteplus.dev/guide/create",
	sections: [
		{
			title: "Vite+ Built-in Templates",
			rows: [
				{
					label: "vite:monorepo",
					description: "Create a new monorepo"
				},
				{
					label: "vite:application",
					description: "Create a new application"
				},
				{
					label: "vite:library",
					description: "Create a new library"
				},
				{
					label: "vite:generator",
					description: "Scaffold a new code generator (monorepo only)"
				}
			]
		},
		{
			title: "Popular Templates (shorthand)",
			rows: [
				{
					label: "vite",
					description: "Official Vite templates (create-vite)"
				},
				{
					label: "@tanstack/start",
					description: "TanStack applications (@tanstack/create-start)"
				},
				{
					label: "next-app",
					description: "Next.js application (create-next-app)"
				},
				{
					label: "nuxt",
					description: "Nuxt application (create-nuxt)"
				},
				{
					label: "react-router",
					description: "React Router application (create-react-router)"
				},
				{
					label: "vue",
					description: "Vue application (create-vue)"
				}
			]
		},
		{
			title: "Examples",
			lines: [
				`  ${accent("vp create")} ${muted("# interactive mode")}`,
				`  ${accent("vp create vite")} ${muted("# shorthand for create-vite")}`,
				`  ${accent("vp create @tanstack/start")} ${muted("# shorthand for @tanstack/create-start")}`,
				`  ${accent("vp create <template> -- <options>")} ${muted("# pass options to the template")}`
			]
		},
		{
			title: "Tip",
			lines: [`  You can use any npm template or git repo with ${accent("vp create")}.`]
		}
	]
});
function parseArgs() {
	const args = process.argv.slice(3);
	const separatorIndex = args.indexOf("--");
	const viteArgs = separatorIndex >= 0 ? args.slice(0, separatorIndex) : args;
	const templateArgs = separatorIndex >= 0 ? args.slice(separatorIndex + 1) : [];
	const parsed = lib_default(viteArgs, {
		alias: { h: "help" },
		boolean: [
			"help",
			"list",
			"all",
			"interactive",
			"hooks",
			"verbose"
		],
		string: [
			"directory",
			"agent",
			"editor"
		],
		default: { interactive: defaultInteractive() }
	});
	return {
		templateName: parsed._[0],
		options: {
			directory: parsed.directory,
			interactive: parsed.interactive,
			list: parsed.list || false,
			help: parsed.help || false,
			verbose: parsed.verbose || false,
			agent: parsed.agent,
			editor: parsed.editor,
			hooks: parsed.hooks
		},
		templateArgs
	};
}
function describeScaffold(templateName, templateArgs) {
	if (templateName === BuiltinTemplate.monorepo) return "Vite+ monorepo";
	if (templateName === BuiltinTemplate.generator) return "generator scaffold";
	if (templateName === BuiltinTemplate.library) return "TypeScript library";
	const selectedTemplate = getTemplateOption(templateArgs);
	if (selectedTemplate) return formatTemplateName(selectedTemplate);
	if (templateName === BuiltinTemplate.application) return "Vite application";
}
function getTemplateOption(args) {
	for (let index = 0; index < args.length; index++) {
		const arg = args[index];
		if (arg === "--template" || arg === "-t") return args[index + 1];
		if (arg.startsWith("--template=")) return arg.slice(11);
	}
}
function hasExplicitTargetDir(args) {
	return args[0] !== void 0 && !args[0].startsWith("-");
}
function formatTemplateName(templateName) {
	const templateAliases = {
		lit: "Lit",
		preact: "Preact",
		react: "React",
		"react-router": "React Router",
		solid: "Solid",
		svelte: "Svelte",
		vanilla: "Vanilla",
		vue: "Vue"
	};
	const isTypeScript = templateName.endsWith("-ts");
	const baseName = isTypeScript ? templateName.slice(0, -3) : templateName;
	return `${templateAliases[baseName] ?? baseName.split(/[-_]/).map((segment) => segment.charAt(0).toUpperCase() + segment.slice(1)).join(" ")} + ${isTypeScript ? "TypeScript" : "JavaScript"}`;
}
function formatDuration(durationMs) {
	if (durationMs < 1e3) return `${Math.max(1, durationMs)}ms`;
	const durationSeconds = durationMs / 1e3;
	if (durationSeconds < 10) return `${durationSeconds.toFixed(1)}s`;
	return `${Math.round(durationSeconds)}s`;
}
function getNextCommand(projectDir, command) {
	if (!projectDir || projectDir === ".") return command;
	return `cd ${projectDir} && ${command}`;
}
function showCreateSummary(options) {
	const { description, installSummary, nextCommand, packageManager, packageManagerVersion, projectDir } = options;
	log$1(`${styleText("magenta", "◇")} Scaffolded ${accent(projectDir)}${description ? ` with ${description}` : ""}`);
	log$1(`${styleText("gray", "•")} Node ${process.versions.node}  ${packageManager} ${packageManagerVersion}`);
	if (installSummary?.status === "installed") log$1(`${styleText("green", "✓")} Dependencies installed in ${formatDuration(installSummary.durationMs)}`);
	log$1(`${styleText("blue", "→")} Next: ${accent(nextCommand)}`);
}
async function main() {
	const { templateName, options, templateArgs } = parseArgs();
	let compactOutput = !options.verbose;
	if (options.help) {
		log$1(vitePlusHeader() + "\n");
		log$1(helpMessage);
		return;
	}
	if (options.list) {
		await showAvailableTemplates();
		return;
	}
	if (!templateName && !options.interactive) {
		console.error(`
A template name is required when running in non-interactive mode

Usage: vp create [TEMPLATE] [OPTIONS] [-- TEMPLATE_OPTIONS]

Example:
  ${muted("# Create a new application in non-interactive mode with a custom target directory")}
  vp create vite:application --no-interactive --directory=apps/my-app

Use \`vp create --list\` to list all available templates, or run \`vp create --help\` for more information.
`);
		process.exit(1);
	}
	if (options.interactive) intro(vitePlusHeader());
	let targetDir = "";
	let packageName = "";
	if (options.directory) {
		const formatted = formatTargetDir(options.directory);
		if (formatted.error) {
			log.error(formatted.error);
			cancelAndExit("The --directory option is invalid", 1);
		}
		targetDir = formatted.directory;
		packageName = formatted.packageName;
	}
	const cwd = process.cwd();
	const workspaceInfoOptional = await detectWorkspace$1(cwd);
	const isMonorepo = workspaceInfoOptional.isMonorepo;
	if (!isMonorepo) workspaceInfoOptional.rootDir = cwd;
	const cwdRelativeToRoot = isMonorepo && workspaceInfoOptional.rootDir !== cwd ? displayRelative(cwd, workspaceInfoOptional.rootDir) : "";
	const isInSubdirectory = cwdRelativeToRoot !== "";
	const cwdUnderParentDir = isInSubdirectory ? workspaceInfoOptional.parentDirs.some((dir) => cwdRelativeToRoot === dir || cwdRelativeToRoot.startsWith(`${dir}/`)) : true;
	const shouldOfferCwdOption = isInSubdirectory && !cwdUnderParentDir;
	let selectedTemplateName = templateName;
	let selectedTemplateArgs = [...templateArgs];
	let selectedAgentTargetPaths;
	let selectedEditor;
	let selectedParentDir;
	let remoteTargetDir;
	let shouldSetupHooks = false;
	if (!selectedTemplateName) {
		const template = await select({
			message: "",
			options: getInitialTemplateOptions(isMonorepo)
		});
		if (Ct(template)) cancelAndExit();
		selectedTemplateName = template;
	}
	const isBuiltinTemplate = selectedTemplateName.startsWith("vite:");
	if (!isBuiltinTemplate) compactOutput = false;
	if (targetDir && !isBuiltinTemplate) cancelAndExit("The --directory option is only available for builtin templates", 1);
	if (selectedTemplateName === BuiltinTemplate.monorepo && isMonorepo) {
		log.info("You are already in a monorepo workspace.\nUse a different template or run this command outside the monorepo");
		cancelAndExit("Cannot create a monorepo inside an existing monorepo", 1);
	}
	if (selectedTemplateName === BuiltinTemplate.generator && !isMonorepo) {
		log.info("The vite:generator template requires a monorepo workspace.\nRun this command inside a Vite+ monorepo, or create one first with `vp create vite:monorepo`");
		cancelAndExit("Cannot create a generator outside a monorepo", 1);
	}
	if (isInSubdirectory && !compactOutput) log.info(`Detected monorepo root at ${accent(workspaceInfoOptional.rootDir)}`);
	if (isMonorepo && options.interactive && !targetDir) {
		let parentDir;
		if (workspaceInfoOptional.parentDirs.length > 0 || isInSubdirectory) {
			const dirOptions = workspaceInfoOptional.parentDirs.map((dir) => ({
				label: `${dir}/`,
				value: dir,
				hint: ""
			}));
			if (shouldOfferCwdOption) dirOptions.push({
				label: `${cwdRelativeToRoot}/ (current directory)`,
				value: cwdRelativeToRoot,
				hint: ""
			});
			dirOptions.push({
				label: "other directory",
				value: "other",
				hint: "Enter a custom target directory"
			});
			const selected = await select({
				message: "Where should the new package be added to the monorepo:",
				options: dirOptions,
				initialValue: shouldOfferCwdOption ? cwdRelativeToRoot : inferParentDir(selectedTemplateName, workspaceInfoOptional) ?? workspaceInfoOptional.parentDirs[0]
			});
			if (Ct(selected)) cancelAndExit();
			if (selected !== "other") parentDir = selected;
		}
		if (!parentDir) {
			const customTargetDir = await text({
				message: "Where should the new package be added to the monorepo:",
				placeholder: "e.g., packages/",
				validate: (value) => {
					return value ? formatTargetDir(value).error : "Target directory is required";
				}
			});
			if (Ct(customTargetDir)) cancelAndExit();
			parentDir = customTargetDir;
		}
		selectedParentDir = parentDir;
	}
	if (isMonorepo && !options.interactive && !targetDir) {
		if (isInSubdirectory && !compactOutput) log.info(`Use ${accent("--directory")} to specify a different target location.`);
		selectedParentDir = inferParentDir(selectedTemplateName, workspaceInfoOptional) ?? workspaceInfoOptional.parentDirs[0];
	}
	if (isGitHubUrl(selectedTemplateName)) if (hasExplicitTargetDir(selectedTemplateArgs)) remoteTargetDir = selectedTemplateArgs[0];
	else {
		const inferredTargetDir = inferGitHubRepoName(selectedTemplateName) ?? "template";
		const remoteTargetBaseDir = selectedParentDir ? path.join(workspaceInfoOptional.rootDir, selectedParentDir) : workspaceInfoOptional.rootDir;
		const defaultTargetDir = suggestAvailableTargetDir(inferredTargetDir, remoteTargetBaseDir);
		if (defaultTargetDir !== inferredTargetDir && options.interactive) log.info(`  Target directory "${inferredTargetDir}" already exists. Suggested: ${accent(defaultTargetDir)}`);
		remoteTargetDir = await promptTargetDir(defaultTargetDir, options.interactive, { cwd: remoteTargetBaseDir });
		selectedTemplateArgs = [remoteTargetDir, ...selectedTemplateArgs];
	}
	if (isBuiltinTemplate && !targetDir) if (selectedTemplateName === BuiltinTemplate.monorepo) {
		const selected = await promptPackageNameAndTargetDir(getRandomProjectName({ fallbackName: "vite-plus-monorepo" }), options.interactive);
		packageName = selected.packageName;
		targetDir = selected.targetDir;
	} else {
		const selected = await promptPackageNameAndTargetDir(getRandomProjectName({
			scope: workspaceInfoOptional.monorepoScope,
			fallbackName: `vite-plus-${selectedTemplateName.split(":")[1]}`
		}), options.interactive);
		packageName = selected.packageName;
		targetDir = selectedParentDir ? path.join(selectedParentDir, selected.targetDir) : selected.targetDir;
	}
	const packageManager = workspaceInfoOptional.packageManager ?? await selectPackageManager(options.interactive, compactOutput);
	const shouldSilencePackageManagerInstallLog = compactOutput || isMonorepo && workspaceInfoOptional.packageManager !== void 0;
	const downloadResult = await downloadPackageManager$1(packageManager, workspaceInfoOptional.packageManagerVersion, options.interactive, shouldSilencePackageManagerInstallLog);
	const workspaceInfo = {
		...workspaceInfoOptional,
		packageManager,
		downloadPackageManager: downloadResult
	};
	const existingAgentTargetPaths = options.agent !== void 0 || !options.interactive ? void 0 : detectExistingAgentTargetPaths(workspaceInfoOptional.rootDir);
	selectedAgentTargetPaths = existingAgentTargetPaths !== void 0 ? existingAgentTargetPaths : await selectAgentTargetPaths({
		interactive: options.interactive,
		agent: options.agent,
		onCancel: () => cancelAndExit()
	});
	selectedEditor = (options.editor || !options.interactive ? void 0 : detectExistingEditor(workspaceInfoOptional.rootDir)) ?? await selectEditor({
		interactive: options.interactive,
		editor: options.editor,
		onCancel: () => cancelAndExit()
	});
	if (!isMonorepo) shouldSetupHooks = await promptGitHooks(options);
	const createProgress = options.interactive && compactOutput ? spinner({ indicator: "timer" }) : void 0;
	let createProgressStarted = false;
	let createProgressMessage = "Scaffolding project";
	const updateCreateProgress = (message) => {
		createProgressMessage = message;
		if (!createProgress) return;
		if (createProgressStarted) {
			createProgress.message(message);
			return;
		}
		createProgress.start(message);
		createProgressStarted = true;
	};
	const clearCreateProgress = () => {
		if (createProgress && createProgressStarted) {
			createProgress.clear();
			createProgressStarted = false;
		}
	};
	const failCreateProgress = (message) => {
		if (createProgress && createProgressStarted) {
			createProgress.error(message);
			createProgressStarted = false;
		}
	};
	const pauseCreateProgress = () => {
		if (createProgress && createProgressStarted) {
			createProgress.pause();
			createProgressStarted = false;
		}
	};
	const resumeCreateProgress = () => {
		if (createProgress && !createProgressStarted) {
			createProgress.resume(createProgressMessage);
			createProgressStarted = true;
		}
	};
	updateCreateProgress("Scaffolding project");
	const templateInfo = discoverTemplate(selectedTemplateName, selectedTemplateArgs, workspaceInfo, options.interactive);
	if (selectedParentDir) templateInfo.parentDir = selectedParentDir;
	if (targetDir) templateInfo.parentDir = void 0;
	if (remoteTargetDir) {
		const projectDir = templateInfo.parentDir ? path.join(templateInfo.parentDir, remoteTargetDir) : remoteTargetDir;
		pauseCreateProgress();
		await checkProjectDirExists(path.join(workspaceInfo.rootDir, projectDir), options.interactive);
		resumeCreateProgress();
	}
	if (templateInfo.command === BuiltinTemplate.monorepo) {
		updateCreateProgress("Creating monorepo");
		await checkProjectDirExists(path.join(workspaceInfo.rootDir, targetDir), options.interactive);
		const result = await executeMonorepoTemplate(workspaceInfo, {
			...templateInfo,
			packageName,
			targetDir
		}, options.interactive, { silent: compactOutput });
		const { projectDir } = result;
		if (result.exitCode !== 0 || !projectDir) {
			failCreateProgress("Scaffolding failed");
			cancelAndExit(`Failed to create monorepo, exit code: ${result.exitCode}`, result.exitCode);
		}
		const fullPath = path.join(workspaceInfo.rootDir, projectDir);
		updateCreateProgress("Writing agent instructions");
		pauseCreateProgress();
		await writeAgentInstructions({
			projectRoot: fullPath,
			targetPaths: selectedAgentTargetPaths,
			interactive: options.interactive,
			silent: compactOutput
		});
		resumeCreateProgress();
		updateCreateProgress("Writing editor configs");
		pauseCreateProgress();
		await writeEditorConfigs({
			projectRoot: fullPath,
			editorId: selectedEditor,
			interactive: options.interactive,
			silent: compactOutput
		});
		resumeCreateProgress();
		workspaceInfo.rootDir = fullPath;
		updateCreateProgress("Integrating monorepo");
		rewriteMonorepo(workspaceInfo, void 0, compactOutput);
		if (shouldSetupHooks) installGitHooks(fullPath, compactOutput);
		updateCreateProgress("Installing dependencies");
		const installSummary = await runViteInstall(fullPath, options.interactive, void 0, { silent: compactOutput });
		updateCreateProgress("Formatting code");
		await runViteFmt(fullPath, options.interactive, void 0, { silent: compactOutput });
		clearCreateProgress();
		showCreateSummary({
			description: describeScaffold(selectedTemplateName, selectedTemplateArgs),
			installSummary,
			nextCommand: getNextCommand(projectDir, `vp dev ${InitialMonorepoAppDir}`),
			packageManager: workspaceInfo.packageManager,
			packageManagerVersion: workspaceInfo.downloadPackageManager.version,
			projectDir
		});
		return;
	}
	let result;
	if (templateInfo.type === TemplateType.builtin) {
		if (!targetDir) {
			const selected = await promptPackageNameAndTargetDir(getRandomProjectName({
				scope: workspaceInfo.monorepoScope,
				fallbackName: `vite-plus-${templateInfo.command.split(":")[1]}`
			}), options.interactive);
			packageName = selected.packageName;
			targetDir = templateInfo.parentDir ? path.join(templateInfo.parentDir, selected.targetDir) : selected.targetDir;
		}
		pauseCreateProgress();
		await checkProjectDirExists(targetDir, options.interactive);
		resumeCreateProgress();
		updateCreateProgress("Generating project");
		result = await executeBuiltinTemplate(workspaceInfo, {
			...templateInfo,
			packageName,
			targetDir
		}, { silent: compactOutput });
	} else {
		updateCreateProgress("Generating project");
		result = await executeRemoteTemplate(workspaceInfo, templateInfo, { silent: compactOutput });
	}
	if (result.exitCode !== 0) {
		failCreateProgress("Scaffolding failed");
		process.exit(result.exitCode);
	}
	const projectDir = result.projectDir;
	if (!projectDir) {
		clearCreateProgress();
		process.exit(0);
	}
	const fullPath = path.join(workspaceInfo.rootDir, projectDir);
	const agentInstructionsRoot = isMonorepo ? workspaceInfo.rootDir : fullPath;
	updateCreateProgress("Writing agent instructions");
	pauseCreateProgress();
	await writeAgentInstructions({
		projectRoot: agentInstructionsRoot,
		targetPaths: selectedAgentTargetPaths,
		interactive: options.interactive,
		silent: compactOutput
	});
	resumeCreateProgress();
	updateCreateProgress("Writing editor configs");
	pauseCreateProgress();
	await writeEditorConfigs({
		projectRoot: fullPath,
		editorId: selectedEditor,
		interactive: options.interactive,
		silent: compactOutput
	});
	resumeCreateProgress();
	let installSummary;
	if (isMonorepo) {
		if (!compactOutput) log.step("Monorepo integration...");
		updateCreateProgress("Integrating into monorepo");
		rewriteMonorepoProject(fullPath, workspaceInfo.packageManager, void 0, compactOutput);
		if (workspaceInfo.packages.length > 0) {
			if (options.interactive) {
				pauseCreateProgress();
				const selectedDepTypeOptions = await multiselect({
					message: `Add workspace dependencies to ${accent(projectDir)}?`,
					options: [
						{ value: DependencyType.dependencies },
						{ value: DependencyType.devDependencies },
						{ value: DependencyType.peerDependencies },
						{ value: DependencyType.optionalDependencies }
					],
					required: false
				});
				let selectedDepTypes = [];
				if (!Ct(selectedDepTypeOptions)) selectedDepTypes = selectedDepTypeOptions;
				for (const selectedDepType of selectedDepTypes) {
					const selected = await multiselect({
						message: `Which packages should be added as ${selectedDepType} to ${success(projectDir)}?`,
						options: workspaceInfo.packages.map((pkg) => ({
							value: pkg.name,
							label: pkg.path
						})),
						required: false
					});
					let selectedDeps = [];
					if (!Ct(selected)) selectedDeps = selected;
					if (selectedDeps.length > 0) updatePackageJsonWithDeps(workspaceInfo.rootDir, projectDir, selectedDeps, selectedDepType);
				}
				resumeCreateProgress();
			}
		}
		updateWorkspaceConfig(projectDir, workspaceInfo);
		updateCreateProgress("Installing dependencies");
		installSummary = await runViteInstall(workspaceInfo.rootDir, options.interactive, void 0, { silent: compactOutput });
		updateCreateProgress("Formatting code");
		await runViteFmt(workspaceInfo.rootDir, options.interactive, [projectDir], { silent: compactOutput });
	} else {
		updateCreateProgress("Applying Vite+ project setup");
		rewriteStandaloneProject(fullPath, workspaceInfo, void 0, compactOutput);
		if (shouldSetupHooks) installGitHooks(fullPath, compactOutput);
		updateCreateProgress("Installing dependencies");
		installSummary = await runViteInstall(fullPath, options.interactive, void 0, { silent: compactOutput });
		updateCreateProgress("Formatting code");
		await runViteFmt(fullPath, options.interactive, void 0, { silent: compactOutput });
	}
	clearCreateProgress();
	showCreateSummary({
		description: describeScaffold(selectedTemplateName, selectedTemplateArgs),
		installSummary,
		nextCommand: isMonorepo ? `vp dev ${projectDir}` : getNextCommand(projectDir, selectedTemplateName === BuiltinTemplate.library ? "vp run dev" : "vp dev"),
		packageManager: workspaceInfo.packageManager,
		packageManagerVersion: workspaceInfo.downloadPackageManager.version,
		projectDir
	});
}
async function showAvailableTemplates() {
	log$1(vitePlusHeader() + "\n");
	log$1(listTemplatesMessage);
}
main().catch((err) => {
	log.error(err.message);
	console.error(err);
	cancelAndExit(`Failed to generate code: ${err.message}`, 1);
});
//#endregion
