// vite-plus@0.1.12/dist/global/package-YAMvX5PJ.js
import { r as readJsonFile } from "./json-BRdVJ52a.js";
import path from "node:path";
import fs from "node:fs";
import { createRequire } from "node:module";
//#region src/utils/package.ts
function getScopeFromPackageName(packageName) {
	if (packageName.startsWith("@")) return packageName.split("/")[0];
	return "";
}
function detectPackageMetadata(projectPath, packageName) {
	try {
		const pkgFilePath = createRequire(path.join(projectPath, "noop.js")).resolve(`${packageName}/package.json`);
		const pkg = JSON.parse(fs.readFileSync(pkgFilePath, "utf8"));
		return {
			name: pkg.name,
			version: pkg.version,
			path: path.dirname(pkgFilePath)
		};
	} catch {
		return;
	}
}
/**
* Read the nearest package.json file from the current directory up to the root directory.
* @param currentDir - The current directory to start searching from.
* @returns The package.json content as a JSON object, or null if no package.json is found.
*/
function readNearestPackageJson(currentDir) {
	do {
		const packageJsonPath = path.join(currentDir, "package.json");
		if (fs.existsSync(packageJsonPath)) return readJsonFile(packageJsonPath);
		currentDir = path.dirname(currentDir);
	} while (currentDir !== path.dirname(currentDir));
	return null;
}
function hasVitePlusDependency(pkg) {
	return Boolean(pkg?.dependencies?.["vite-plus"] || pkg?.devDependencies?.["vite-plus"]);
}
//#endregion
export { readNearestPackageJson as i, getScopeFromPackageName as n, hasVitePlusDependency as r, detectPackageMetadata as t };
