// vite-plus@0.1.12/dist/global/version.js
import { s as VITE_PLUS_NAME } from "./json-BRdVJ52a.js";
import { r as hasVitePlusDependency, t as detectPackageMetadata } from "./package-YAMvX5PJ.js";
import { i as log, n as accent, t as renderCliDoc } from "./help-CbTzUdXc.js";
import path from "node:path";
import fs from "node:fs";
import { vitePlusHeader } from "../../binding/index.js";
import { createRequire } from "node:module";
//#region src/version.ts
const require = createRequire(import.meta.url);
function getGlobalVersion() {
	return process.env.VITE_PLUS_GLOBAL_VERSION ?? null;
}
function getCliVersion() {
	return resolvePackageJson(VITE_PLUS_NAME, process.cwd())?.version ?? null;
}
function getLocalMetadata(cwd) {
	if (!isVitePlusDeclaredInAncestors(cwd)) return null;
	return detectPackageMetadata(cwd, "vite-plus") ?? null;
}
function isVitePlusDeclaredInAncestors(cwd) {
	let currentDir = path.resolve(cwd);
	while (true) {
		const pkg = readPackageJsonFromPath(path.join(currentDir, "package.json"));
		if (pkg && hasVitePlusDependency(pkg)) return true;
		const parentDir = path.dirname(currentDir);
		if (parentDir === currentDir) break;
		currentDir = parentDir;
	}
	return false;
}
function readPackageJsonFromPath(packageJsonPath) {
	try {
		return JSON.parse(fs.readFileSync(packageJsonPath, "utf8"));
	} catch {
		return null;
	}
}
function resolvePackageJson(packageName, baseDir) {
	try {
		return readPackageJsonFromPath(require.resolve(`${packageName}/package.json`, { paths: [baseDir] }));
	} catch {
		try {
			const mainPath = require.resolve(packageName, { paths: [baseDir] });
			let dir = path.dirname(mainPath);
			while (dir !== path.dirname(dir)) {
				const pkg = readPackageJsonFromPath(path.join(dir, "package.json"));
				if (pkg) return pkg;
				dir = path.dirname(dir);
			}
		} catch {}
		return null;
	}
}
function resolveToolVersion(tool, localPackagePath) {
	const pkg = resolvePackageJson(tool.packageName, localPackagePath);
	const bundledVersion = tool.bundledVersionKey ? pkg?.bundledVersions?.[tool.bundledVersionKey] ?? null : null;
	if (bundledVersion) return bundledVersion;
	const version = pkg?.version ?? null;
	if (version) return version;
	if (tool.fallbackPackageJson) return readPackageJsonFromPath(path.join(localPackagePath, tool.fallbackPackageJson))?.version ?? null;
	return null;
}
/**
* Print version information
*/
async function printVersion(cwd) {
	const globalVersion = getGlobalVersion();
	const cliVersion = getCliVersion();
	const localMetadata = getLocalMetadata(cwd);
	const localVersion = localMetadata?.version ?? null;
	const vpVersion = globalVersion ?? cliVersion ?? localVersion ?? "unknown";
	log(vitePlusHeader());
	log("");
	log(`vp v${vpVersion}\n`);
	const sections = [{
		title: "Local vite-plus",
		rows: [{
			label: accent("vite-plus"),
			description: localVersion ? `v${localVersion}` : "Not found"
		}]
	}];
	const tools = [
		{
			displayName: "vite",
			packageName: "@voidzero-dev/vite-plus-core",
			bundledVersionKey: "vite"
		},
		{
			displayName: "rolldown",
			packageName: "@voidzero-dev/vite-plus-core",
			bundledVersionKey: "rolldown"
		},
		{
			displayName: "vitest",
			packageName: "@voidzero-dev/vite-plus-test",
			bundledVersionKey: "vitest"
		},
		{
			displayName: "oxfmt",
			packageName: "oxfmt"
		},
		{
			displayName: "oxlint",
			packageName: "oxlint"
		},
		{
			displayName: "oxlint-tsgolint",
			packageName: "oxlint-tsgolint"
		},
		{
			displayName: "tsdown",
			packageName: "@voidzero-dev/vite-plus-core",
			bundledVersionKey: "tsdown"
		}
	];
	if (localMetadata) {
		const resolvedTools = tools.map((tool) => ({
			tool,
			version: resolveToolVersion(tool, localMetadata.path)
		}));
		sections.push({
			title: "Tools",
			rows: resolvedTools.map(({ tool, version }) => ({
				label: accent(tool.displayName),
				description: version ? `v${version}` : "Not found"
			}))
		});
	}
	log(renderCliDoc({ sections }));
}
await printVersion(process.cwd());
//#endregion
export { printVersion };
